{
    "requirements": [],
    "sbg:toolkit": "VarScan2",
    "sbg:validationErrors": [],
    "sbg:categories": [
        "Variant-Calling"
    ],
    "class": "Workflow",
    "label": "VarScan2 Workflow from BAM",
    "sbg:createdBy": "bixqa",
    "sbg:image_url": "https://brood.sbgenomics.com/static/bixqa/qa-load-2017-07-31-18/varscan2-workflow-from-bam-v2-3-9/29.png",
    "cwlVersion": "sbg:draft-2",
    "sbg:revisionsInfo": [
        {
            "sbg:revision": 0,
            "sbg:revisionNotes": null,
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 1,
            "sbg:revisionNotes": null,
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 2,
            "sbg:revisionNotes": null,
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 3,
            "sbg:revisionNotes": null,
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 4,
            "sbg:revisionNotes": null,
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 5,
            "sbg:revisionNotes": null,
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 6,
            "sbg:revisionNotes": null,
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 7,
            "sbg:revisionNotes": null,
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 8,
            "sbg:revisionNotes": null,
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 9,
            "sbg:revisionNotes": null,
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 10,
            "sbg:revisionNotes": null,
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 11,
            "sbg:revisionNotes": null,
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 12,
            "sbg:revisionNotes": null,
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 13,
            "sbg:revisionNotes": "Added suggested file.",
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 14,
            "sbg:revisionNotes": "Updated All tools.",
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 15,
            "sbg:revisionNotes": "SAMtools Index FASTA check box",
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 16,
            "sbg:revisionNotes": null,
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 17,
            "sbg:revisionNotes": "update faidx from public tools",
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 18,
            "sbg:revisionNotes": "update pileup pipeline",
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 19,
            "sbg:revisionNotes": "Add description",
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 20,
            "sbg:revisionNotes": null,
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 21,
            "sbg:revisionNotes": "Add common issues",
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 22,
            "sbg:revisionNotes": "disable BAQ calculations",
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 23,
            "sbg:revisionNotes": "update description",
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 24,
            "sbg:revisionNotes": "Add filter to mpileup tool to fitler zero-coverage lines",
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 25,
            "sbg:revisionNotes": "update pileup parameters to disable BAQ computation",
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 26,
            "sbg:revisionNotes": "Add filtering option in description",
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 27,
            "sbg:revisionNotes": "Updated all tools and fixed issue with metadata propagation.",
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 28,
            "sbg:revisionNotes": "fix metadata issue",
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        },
        {
            "sbg:revision": 29,
            "sbg:revisionNotes": "fix metadata issue when it is empty",
            "sbg:modifiedBy": "bixqa",
            "sbg:modifiedOn": 1501518730
        }
    ],
    "sbg:modifiedOn": 1501518730,
    "sbg:license": "Non-Profit Open Software License 3.0 (Non-Profit OSL 3.0)",
    "sbg:canvas_x": 41,
    "inputs": [
        {
            "sbg:x": 125.42502593994145,
            "label": "input_fasta_file",
            "sbg:suggestedValue": {
                "path": "5772b6d8507c1752674486e6",
                "name": "human_g1k_v37_decoy.fasta",
                "class": "File"
            },
            "sbg:fileTypes": "FASTA,FA,GZ",
            "id": "#input_fasta_file",
            "sbg:y": 359.2647399902345,
            "type": [
                "File"
            ]
        },
        {
            "type": [
                "File"
            ],
            "sbg:x": 137.98312704903555,
            "label": "#Tumor_BAM",
            "sbg:fileTypes": "BAM",
            "id": "#Tumor_BAM",
            "sbg:y": 103.4453937942539,
            "sbg:includeInPorts": true
        },
        {
            "type": [
                "File"
            ],
            "sbg:x": 123.05578613281253,
            "label": "#Normal_BAM",
            "sbg:fileTypes": "BAM",
            "id": "#Normal_BAM",
            "sbg:y": 463.63250732421886,
            "sbg:includeInPorts": true
        }
    ],
    "hints": [
        {
            "value": "c3.8xlarge",
            "class": "sbg:AWSInstanceType"
        },
        {
            "value": "n1-standard-32;pd-ssd;4096",
            "class": "sbg:GoogleInstanceType"
        }
    ],
    "sbg:revision": 29,
    "id": "https://api.sbgenomics.com/v2/apps/bixqa/qa-load-2017-07-31-18/varscan2-workflow-from-bam-v2-3-9/29/raw/",
    "sbg:latestRevision": 29,
    "steps": [
        {
            "sbg:x": 1833.4655493667708,
            "outputs": [
                {
                    "id": "#SBG_VCF_Reorder.reordered_vcf"
                }
            ],
            "run": {
                "sbg:toolkit": "SBGTools",
                "sbg:contributors": [
                    "bix-demo"
                ],
                "requirements": [],
                "sbg:categories": [
                    "VCF-Processing"
                ],
                "label": "SBG VCF Reorder",
                "sbg:createdBy": "bix-demo",
                "sbg:cmdPreview": "python /opt/sbg_vcf_reorder.py -v vcf.ext -r reference.ext -s species -m 0 vcf.ext reference.ext",
                "arguments": [],
                "cwlVersion": "sbg:draft-2",
                "sbg:validationErrors": [],
                "y": 309.31051458093856,
                "sbg:revisionsInfo": [
                    {
                        "sbg:revision": 0,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911304
                    },
                    {
                        "sbg:revision": 1,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911305
                    }
                ],
                "sbg:license": "Apache License 2.0",
                "class": "CommandLineTool",
                "inputs": [
                    {
                        "label": "Vcf",
                        "required": true,
                        "inputBinding": {
                            "prefix": "-v",
                            "sbg:cmdInclude": true,
                            "separate": true
                        },
                        "description": "A VCF file for reordering.",
                        "id": "#vcf",
                        "type": [
                            "File"
                        ]
                    },
                    {
                        "label": "Species",
                        "required": false,
                        "inputBinding": {
                            "prefix": "-s",
                            "sbg:cmdInclude": true,
                            "separate": true
                        },
                        "description": "Species name.",
                        "id": "#species",
                        "type": [
                            "null",
                            "string"
                        ]
                    },
                    {
                        "label": "Reference",
                        "required": true,
                        "inputBinding": {
                            "prefix": "-r",
                            "sbg:cmdInclude": true,
                            "separate": true
                        },
                        "description": "Reference genome in FASTA format.",
                        "id": "#reference",
                        "type": [
                            "File"
                        ]
                    },
                    {
                        "label": "Memory",
                        "required": false,
                        "inputBinding": {
                            "prefix": "-m",
                            "sbg:cmdInclude": true,
                            "separate": true
                        },
                        "description": "Amount of memory in GB to allocate for the process.",
                        "id": "#memory",
                        "type": [
                            "null",
                            "int"
                        ]
                    }
                ],
                "hints": [
                    {
                        "dockerImageId": "",
                        "class": "DockerRequirement",
                        "dockerPull": "images.sbgenomics.com/tziotas/sbg_vcf_reorder:1.0.3"
                    },
                    {
                        "value": 1,
                        "class": "sbg:CPURequirement"
                    },
                    {
                        "value": 4096,
                        "class": "sbg:MemRequirement"
                    }
                ],
                "sbg:revision": 1,
                "sbg:modifiedOn": 1450911305,
                "id": "bix-demo/sbgtools-demo/sbg-vcf-reorder/1",
                "sbg:latestRevision": 1,
                "temporaryFailCodes": [],
                "successCodes": [],
                "sbg:toolAuthor": "Seven Bridges Genomics",
                "outputs": [
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Reordered VCF",
                        "outputBinding": {
                            "glob": "*.reordered.vcf",
                            "sbg:metadata": {},
                            "sbg:inheritMetadataFrom": "#vcf"
                        },
                        "sbg:fileTypes": "VCF",
                        "id": "#reordered_vcf",
                        "description": "Reordered VCF"
                    }
                ],
                "sbg:modifiedBy": "bix-demo",
                "sbg:id": "admin/sbg-public-data/sbg-vcf-reorder/1",
                "sbg:job": {
                    "allocatedResources": {
                        "cpu": 1,
                        "mem": 4096
                    },
                    "inputs": {
                        "vcf": {
                            "path": "vcf.ext",
                            "size": 0,
                            "class": "File",
                            "secondaryFiles": []
                        },
                        "reference": {
                            "path": "reference.ext",
                            "size": 0,
                            "class": "File",
                            "secondaryFiles": []
                        },
                        "memory": 0,
                        "species": "species"
                    }
                },
                "stdin": "",
                "sbg:project": "bix-demo/sbgtools-demo",
                "description": "VCF Reorder is a tool for re-ordering variants in VCF files by chromosome.",
                "x": 1833.4655493667708,
                "appUrl": "/u/bix-demo/sbgtools-demo/apps/#bix-demo/sbgtools-demo/sbg-vcf-reorder/1",
                "sbg:createdOn": 1450911304,
                "baseCommand": [
                    "python",
                    "/opt/sbg_vcf_reorder.py"
                ],
                "sbg:sbgMaintained": false,
                "stdout": ""
            },
            "inputs": [
                {
                    "id": "#SBG_VCF_Reorder.vcf",
                    "source": [
                        "#VCFtools_Sort.output_file"
                    ]
                },
                {
                    "id": "#SBG_VCF_Reorder.species"
                },
                {
                    "id": "#SBG_VCF_Reorder.reference",
                    "source": [
                        "#SAMtools_Index_FASTA.output_fasta_file"
                    ]
                },
                {
                    "id": "#SBG_VCF_Reorder.memory"
                }
            ],
            "id": "#SBG_VCF_Reorder",
            "sbg:y": 309.31051458093856
        },
        {
            "sbg:x": 1734.4195431372643,
            "outputs": [
                {
                    "id": "#SBG_VCF_Reorder_1.reordered_vcf"
                }
            ],
            "run": {
                "sbg:toolkit": "SBGTools",
                "requirements": [],
                "sbg:validationErrors": [],
                "sbg:categories": [
                    "VCF-Processing"
                ],
                "label": "SBG VCF Reorder",
                "sbg:createdBy": "bix-demo",
                "sbg:cmdPreview": "python /opt/sbg_vcf_reorder.py -v vcf.ext -r reference.ext -s species -m 0 vcf.ext reference.ext",
                "arguments": [],
                "sbg:revisionsInfo": [
                    {
                        "sbg:revision": 0,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911304
                    },
                    {
                        "sbg:revision": 1,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911305
                    }
                ],
                "y": 454.08287316889977,
                "sbg:modifiedOn": 1450911305,
                "sbg:license": "Apache License 2.0",
                "class": "CommandLineTool",
                "inputs": [
                    {
                        "label": "Vcf",
                        "required": true,
                        "inputBinding": {
                            "prefix": "-v",
                            "sbg:cmdInclude": true,
                            "separate": true
                        },
                        "description": "A VCF file for reordering.",
                        "id": "#vcf",
                        "type": [
                            "File"
                        ]
                    },
                    {
                        "inputBinding": {
                            "prefix": "-s",
                            "sbg:cmdInclude": true,
                            "separate": true
                        },
                        "description": "Species name.",
                        "label": "Species",
                        "id": "#species",
                        "type": [
                            "null",
                            "string"
                        ]
                    },
                    {
                        "label": "Reference",
                        "required": true,
                        "inputBinding": {
                            "prefix": "-r",
                            "sbg:cmdInclude": true,
                            "separate": true
                        },
                        "description": "Reference genome in FASTA format.",
                        "id": "#reference",
                        "type": [
                            "File"
                        ]
                    },
                    {
                        "inputBinding": {
                            "prefix": "-m",
                            "sbg:cmdInclude": true,
                            "separate": true
                        },
                        "description": "Amount of memory in GB to allocate for the process.",
                        "label": "Memory",
                        "id": "#memory",
                        "type": [
                            "null",
                            "int"
                        ]
                    }
                ],
                "hints": [
                    {
                        "dockerImageId": "",
                        "class": "DockerRequirement",
                        "dockerPull": "images.sbgenomics.com/tziotas/sbg_vcf_reorder:1.0.3"
                    },
                    {
                        "value": 1,
                        "class": "sbg:CPURequirement"
                    },
                    {
                        "value": 4096,
                        "class": "sbg:MemRequirement"
                    }
                ],
                "sbg:revision": 1,
                "baseCommand": [
                    "python",
                    "/opt/sbg_vcf_reorder.py"
                ],
                "id": "bix-demo/sbgtools-demo/sbg-vcf-reorder/1",
                "sbg:latestRevision": 1,
                "temporaryFailCodes": [],
                "successCodes": [],
                "sbg:toolAuthor": "Seven Bridges Genomics",
                "outputs": [
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Reordered VCF",
                        "outputBinding": {
                            "glob": "*.reordered.vcf",
                            "sbg:metadata": {},
                            "sbg:inheritMetadataFrom": "#vcf"
                        },
                        "sbg:fileTypes": "VCF",
                        "id": "#reordered_vcf",
                        "description": "Reordered VCF"
                    }
                ],
                "sbg:modifiedBy": "bix-demo",
                "sbg:id": "admin/sbg-public-data/sbg-vcf-reorder/1",
                "sbg:job": {
                    "allocatedResources": {
                        "cpu": 1,
                        "mem": 4096
                    },
                    "inputs": {
                        "vcf": {
                            "path": "vcf.ext",
                            "size": 0,
                            "class": "File",
                            "secondaryFiles": []
                        },
                        "reference": {
                            "path": "reference.ext",
                            "size": 0,
                            "class": "File",
                            "secondaryFiles": []
                        },
                        "memory": 0,
                        "species": "species"
                    }
                },
                "stdin": "",
                "sbg:project": "bix-demo/sbgtools-demo",
                "description": "VCF Reorder is a tool for re-ordering variants in VCF files by chromosome.",
                "x": 1734.4195431372643,
                "sbg:createdOn": 1450911304,
                "cwlVersion": "sbg:draft-2",
                "sbg:contributors": [
                    "bix-demo"
                ],
                "sbg:sbgMaintained": false,
                "stdout": ""
            },
            "inputs": [
                {
                    "id": "#SBG_VCF_Reorder_1.vcf",
                    "source": [
                        "#VCFtools_Sort_1.output_file"
                    ]
                },
                {
                    "id": "#SBG_VCF_Reorder_1.species"
                },
                {
                    "id": "#SBG_VCF_Reorder_1.reference",
                    "source": [
                        "#SAMtools_Index_FASTA.output_fasta_file"
                    ]
                },
                {
                    "id": "#SBG_VCF_Reorder_1.memory"
                }
            ],
            "id": "#SBG_VCF_Reorder_1",
            "sbg:y": 454.08287316889977
        },
        {
            "sbg:x": 297.47902680951967,
            "outputs": [
                {
                    "id": "#BamTools_Index_Tumor.output_bam_file"
                },
                {
                    "id": "#BamTools_Index_Tumor.generated_index"
                }
            ],
            "run": {
                "sbg:toolkit": "BamTools",
                "requirements": [
                    {
                        "requirements": [
                            {
                                "class": "DockerRequirement",
                                "dockerPull": "rabix/js-engine"
                            }
                        ],
                        "id": "#cwl-js-engine",
                        "class": "ExpressionEngineRequirement"
                    }
                ],
                "sbg:validationErrors": [],
                "sbg:categories": [
                    "SAM/BAM-Processing",
                    "Indexing"
                ],
                "label": "BamTools Index Tumor",
                "sbg:createdBy": "bix-demo",
                "sbg:cmdPreview": "/opt/bamtools/bin/bamtools index  -in input_bam.bam",
                "sbg:image_url": null,
                "sbg:links": [
                    {
                        "label": "Homepage",
                        "id": "https://github.com/pezmaster31/bamtools"
                    },
                    {
                        "label": "Wiki",
                        "id": "https://github.com/pezmaster31/bamtools/wiki"
                    }
                ],
                "sbg:revisionsInfo": [
                    {
                        "sbg:revision": 0,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911459
                    },
                    {
                        "sbg:revision": 1,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911459
                    },
                    {
                        "sbg:revision": 2,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1455724415
                    },
                    {
                        "sbg:revision": 3,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1455724618
                    },
                    {
                        "sbg:revision": 4,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1455726303
                    },
                    {
                        "sbg:revision": 5,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1456328383
                    },
                    {
                        "sbg:revision": 6,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1459177904
                    },
                    {
                        "sbg:revision": 7,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1462383521
                    },
                    {
                        "sbg:revision": 8,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1462383765
                    },
                    {
                        "sbg:revision": 9,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1462384650
                    },
                    {
                        "sbg:revision": 10,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1466421734
                    },
                    {
                        "sbg:revision": 11,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1466777070
                    },
                    {
                        "sbg:revision": 12,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1466777658
                    },
                    {
                        "sbg:revision": 13,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1466778871
                    },
                    {
                        "sbg:revision": 14,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1466779762
                    },
                    {
                        "sbg:revision": 15,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1467116120
                    },
                    {
                        "sbg:revision": 16,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1467119021
                    },
                    {
                        "sbg:revision": 17,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1467119608
                    },
                    {
                        "sbg:revision": 18,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1467119913
                    },
                    {
                        "sbg:revision": 19,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1467119936
                    },
                    {
                        "sbg:revision": 20,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1467210836
                    }
                ],
                "y": 128.1512895184353,
                "sbg:modifiedOn": 1467210836,
                "sbg:license": "The MIT License",
                "sbg:modifiedBy": "ognjenm",
                "class": "CommandLineTool",
                "inputs": [
                    {
                        "sbg:fileTypes": "BAM",
                        "sbg:stageInput": "link",
                        "label": "Input BAM file",
                        "required": true,
                        "description": "The input BAM file.",
                        "id": "#input_bam_file",
                        "sbg:category": "Input & Output",
                        "type": [
                            "File"
                        ]
                    },
                    {
                        "sbg:fileTypes": "BAI",
                        "sbg:stageInput": "link",
                        "label": "Input BAI(BAM index) file",
                        "required": false,
                        "description": "Input BAI(BAM index) file.",
                        "id": "#input_bai_file",
                        "sbg:category": "File input",
                        "type": [
                            "null",
                            "File"
                        ]
                    },
                    {
                        "description": "Don't output indexed data file. The default value is [FALSE].",
                        "id": "#indexed_data",
                        "label": "Don't output indexed data file",
                        "sbg:category": "Input & Output",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "description": "Create (non-standard) BamTools index file (*.BTI). Default behavior is to create standard BAM index (*.BAI).",
                        "id": "#bti_format",
                        "label": "BTI format",
                        "sbg:category": "Input & Output",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    }
                ],
                "hints": [
                    {
                        "dockerImageId": "f808163d4cd3",
                        "class": "DockerRequirement",
                        "dockerPull": "images.sbgenomics.com/markop/bamtools:2.4.0"
                    },
                    {
                        "value": 1,
                        "class": "sbg:CPURequirement"
                    },
                    {
                        "value": 1000,
                        "class": "sbg:MemRequirement"
                    }
                ],
                "sbg:revision": 20,
                "baseCommand": [
                    {
                        "script": "{\n  if ($job.inputs.input_bai_file){\n \treturn\"echo Skipping index step because BAI file is provided on the input.\"\n  }\n  else{\n    return \"/opt/bamtools/bin/bamtools index\"\n  }\n}",
                        "engine": "#cwl-js-engine",
                        "class": "Expression"
                    }
                ],
                "id": "bix-demo/bamtools-2-4-0-demo/bamtools-index-2-4-0/20",
                "sbg:latestRevision": 20,
                "temporaryFailCodes": [],
                "sbg:toolkitVersion": "2.4.0",
                "sbg:toolAuthor": "Derek Barnett, Erik Garrison, Gabor Marth, and Michael Stromberg",
                "successCodes": [],
                "outputs": [
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Output BAM file",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  if ($job.inputs.indexed_data === true){\n\treturn ''\n  } else {\n    return $job.inputs.input_bam_file.path.split(\"/\").pop()\n  }\n}\n\n\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "secondaryFiles": [
                                ".bai",
                                ".bti"
                            ],
                            "sbg:inheritMetadataFrom": "#input_bam_file"
                        },
                        "description": "Output BAM file with index (BAI or BTI) file as a secondary file (not visible in the output list of the task).",
                        "fileTypes": "BAM,BAI,BTI",
                        "id": "#output_bam_file",
                        "sbg:fileTypes": "BAM"
                    },
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Generated index file",
                        "outputBinding": {
                            "glob": "{*.bai,*.bti}",
                            "sbg:inheritMetadataFrom": "#input_bam_file"
                        },
                        "fileTypes": "BAI,BTI",
                        "id": "#generated_index",
                        "description": "Generated index file (without the indexed data)."
                    }
                ],
                "arguments": [
                    {
                        "position": 0,
                        "valueFrom": {
                            "script": "{\n  if (!(typeof $job.inputs.input_bai_file !== \"undefined\")){\n    return \"-in \" + $job.inputs.input_bam_file.path.split('/').pop() \n  }\n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "separate": true
                    },
                    {
                        "position": 1,
                        "valueFrom": {
                            "script": "{\n  if (!(typeof $job.inputs.input_bai_file !== \"undefined\"))\n    if ((typeof $job.inputs.bti_format !== \"undefined\") && ($job.inputs.bti_format))\n    \treturn \"-bti\"  \n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "separate": true
                    }
                ],
                "sbg:id": "admin/sbg-public-data/bamtools-index-2-4-0/30",
                "sbg:job": {
                    "allocatedResources": {
                        "cpu": 1,
                        "mem": 1000
                    },
                    "inputs": {
                        "bti_format": false,
                        "indexed_data": false,
                        "input_bam_file": {
                            "path": "input/input_bam.bam",
                            "size": 0,
                            "class": "File",
                            "secondaryFiles": []
                        }
                    }
                },
                "stdin": "",
                "sbg:homepage": "https://github.com/pezmaster31/bamtools/wiki",
                "sbg:project": "bix-demo/bamtools-2-4-0-demo",
                "description": "BamTools Index creates an index file (BAI or BTI) for a BAM file. If BAI file is present on the input the tool will skip indexing step and output BAM with provided BAI file.\n\n**Common issues:** Providing a BAI file on input will result in a pass-through without execution, even if a different index format is requested on the output (BTI instead of BAI).",
                "x": 297.47902680951967,
                "sbg:createdOn": 1450911459,
                "cwlVersion": "sbg:draft-2",
                "sbg:contributors": [
                    "bix-demo",
                    "markop",
                    "ognjenm"
                ],
                "sbg:sbgMaintained": false,
                "stdout": ""
            },
            "inputs": [
                {
                    "id": "#BamTools_Index_Tumor.input_bam_file",
                    "source": [
                        "#Tumor_BAM"
                    ]
                },
                {
                    "id": "#BamTools_Index_Tumor.input_bai_file"
                },
                {
                    "id": "#BamTools_Index_Tumor.indexed_data"
                },
                {
                    "id": "#BamTools_Index_Tumor.bti_format"
                }
            ],
            "id": "#BamTools_Index_Tumor",
            "sbg:y": 128.1512895184353
        },
        {
            "sbg:x": 320.6301555060781,
            "outputs": [
                {
                    "id": "#BamTools_Index_Normal.output_bam_file"
                },
                {
                    "id": "#BamTools_Index_Normal.generated_index"
                }
            ],
            "run": {
                "sbg:toolkit": "BamTools",
                "requirements": [
                    {
                        "requirements": [
                            {
                                "class": "DockerRequirement",
                                "dockerPull": "rabix/js-engine"
                            }
                        ],
                        "id": "#cwl-js-engine",
                        "class": "ExpressionEngineRequirement"
                    }
                ],
                "sbg:validationErrors": [],
                "sbg:categories": [
                    "SAM/BAM-Processing",
                    "Indexing"
                ],
                "label": "BamTools Index Normal",
                "sbg:createdBy": "bix-demo",
                "sbg:cmdPreview": "/opt/bamtools/bin/bamtools index  -in input_bam.bam",
                "sbg:image_url": null,
                "sbg:links": [
                    {
                        "label": "Homepage",
                        "id": "https://github.com/pezmaster31/bamtools"
                    },
                    {
                        "label": "Wiki",
                        "id": "https://github.com/pezmaster31/bamtools/wiki"
                    }
                ],
                "sbg:revisionsInfo": [
                    {
                        "sbg:revision": 0,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911459
                    },
                    {
                        "sbg:revision": 1,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911459
                    },
                    {
                        "sbg:revision": 2,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1455724415
                    },
                    {
                        "sbg:revision": 3,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1455724618
                    },
                    {
                        "sbg:revision": 4,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1455726303
                    },
                    {
                        "sbg:revision": 5,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1456328383
                    },
                    {
                        "sbg:revision": 6,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1459177904
                    },
                    {
                        "sbg:revision": 7,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1462383521
                    },
                    {
                        "sbg:revision": 8,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1462383765
                    },
                    {
                        "sbg:revision": 9,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1462384650
                    },
                    {
                        "sbg:revision": 10,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1466421734
                    },
                    {
                        "sbg:revision": 11,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1466777070
                    },
                    {
                        "sbg:revision": 12,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1466777658
                    },
                    {
                        "sbg:revision": 13,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1466778871
                    },
                    {
                        "sbg:revision": 14,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1466779762
                    },
                    {
                        "sbg:revision": 15,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1467116120
                    },
                    {
                        "sbg:revision": 16,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1467119021
                    },
                    {
                        "sbg:revision": 17,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1467119608
                    },
                    {
                        "sbg:revision": 18,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1467119913
                    },
                    {
                        "sbg:revision": 19,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1467119936
                    },
                    {
                        "sbg:revision": 20,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1467210836
                    }
                ],
                "y": 488.79836145665394,
                "sbg:modifiedOn": 1467210836,
                "sbg:license": "The MIT License",
                "sbg:modifiedBy": "ognjenm",
                "class": "CommandLineTool",
                "inputs": [
                    {
                        "sbg:fileTypes": "BAM",
                        "sbg:stageInput": "link",
                        "label": "Input BAM file",
                        "required": true,
                        "description": "The input BAM file.",
                        "id": "#input_bam_file",
                        "sbg:category": "Input & Output",
                        "type": [
                            "File"
                        ]
                    },
                    {
                        "sbg:fileTypes": "BAI",
                        "sbg:stageInput": "link",
                        "label": "Input BAI(BAM index) file",
                        "required": false,
                        "description": "Input BAI(BAM index) file.",
                        "id": "#input_bai_file",
                        "sbg:category": "File input",
                        "type": [
                            "null",
                            "File"
                        ]
                    },
                    {
                        "description": "Don't output indexed data file. The default value is [FALSE].",
                        "id": "#indexed_data",
                        "label": "Don't output indexed data file",
                        "sbg:category": "Input & Output",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "description": "Create (non-standard) BamTools index file (*.BTI). Default behavior is to create standard BAM index (*.BAI).",
                        "id": "#bti_format",
                        "label": "BTI format",
                        "sbg:category": "Input & Output",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    }
                ],
                "hints": [
                    {
                        "dockerImageId": "f808163d4cd3",
                        "class": "DockerRequirement",
                        "dockerPull": "images.sbgenomics.com/markop/bamtools:2.4.0"
                    },
                    {
                        "value": 1,
                        "class": "sbg:CPURequirement"
                    },
                    {
                        "value": 1000,
                        "class": "sbg:MemRequirement"
                    }
                ],
                "sbg:revision": 20,
                "baseCommand": [
                    {
                        "script": "{\n  if ($job.inputs.input_bai_file){\n \treturn\"echo Skipping index step because BAI file is provided on the input.\"\n  }\n  else{\n    return \"/opt/bamtools/bin/bamtools index\"\n  }\n}",
                        "engine": "#cwl-js-engine",
                        "class": "Expression"
                    }
                ],
                "id": "bix-demo/bamtools-2-4-0-demo/bamtools-index-2-4-0/20",
                "sbg:latestRevision": 20,
                "temporaryFailCodes": [],
                "sbg:toolkitVersion": "2.4.0",
                "sbg:toolAuthor": "Derek Barnett, Erik Garrison, Gabor Marth, and Michael Stromberg",
                "successCodes": [],
                "outputs": [
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Output BAM file",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  if ($job.inputs.indexed_data === true){\n\treturn ''\n  } else {\n    return $job.inputs.input_bam_file.path.split(\"/\").pop()\n  }\n}\n\n\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "secondaryFiles": [
                                ".bai",
                                ".bti"
                            ],
                            "sbg:inheritMetadataFrom": "#input_bam_file"
                        },
                        "description": "Output BAM file with index (BAI or BTI) file as a secondary file (not visible in the output list of the task).",
                        "fileTypes": "BAM,BAI,BTI",
                        "id": "#output_bam_file",
                        "sbg:fileTypes": "BAM"
                    },
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Generated index file",
                        "outputBinding": {
                            "glob": "{*.bai,*.bti}",
                            "sbg:inheritMetadataFrom": "#input_bam_file"
                        },
                        "fileTypes": "BAI,BTI",
                        "id": "#generated_index",
                        "description": "Generated index file (without the indexed data)."
                    }
                ],
                "arguments": [
                    {
                        "position": 0,
                        "valueFrom": {
                            "script": "{\n  if (!(typeof $job.inputs.input_bai_file !== \"undefined\")){\n    return \"-in \" + $job.inputs.input_bam_file.path.split('/').pop() \n  }\n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "separate": true
                    },
                    {
                        "position": 1,
                        "valueFrom": {
                            "script": "{\n  if (!(typeof $job.inputs.input_bai_file !== \"undefined\"))\n    if ((typeof $job.inputs.bti_format !== \"undefined\") && ($job.inputs.bti_format))\n    \treturn \"-bti\"  \n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "separate": true
                    }
                ],
                "sbg:id": "admin/sbg-public-data/bamtools-index-2-4-0/30",
                "sbg:job": {
                    "allocatedResources": {
                        "cpu": 1,
                        "mem": 1000
                    },
                    "inputs": {
                        "bti_format": false,
                        "indexed_data": false,
                        "input_bam_file": {
                            "path": "input/input_bam.bam",
                            "size": 0,
                            "class": "File",
                            "secondaryFiles": []
                        }
                    }
                },
                "stdin": "",
                "sbg:homepage": "https://github.com/pezmaster31/bamtools/wiki",
                "sbg:project": "bix-demo/bamtools-2-4-0-demo",
                "description": "BamTools Index creates an index file (BAI or BTI) for a BAM file. If BAI file is present on the input the tool will skip indexing step and output BAM with provided BAI file.\n\n**Common issues:** Providing a BAI file on input will result in a pass-through without execution, even if a different index format is requested on the output (BTI instead of BAI).",
                "x": 320.6301555060781,
                "sbg:createdOn": 1450911459,
                "cwlVersion": "sbg:draft-2",
                "sbg:contributors": [
                    "bix-demo",
                    "markop",
                    "ognjenm"
                ],
                "sbg:sbgMaintained": false,
                "stdout": ""
            },
            "inputs": [
                {
                    "id": "#BamTools_Index_Normal.input_bam_file",
                    "source": [
                        "#Normal_BAM"
                    ]
                },
                {
                    "id": "#BamTools_Index_Normal.input_bai_file"
                },
                {
                    "id": "#BamTools_Index_Normal.indexed_data"
                },
                {
                    "id": "#BamTools_Index_Normal.bti_format"
                }
            ],
            "id": "#BamTools_Index_Normal",
            "sbg:y": 488.79836145665394
        },
        {
            "scatter": "#VarScan2_ProcessSomatic_SNPs.status_file",
            "inputs": [
                {
                    "id": "#VarScan2_ProcessSomatic_SNPs.status_file",
                    "source": [
                        "#VarScan2_Filter_1.filtered_snps"
                    ]
                },
                {
                    "id": "#VarScan2_ProcessSomatic_SNPs.p_value"
                },
                {
                    "id": "#VarScan2_ProcessSomatic_SNPs.min_tumor_freq"
                },
                {
                    "id": "#VarScan2_ProcessSomatic_SNPs.max_normal_freq"
                }
            ],
            "outputs": [
                {
                    "id": "#VarScan2_ProcessSomatic_SNPs.somatic_hc"
                },
                {
                    "id": "#VarScan2_ProcessSomatic_SNPs.somatic"
                },
                {
                    "id": "#VarScan2_ProcessSomatic_SNPs.germline_hc"
                },
                {
                    "id": "#VarScan2_ProcessSomatic_SNPs.germline"
                },
                {
                    "id": "#VarScan2_ProcessSomatic_SNPs.LOH_hc"
                },
                {
                    "id": "#VarScan2_ProcessSomatic_SNPs.LOH"
                }
            ],
            "sbg:x": 1303.5919555100813,
            "run": {
                "sbg:toolkit": "VarScan",
                "requirements": [
                    {
                        "requirements": [
                            {
                                "class": "DockerRequirement",
                                "dockerPull": "rabix/js-engine"
                            }
                        ],
                        "id": "#cwl-js-engine",
                        "class": "ExpressionEngineRequirement"
                    }
                ],
                "sbg:validationErrors": [],
                "sbg:categories": [
                    "VCF-Processing"
                ],
                "label": "VarScan2 ProcessSomatic SNPs",
                "sbg:createdBy": "bix-demo",
                "sbg:cmdPreview": "java -jar -Xmx2048m /opt/VarScan2.3.9.jar processSomatic",
                "sbg:image_url": null,
                "sbg:links": [
                    {
                        "label": "Homepage",
                        "id": "http://varscan.sourceforge.net/using-varscan.html"
                    },
                    {
                        "label": "Source Code",
                        "id": "http://dkoboldt.github.io/varscan"
                    },
                    {
                        "label": "Wiki",
                        "id": "http://varscan.sourceforge.net/somatic-calling.html"
                    },
                    {
                        "label": "Download",
                        "id": "http://sourceforge.net/projects/varscan/files/VarScan.v2.3.9.jar/download"
                    },
                    {
                        "label": "Publication",
                        "id": "http://www.ncbi.nlm.nih.gov/pubmed/19542151"
                    },
                    {
                        "label": "Documentation",
                        "id": "http://varscan.sourceforge.net/doc/index.html"
                    }
                ],
                "sbg:revisionsInfo": [
                    {
                        "sbg:revision": 0,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911531
                    },
                    {
                        "sbg:revision": 1,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911533
                    },
                    {
                        "sbg:revision": 2,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1467636280
                    },
                    {
                        "sbg:revision": 3,
                        "sbg:revisionNotes": "Update PR",
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1468328200
                    },
                    {
                        "sbg:revision": 4,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "djordje_klisic",
                        "sbg:modifiedOn": 1469457737
                    },
                    {
                        "sbg:revision": 5,
                        "sbg:revisionNotes": "Staged",
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1469550974
                    }
                ],
                "y": 295.4100603842338,
                "sbg:modifiedOn": 1469550974,
                "sbg:license": "Non-Profit Open Software License 3.0",
                "sbg:modifiedBy": "markop",
                "class": "CommandLineTool",
                "inputs": [
                    {
                        "sbg:fileTypes": "VCF, TXT",
                        "sbg:stageInput": "link",
                        "type": [
                            "null",
                            "File"
                        ],
                        "required": false,
                        "inputBinding": {
                            "sbg:cmdInclude": true,
                            "position": -1,
                            "separate": true
                        },
                        "description": "The VarScan output file for SNPs or INDELs.",
                        "id": "#status_file",
                        "sbg:category": "File type inputs",
                        "label": "The VarScan output file for SNPs or INDELs"
                    },
                    {
                        "label": "P-value for high-confidence calling",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--p-value",
                            "sbg:cmdInclude": true,
                            "position": 5,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "0.07",
                        "description": "P-value for high-confidence calling.",
                        "id": "#p_value",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "float"
                        ]
                    },
                    {
                        "label": "Minimum variant allele frequency in tumor",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--min-tumor-freq",
                            "sbg:cmdInclude": true,
                            "position": 3,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "0.10",
                        "description": "Minimum variant allele frequency in tumor.",
                        "id": "#min_tumor_freq",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "float"
                        ]
                    },
                    {
                        "label": "Maximum variant allele frequency in normal",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--max-normal-freq",
                            "sbg:cmdInclude": true,
                            "position": 4,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "0.05",
                        "description": "Maximum variant allele frequency in normal.",
                        "id": "#max_normal_freq",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "float"
                        ]
                    }
                ],
                "hints": [
                    {
                        "dockerImageId": "e9fc75d12305",
                        "class": "DockerRequirement",
                        "dockerPull": "images.sbgenomics.com/djordje_klisic/varscan2:2.3.9"
                    },
                    {
                        "value": 1,
                        "class": "sbg:CPURequirement"
                    },
                    {
                        "value": {
                            "script": "{\n  if ($job.inputs.mem_mb)\n  {\n    return $job.inputs.mem_mb\n  }\n  \n  else\n  {\n    return 2048\n  }\n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "class": "sbg:MemRequirement"
                    }
                ],
                "sbg:revision": 5,
                "id": "bix-demo/varscan2-v2-3-9-demo/varscan2-processsomatic-v2-3-9/5",
                "sbg:latestRevision": 5,
                "temporaryFailCodes": [],
                "sbg:toolkitVersion": "2.3.9",
                "sbg:toolAuthor": "Genome Institute at Washington University",
                "successCodes": [],
                "sbg:revisionNotes": "Staged",
                "arguments": [],
                "appUrl": "/u/bix-demo/varscan2-v2-3-9-demo/apps/#bix-demo/varscan2-v2-3-9-demo/varscan2-processsomatic-v2-3-9/5",
                "baseCommand": [
                    "java",
                    "-jar",
                    {
                        "script": "{\n  if ($job.inputs.mem_mb)\n  {\n    return \"-Xmx\" + ($job.inputs.mem_mb).toString() + \"m\"\n  }\n  else\n  {\n    return \"-Xmx2048m\"\n  }\n}",
                        "engine": "#cwl-js-engine",
                        "class": "Expression"
                    },
                    "/opt/VarScan2.3.9.jar",
                    "processSomatic"
                ],
                "stdin": "",
                "sbg:project": "bix-demo/varscan2-v2-3-9-demo",
                "description": "The VarScan2 ProcessSomatic tool allow user to segregate different types of variants within the VCF file produced by the VarScan2 Somatic tool. Variants are segregated into three groups: germline, LOH (loss of heterozygosity), and somatic. Each of these groups will have two files on the tool output: one contains all variants from this group; the second is the subset of these variants that is considered high confidence. The output format of the tool is VCF by default.\n\n**The script will not work if the input is a list of \"called SNPs\" in a Varscan format.**",
                "x": 1303.5919555100813,
                "sbg:id": "admin/sbg-public-data/varscan2-processsomatic-v2-3-9/5",
                "sbg:createdOn": 1450911531,
                "sbg:job": {
                    "allocatedResources": {
                        "cpu": 1,
                        "mem": 2048
                    },
                    "inputs": {
                        "min_tumor_freq": 0,
                        "p_value": 0,
                        "max_normal_freq": 0,
                        "status_file": {
                            "path": "input/status_file.ext",
                            "size": 0,
                            "class": "File",
                            "secondaryFiles": []
                        }
                    }
                },
                "outputs": [
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "High confidence somatic variants",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  filename = $job.inputs.status_file.path\n  basename = filename.split('.').slice(0, filename.split('.').length-1).join('.')\n\n  return basename+'.Somatic.hc.vcf'\n}\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:inheritMetadataFrom": "#status_file"
                        },
                        "description": "File with high confidence variants that are somatic by origin.",
                        "fileTypes": "VCF",
                        "id": "#somatic_hc",
                        "sbg:fileTypes": "VCF, TXT"
                    },
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Somatic variants",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  filename = $job.inputs.status_file.path\n  basename = filename.split('.').slice(0, filename.split('.').length-1).join('.')\n\n  return basename+'.Somatic.vcf'\n}\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:inheritMetadataFrom": "#status_file"
                        },
                        "description": "File with variants that are somatic by origin.",
                        "fileTypes": "VCF",
                        "id": "#somatic",
                        "sbg:fileTypes": "VCF, TXT"
                    },
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "High confidence germline variants",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  filename = $job.inputs.status_file.path\n  basename = filename.split('.').slice(0, filename.split('.').length-1).join('.')\n\n  return basename+'.Germline.hc.vcf'\n}\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:inheritMetadataFrom": "#status_file"
                        },
                        "description": "File with high confidence variants that are germline by origin.",
                        "fileTypes": "VCF",
                        "id": "#germline_hc",
                        "sbg:fileTypes": "VCF, TXT"
                    },
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Germline variants",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  filename = $job.inputs.status_file.path\n  basename = filename.split('.').slice(0, filename.split('.').length-1).join('.')\n\n  return basename+'.Germline.vcf'\n}\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:inheritMetadataFrom": "#status_file"
                        },
                        "description": "File with variants that are germline by origin.",
                        "fileTypes": "VCF",
                        "id": "#germline",
                        "sbg:fileTypes": "VCF, TXT"
                    },
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "High confidence loss of heterozygosity variants",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  filename = $job.inputs.status_file.path\n  basename = filename.split('.').slice(0, filename.split('.').length-1).join('.')\n\n  return basename+'.LOH.hc.vcf'\n}\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:inheritMetadataFrom": "#status_file"
                        },
                        "description": "File with high confidence variants that are loss of heterozygosity by origin.",
                        "fileTypes": "VCF",
                        "id": "#LOH_hc",
                        "sbg:fileTypes": "VCF, TXT"
                    },
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Loss of heterozygosity variants",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  filename = $job.inputs.status_file.path\n  basename = filename.split('.').slice(0, filename.split('.').length-1).join('.')\n\n  return basename+'.LOH.vcf'\n}\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:inheritMetadataFrom": "#status_file"
                        },
                        "description": "File with variants that are loss of heterozygosity by origin.",
                        "fileTypes": "VCF",
                        "id": "#LOH",
                        "sbg:fileTypes": "VCF, TXT"
                    }
                ],
                "cwlVersion": "sbg:draft-2",
                "sbg:contributors": [
                    "markop",
                    "bix-demo",
                    "djordje_klisic"
                ],
                "sbg:sbgMaintained": false,
                "stdout": ""
            },
            "id": "#VarScan2_ProcessSomatic_SNPs",
            "sbg:y": 295.4100603842338
        },
        {
            "scatter": "#VarScan2_ProcessSomatic_INDELs.status_file",
            "inputs": [
                {
                    "id": "#VarScan2_ProcessSomatic_INDELs.status_file",
                    "source": [
                        "#VarScan2_Somatic_1.somatic_indels"
                    ]
                },
                {
                    "id": "#VarScan2_ProcessSomatic_INDELs.p_value"
                },
                {
                    "id": "#VarScan2_ProcessSomatic_INDELs.min_tumor_freq"
                },
                {
                    "id": "#VarScan2_ProcessSomatic_INDELs.max_normal_freq"
                }
            ],
            "outputs": [
                {
                    "id": "#VarScan2_ProcessSomatic_INDELs.somatic_hc"
                },
                {
                    "id": "#VarScan2_ProcessSomatic_INDELs.somatic"
                },
                {
                    "id": "#VarScan2_ProcessSomatic_INDELs.germline_hc"
                },
                {
                    "id": "#VarScan2_ProcessSomatic_INDELs.germline"
                },
                {
                    "id": "#VarScan2_ProcessSomatic_INDELs.LOH_hc"
                },
                {
                    "id": "#VarScan2_ProcessSomatic_INDELs.LOH"
                }
            ],
            "sbg:x": 1104.9663085937502,
            "run": {
                "sbg:toolkit": "VarScan",
                "requirements": [
                    {
                        "requirements": [
                            {
                                "class": "DockerRequirement",
                                "dockerPull": "rabix/js-engine"
                            }
                        ],
                        "id": "#cwl-js-engine",
                        "class": "ExpressionEngineRequirement"
                    }
                ],
                "sbg:validationErrors": [],
                "sbg:categories": [
                    "VCF-Processing"
                ],
                "label": "VarScan2 ProcessSomatic INDELs",
                "sbg:createdBy": "bix-demo",
                "sbg:cmdPreview": "java -jar -Xmx2048m /opt/VarScan2.3.9.jar processSomatic",
                "sbg:image_url": null,
                "sbg:links": [
                    {
                        "label": "Homepage",
                        "id": "http://varscan.sourceforge.net/using-varscan.html"
                    },
                    {
                        "label": "Source Code",
                        "id": "http://dkoboldt.github.io/varscan"
                    },
                    {
                        "label": "Wiki",
                        "id": "http://varscan.sourceforge.net/somatic-calling.html"
                    },
                    {
                        "label": "Download",
                        "id": "http://sourceforge.net/projects/varscan/files/VarScan.v2.3.9.jar/download"
                    },
                    {
                        "label": "Publication",
                        "id": "http://www.ncbi.nlm.nih.gov/pubmed/19542151"
                    },
                    {
                        "label": "Documentation",
                        "id": "http://varscan.sourceforge.net/doc/index.html"
                    }
                ],
                "sbg:revisionsInfo": [
                    {
                        "sbg:revision": 0,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911531
                    },
                    {
                        "sbg:revision": 1,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911533
                    },
                    {
                        "sbg:revision": 2,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1467636280
                    },
                    {
                        "sbg:revision": 3,
                        "sbg:revisionNotes": "Update PR",
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1468328200
                    },
                    {
                        "sbg:revision": 4,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "djordje_klisic",
                        "sbg:modifiedOn": 1469457737
                    },
                    {
                        "sbg:revision": 5,
                        "sbg:revisionNotes": "Staged",
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1469550974
                    }
                ],
                "y": 440.92810058593767,
                "sbg:modifiedOn": 1469550974,
                "sbg:license": "Non-Profit Open Software License 3.0",
                "sbg:modifiedBy": "markop",
                "class": "CommandLineTool",
                "inputs": [
                    {
                        "sbg:fileTypes": "VCF, TXT",
                        "sbg:stageInput": "link",
                        "type": [
                            "null",
                            "File"
                        ],
                        "required": false,
                        "inputBinding": {
                            "sbg:cmdInclude": true,
                            "position": -1,
                            "separate": true
                        },
                        "description": "The VarScan output file for SNPs or INDELs.",
                        "id": "#status_file",
                        "sbg:category": "File type inputs",
                        "label": "The VarScan output file for SNPs or INDELs"
                    },
                    {
                        "label": "P-value for high-confidence calling",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--p-value",
                            "sbg:cmdInclude": true,
                            "position": 5,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "0.07",
                        "description": "P-value for high-confidence calling.",
                        "id": "#p_value",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "float"
                        ]
                    },
                    {
                        "label": "Minimum variant allele frequency in tumor",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--min-tumor-freq",
                            "sbg:cmdInclude": true,
                            "position": 3,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "0.10",
                        "description": "Minimum variant allele frequency in tumor.",
                        "id": "#min_tumor_freq",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "float"
                        ]
                    },
                    {
                        "label": "Maximum variant allele frequency in normal",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--max-normal-freq",
                            "sbg:cmdInclude": true,
                            "position": 4,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "0.05",
                        "description": "Maximum variant allele frequency in normal.",
                        "id": "#max_normal_freq",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "float"
                        ]
                    }
                ],
                "hints": [
                    {
                        "dockerImageId": "e9fc75d12305",
                        "class": "DockerRequirement",
                        "dockerPull": "images.sbgenomics.com/djordje_klisic/varscan2:2.3.9"
                    },
                    {
                        "value": 1,
                        "class": "sbg:CPURequirement"
                    },
                    {
                        "value": {
                            "script": "{\n  if ($job.inputs.mem_mb)\n  {\n    return $job.inputs.mem_mb\n  }\n  \n  else\n  {\n    return 2048\n  }\n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "class": "sbg:MemRequirement"
                    }
                ],
                "sbg:revision": 5,
                "id": "bix-demo/varscan2-v2-3-9-demo/varscan2-processsomatic-v2-3-9/5",
                "sbg:latestRevision": 5,
                "temporaryFailCodes": [],
                "sbg:toolkitVersion": "2.3.9",
                "sbg:toolAuthor": "Genome Institute at Washington University",
                "successCodes": [],
                "sbg:revisionNotes": "Staged",
                "arguments": [],
                "appUrl": "/u/bix-demo/varscan2-v2-3-9-demo/apps/#bix-demo/varscan2-v2-3-9-demo/varscan2-processsomatic-v2-3-9/5",
                "baseCommand": [
                    "java",
                    "-jar",
                    {
                        "script": "{\n  if ($job.inputs.mem_mb)\n  {\n    return \"-Xmx\" + ($job.inputs.mem_mb).toString() + \"m\"\n  }\n  else\n  {\n    return \"-Xmx2048m\"\n  }\n}",
                        "engine": "#cwl-js-engine",
                        "class": "Expression"
                    },
                    "/opt/VarScan2.3.9.jar",
                    "processSomatic"
                ],
                "stdin": "",
                "sbg:project": "bix-demo/varscan2-v2-3-9-demo",
                "description": "The VarScan2 ProcessSomatic tool allow user to segregate different types of variants within the VCF file produced by the VarScan2 Somatic tool. Variants are segregated into three groups: germline, LOH (loss of heterozygosity), and somatic. Each of these groups will have two files on the tool output: one contains all variants from this group; the second is the subset of these variants that is considered high confidence. The output format of the tool is VCF by default.\n\n**The script will not work if the input is a list of \"called SNPs\" in a Varscan format.**",
                "x": 1104.9663085937502,
                "sbg:id": "admin/sbg-public-data/varscan2-processsomatic-v2-3-9/5",
                "sbg:createdOn": 1450911531,
                "sbg:job": {
                    "allocatedResources": {
                        "cpu": 1,
                        "mem": 2048
                    },
                    "inputs": {
                        "min_tumor_freq": 0,
                        "p_value": 0,
                        "max_normal_freq": 0,
                        "status_file": {
                            "path": "input/status_file.ext",
                            "size": 0,
                            "class": "File",
                            "secondaryFiles": []
                        }
                    }
                },
                "outputs": [
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "High confidence somatic variants",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  filename = $job.inputs.status_file.path\n  basename = filename.split('.').slice(0, filename.split('.').length-1).join('.')\n\n  return basename+'.Somatic.hc.vcf'\n}\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:inheritMetadataFrom": "#status_file"
                        },
                        "description": "File with high confidence variants that are somatic by origin.",
                        "fileTypes": "VCF",
                        "id": "#somatic_hc",
                        "sbg:fileTypes": "VCF, TXT"
                    },
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Somatic variants",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  filename = $job.inputs.status_file.path\n  basename = filename.split('.').slice(0, filename.split('.').length-1).join('.')\n\n  return basename+'.Somatic.vcf'\n}\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:inheritMetadataFrom": "#status_file"
                        },
                        "description": "File with variants that are somatic by origin.",
                        "fileTypes": "VCF",
                        "id": "#somatic",
                        "sbg:fileTypes": "VCF, TXT"
                    },
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "High confidence germline variants",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  filename = $job.inputs.status_file.path\n  basename = filename.split('.').slice(0, filename.split('.').length-1).join('.')\n\n  return basename+'.Germline.hc.vcf'\n}\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:inheritMetadataFrom": "#status_file"
                        },
                        "description": "File with high confidence variants that are germline by origin.",
                        "fileTypes": "VCF",
                        "id": "#germline_hc",
                        "sbg:fileTypes": "VCF, TXT"
                    },
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Germline variants",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  filename = $job.inputs.status_file.path\n  basename = filename.split('.').slice(0, filename.split('.').length-1).join('.')\n\n  return basename+'.Germline.vcf'\n}\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:inheritMetadataFrom": "#status_file"
                        },
                        "description": "File with variants that are germline by origin.",
                        "fileTypes": "VCF",
                        "id": "#germline",
                        "sbg:fileTypes": "VCF, TXT"
                    },
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "High confidence loss of heterozygosity variants",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  filename = $job.inputs.status_file.path\n  basename = filename.split('.').slice(0, filename.split('.').length-1).join('.')\n\n  return basename+'.LOH.hc.vcf'\n}\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:inheritMetadataFrom": "#status_file"
                        },
                        "description": "File with high confidence variants that are loss of heterozygosity by origin.",
                        "fileTypes": "VCF",
                        "id": "#LOH_hc",
                        "sbg:fileTypes": "VCF, TXT"
                    },
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Loss of heterozygosity variants",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  filename = $job.inputs.status_file.path\n  basename = filename.split('.').slice(0, filename.split('.').length-1).join('.')\n\n  return basename+'.LOH.vcf'\n}\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:inheritMetadataFrom": "#status_file"
                        },
                        "description": "File with variants that are loss of heterozygosity by origin.",
                        "fileTypes": "VCF",
                        "id": "#LOH",
                        "sbg:fileTypes": "VCF, TXT"
                    }
                ],
                "cwlVersion": "sbg:draft-2",
                "sbg:contributors": [
                    "markop",
                    "bix-demo",
                    "djordje_klisic"
                ],
                "sbg:sbgMaintained": false,
                "stdout": ""
            },
            "id": "#VarScan2_ProcessSomatic_INDELs",
            "sbg:y": 440.92810058593767
        },
        {
            "scatter": "#VarScan2_Somatic_1.tumor_pileup",
            "inputs": [
                {
                    "id": "#VarScan2_Somatic_1.validation"
                },
                {
                    "id": "#VarScan2_Somatic_1.tumor_purity"
                },
                {
                    "id": "#VarScan2_Somatic_1.tumor_pileup",
                    "source": [
                        "#SAMtools_Mpileup_parallel.output_file"
                    ]
                },
                {
                    "id": "#VarScan2_Somatic_1.strand_filter"
                },
                {
                    "id": "#VarScan2_Somatic_1.somatic_p_value"
                },
                {
                    "id": "#VarScan2_Somatic_1.p_value"
                },
                {
                    "id": "#VarScan2_Somatic_1.normal_purity"
                },
                {
                    "id": "#VarScan2_Somatic_1.normal_pileup",
                    "source": [
                        "#SAMtools_Mpileup_parallel_1.output_file"
                    ]
                },
                {
                    "id": "#VarScan2_Somatic_1.min_var_freq"
                },
                {
                    "id": "#VarScan2_Somatic_1.min_freq_for_hom"
                },
                {
                    "id": "#VarScan2_Somatic_1.min_coverage_tumor"
                },
                {
                    "id": "#VarScan2_Somatic_1.min_coverage_normal"
                },
                {
                    "id": "#VarScan2_Somatic_1.min_coverage"
                },
                {
                    "id": "#VarScan2_Somatic_1.mem_mb"
                },
                {
                    "id": "#VarScan2_Somatic_1.intervals_files_scatter"
                }
            ],
            "outputs": [
                {
                    "id": "#VarScan2_Somatic_1.validation_output"
                },
                {
                    "id": "#VarScan2_Somatic_1.somatic_snps"
                },
                {
                    "id": "#VarScan2_Somatic_1.somatic_indels"
                }
            ],
            "sbg:x": 865.8330202202162,
            "run": {
                "sbg:toolkit": "VarScan",
                "requirements": [
                    {
                        "requirements": [
                            {
                                "class": "DockerRequirement",
                                "dockerPull": "rabix/js-engine"
                            }
                        ],
                        "class": "ExpressionEngineRequirement",
                        "engineCommand": "cwl-engine.js",
                        "id": "#cwl-js-engine"
                    }
                ],
                "sbg:validationErrors": [],
                "sbg:categories": [
                    "Variant-Calling"
                ],
                "label": "VarScan2 Somatic",
                "sbg:createdBy": "bix-demo",
                "sbg:cmdPreview": "java -jar -Xmx2048m /opt/VarScan2.3.9.jar somatic  /path/to/normal_pileup-1.ext /path/to/tumor_pileup.ext --output-snp tumor_pileup_SNPs.vcf --output-indel tumor_pileup_INDELs.vcf  --output-vcf",
                "sbg:image_url": null,
                "sbg:links": [
                    {
                        "label": "Homepage",
                        "id": "http://varscan.sourceforge.net/using-varscan.html"
                    },
                    {
                        "label": "Source Code",
                        "id": "http://dkoboldt.github.io/varscan"
                    },
                    {
                        "label": "Wiki",
                        "id": "http://varscan.sourceforge.net/somatic-calling.html"
                    },
                    {
                        "label": "Download",
                        "id": "http://sourceforge.net/projects/varscan/files/VarScan.v2.3.9.jar/download"
                    },
                    {
                        "label": "Publication",
                        "id": "http://www.ncbi.nlm.nih.gov/pubmed/19542151"
                    },
                    {
                        "label": "Documentation",
                        "id": "http://varscan.sourceforge.net/doc/index.html"
                    }
                ],
                "sbg:revisionsInfo": [
                    {
                        "sbg:revision": 0,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911535
                    },
                    {
                        "sbg:revision": 1,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911535
                    },
                    {
                        "sbg:revision": 2,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1467635695
                    },
                    {
                        "sbg:revision": 3,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1467635901
                    },
                    {
                        "sbg:revision": 4,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1468277658
                    },
                    {
                        "sbg:revision": 5,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1468325180
                    },
                    {
                        "sbg:revision": 6,
                        "sbg:revisionNotes": "Added \"validation\" param.",
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1468326080
                    },
                    {
                        "sbg:revision": 7,
                        "sbg:revisionNotes": "output_vcf moved from args to inputs.",
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1468326509
                    },
                    {
                        "sbg:revision": 8,
                        "sbg:revisionNotes": "Updated GLOB.",
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1468327489
                    },
                    {
                        "sbg:revision": 9,
                        "sbg:revisionNotes": "output-vcf back to args",
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1471276668
                    },
                    {
                        "sbg:revision": 10,
                        "sbg:revisionNotes": "Merge the usage of somatic and somatic parallel in one tool. This will replace somatic parallel in the varscan workflow.",
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1473857440
                    },
                    {
                        "sbg:revision": 11,
                        "sbg:revisionNotes": "Adding some warning messages for scattering mode usage .",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1473927287
                    }
                ],
                "y": 362.2971600713027,
                "sbg:modifiedOn": 1473927287,
                "sbg:license": "Non-Profit Open Software License 3.0",
                "sbg:modifiedBy": "marouf",
                "class": "CommandLineTool",
                "inputs": [
                    {
                        "label": "Validation:If set to 1, outputs all compared positions even if non-variant",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--validation",
                            "sbg:cmdInclude": true,
                            "position": 15,
                            "separate": true
                        },
                        "description": "Validation:If set to 1, outputs all compared positions even if non-variant.",
                        "id": "#validation",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "label": "Estimated purity for tumor",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--tumor-purity",
                            "sbg:cmdInclude": true,
                            "position": 11,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "1.00",
                        "description": "Estimated purity (tumor content) of tumor tissue sample sample.",
                        "id": "#tumor_purity",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "float"
                        ]
                    },
                    {
                        "sbg:fileTypes": "PILEUP, TXT",
                        "label": "The SAMtools pileup file for tumor",
                        "required": true,
                        "description": "The SAMtools pileup file for tumor tissue sample. Preferably use SAMtools mpileup command from the version 0.1.18 and older.",
                        "id": "#tumor_pileup",
                        "sbg:category": "Input file types",
                        "type": [
                            "File"
                        ]
                    },
                    {
                        "label": "Strand-filter: If set to 1, removes variants with >90% strand bias",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--strand-filter",
                            "sbg:cmdInclude": true,
                            "position": 14
                        },
                        "description": "Strand-filter: If set to 1, removes variants with >90% strand bias.",
                        "id": "#strand_filter",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "label": "P-value threshold to call a somatic site",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--somatic-p-value",
                            "sbg:cmdInclude": true,
                            "position": 13,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "0.05",
                        "description": "P-value threshold to call a somatic site.",
                        "id": "#somatic_p_value",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "float"
                        ]
                    },
                    {
                        "label": "P-value threshold to call a heterozygote",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--p-value",
                            "sbg:cmdInclude": true,
                            "position": 12,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "0.99",
                        "description": "P-value threshold to call a heterozygote.",
                        "id": "#p_value",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "float"
                        ]
                    },
                    {
                        "label": "Estimated purity for normal",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--normal-purity",
                            "sbg:cmdInclude": true,
                            "position": 10,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "1.00",
                        "description": "Estimated purity (non-tumor content) of normal tissue sample sample.",
                        "id": "#normal_purity",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "float"
                        ]
                    },
                    {
                        "sbg:fileTypes": "PILEUP, TXT",
                        "label": "The SAMtools pileup file for normal",
                        "required": true,
                        "description": "The SAMtools pileup file for normal tissue sample. Preferably use SAMtools mpileup command from the version 0.1.18 and older. Array of files could be provided only in case of scattering which is the case of varscan pipeline from BAM. So please provide only single file on this input.",
                        "id": "#normal_pileup",
                        "sbg:category": "Input file types",
                        "type": [
                            {
                                "name": "normal_pileup",
                                "type": "array",
                                "items": "File"
                            }
                        ]
                    },
                    {
                        "label": "Minimum variant frequency to call a heterozygote",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--min-var-freq",
                            "sbg:cmdInclude": true,
                            "position": 8,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "0.10",
                        "description": "Minimum variant frequency to call a heterozygote.",
                        "id": "#min_var_freq",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "float"
                        ]
                    },
                    {
                        "label": "Minimum frequency to call homozygote",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--min-freq-for-hom",
                            "sbg:cmdInclude": true,
                            "position": 9,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "0.75",
                        "description": "Minimum frequency to call homozygote.",
                        "id": "#min_freq_for_hom",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "float"
                        ]
                    },
                    {
                        "label": "Minimum coverage in tumor to call somatic",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--min-coverage-tumor",
                            "sbg:cmdInclude": true,
                            "position": 6,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "6",
                        "description": "Minimum coverage in tumor tissue sample to call somatic.",
                        "id": "#min_coverage_tumor",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "label": "Minimum coverage in normal to call somatic",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--min-coverage-normal",
                            "sbg:cmdInclude": true,
                            "position": 5,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "8",
                        "description": "Minimum coverage in normal tissue sample to call somatic.",
                        "id": "#min_coverage_normal",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "label": "Minimum coverage in normal and tumor to call variant",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--min-coverage",
                            "sbg:cmdInclude": true,
                            "position": 4,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "8",
                        "description": "Minimum coverage in normal and tumor to call variant.",
                        "id": "#min_coverage",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "label": "Java Xmx%m requirement",
                        "required": false,
                        "sbg:toolDefaultValue": "2048",
                        "description": "RAM requirement for the java process execution.",
                        "id": "#mem_mb",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "sbg:stageInput": null,
                        "id": "#intervals_files_scatter",
                        "label": "Intervals files names used in creating of pileup files",
                        "required": false,
                        "type": [
                            "null",
                            "string"
                        ]
                    }
                ],
                "hints": [
                    {
                        "dockerImageId": "e9fc75d12305",
                        "class": "DockerRequirement",
                        "dockerPull": "images.sbgenomics.com/djordje_klisic/varscan2:2.3.9"
                    },
                    {
                        "value": 1,
                        "class": "sbg:CPURequirement"
                    },
                    {
                        "value": {
                            "script": "{\n  if ($job.inputs.mem_mb)\n  {\n    return $job.inputs.mem_mb\n  }\n  \n  else\n  {\n    return 2048\n  }\n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "class": "sbg:MemRequirement"
                    }
                ],
                "sbg:revision": 11,
                "id": "bix-demo/varscan2-v2-3-9-demo/varscan2-somatic-v2-3-9/11",
                "sbg:latestRevision": 11,
                "temporaryFailCodes": [],
                "sbg:toolkitVersion": "2.3.9",
                "sbg:toolAuthor": "Genome Institute at Washington University",
                "successCodes": [],
                "sbg:revisionNotes": "Adding some warning messages for scattering mode usage .",
                "arguments": [
                    {
                        "prefix": "--output-snp",
                        "position": 2,
                        "valueFrom": {
                            "script": "{\n  \n  filename = [].concat($job.inputs.tumor_pileup)[0].path\n  \n  \n  basename = filename.split('.').slice(0, filename.split('.').length-1).join('.').replace(/^.*[\\\\\\/]/, '')\n\n  return basename+'_SNPs.vcf'\n}\n",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "separate": true
                    },
                    {
                        "prefix": "--output-indel",
                        "position": 3,
                        "valueFrom": {
                            "script": "{\n  filename = [].concat($job.inputs.tumor_pileup)[0].path\n   \n  \n  basename = filename.split('.').slice(0, filename.split('.').length-1).join('.').replace(/^.*[\\\\\\/]/, '')\n\n  return basename+'_INDELs.vcf'\n}\n",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "separate": true
                    },
                    {
                        "prefix": "",
                        "position": 16,
                        "valueFrom": "--output-vcf",
                        "separate": true
                    },
                    {
                        "position": 0,
                        "valueFrom": {
                            "script": "{ \n  \n  //scattering using tumor_pileup input\n  if($job.inputs.normal_pileup.length>1)\n  {\n    normal = \"0\"\n \n    //find the normal associated to tumor\n    tumor_scatteredUsing = [].concat($job.inputs.tumor_pileup)[0].metadata.ScatteredUsing.replace(/\\s/g, '')\n\ttumor_path = [].concat($job.inputs.tumor_pileup)[0].path\n    \n    for (fn in $job.inputs.normal_pileup)\n    {\n      if($job.inputs.normal_pileup[fn].metadata.ScatteredUsing)\n      {\n          if ($job.inputs.normal_pileup[fn].metadata.ScatteredUsing.replace(/\\s/g, '') == tumor_scatteredUsing)\n          {\n\n            normal = $job.inputs.normal_pileup[fn].path\n          }            \n\n      }\n      \n    }\n    \n\tif (typeof normal !== \"0\" && typeof tumor_path !== \"0\") // Any scope\n    \treturn normal +\" \" +tumor_path\n\n  }\n  else\n  {\n  \treturn [].concat($job.inputs.normal_pileup)[0].path+ \" \"+ $job.inputs.tumor_pileup.path\n  \n  }\n  \n}\n\n",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "separate": true
                    }
                ],
                "appUrl": "/u/bix-demo/varscan2-v2-3-9-demo/apps/#bix-demo/varscan2-v2-3-9-demo/varscan2-somatic-v2-3-9/11",
                "baseCommand": [
                    "java",
                    "-jar",
                    {
                        "script": "{\n  if ($job.inputs.mem_mb)\n  {\n    return \"-Xmx\" + ($job.inputs.mem_mb).toString() + \"m\"\n  }\n  else\n  {\n    return \"-Xmx2048m\"\n  }\n}",
                        "engine": "#cwl-js-engine",
                        "class": "Expression"
                    },
                    "/opt/VarScan2.3.9.jar",
                    "somatic"
                ],
                "stdin": "",
                "sbg:project": "bix-demo/varscan2-v2-3-9-demo",
                "description": "VarScan calls somatic variants (SNPs and INDELs) using a heuristic method and a statistical test based on the number of aligned reads supporting each allele. \n\nThe VarScan2 Somatic command calls variants and identifies their somatic status (Germline/LOH/Somatic) using pileup files from a matched tumor-normal pair generated using SAMtools mpileup. \n\nMore specific options (e.g. min-coverage-normal) will override the default or specified value of less specific options (e.g. min-coverage). \n\nThe normal and tumor purity values should be a value between 0 and 1. The default (1) implies that the normal is 100% pure with no contaminating tumor cells, and the tumor is 100% pure with no contaminating stromal or other non-malignant cells. Please change the tumor-purity to something less than 1 if you have a low-purity tumor sample and thus expect lower variant allele frequencies for mutations. Please change the normal-purity to something less than 1 only if it is possible that there will be some tumor content in your \"normal\" sample, e.g. adjacent normal tissue for a solid tumor, malignant blood cells in the skin punch normal for some liquid tumors, etc. \n\nThere are two p-value options: the first, p-value, is the significance threshold for the first-pass algorithm that determines, for each position, if either normal or tumor is variant at that position; the second, somatic-p-value, is more important, as this is the threshold below which read count differences between tumor and normal are deemed significant enough to classify the sample as a somatic mutation or an LOH event.\n\nAnother mode for scattering is already applied in CWL implementation of this tool. This mode is used for varscan pipeline from BAM and is designed only to work with this pipeline.",
                "x": 865.8330202202162,
                "sbg:id": "admin/sbg-public-data/varscan2-somatic-v2-3-9/11",
                "sbg:createdOn": 1450911535,
                "sbg:job": {
                    "allocatedResources": {
                        "cpu": 1,
                        "mem": 2048
                    },
                    "inputs": {
                        "mem_mb": null,
                        "somatic_p_value": null,
                        "min_freq_for_hom": null,
                        "tumor_purity": null,
                        "intervals_files_scatter": "10",
                        "tumor_pileup": {
                            "metadata": {
                                "ScatteredUsing": "10"
                            },
                            "path": "/path/to/tumor_pileup.ext",
                            "size": 0,
                            "class": "File",
                            "secondaryFiles": []
                        },
                        "p_value": null,
                        "min_var_freq": null,
                        "normal_purity": null,
                        "validation": false,
                        "normal_pileup": [
                            {
                                "metadata": {
                                    "ScatteredUsing": "10",
                                    "filetype": "\"bam\""
                                },
                                "path": "/path/to/normal_pileup-1.ext",
                                "size": 5,
                                "class": "File",
                                "secondaryFiles": []
                            },
                            {
                                "metadata": {
                                    "ScatteredUsing": "20"
                                },
                                "path": "/path/to/normal_pileup-2.ext",
                                "size": 0,
                                "class": "File",
                                "secondaryFiles": []
                            }
                        ],
                        "strand_filter": false
                    }
                },
                "outputs": [
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Validation output",
                        "outputBinding": {
                            "glob": "*.validation",
                            "sbg:inheritMetadataFrom": "#tumor_pileup"
                        },
                        "sbg:fileTypes": "VALIDATION",
                        "id": "#validation_output",
                        "description": "the --validation flag is turned on, a file (.validation) will also be generated containing all positions that were called."
                    },
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Output file for SNP calls",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  \n  filename = [].concat($job.inputs.tumor_pileup)[0].path\n  \n  \n  basename = filename.split('.').slice(0, filename.split('.').length-1).join('.').replace(/^.*[\\\\\\/]/, '')\n\n  return basename+'_SNPs.vcf'\n}\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:metadata": {
                                "__inherit__": "normal_pileup"
                            },
                            "sbg:inheritMetadataFrom": "#tumor_pileup"
                        },
                        "sbg:fileTypes": "VCF",
                        "id": "#somatic_snps",
                        "description": "Output file for SNP calls."
                    },
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Output file for INDEL calls",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  filename = [].concat($job.inputs.tumor_pileup)[0].path\n   \n  \n  basename = filename.split('.').slice(0, filename.split('.').length-1).join('.').replace(/^.*[\\\\\\/]/, '')\n\n  return basename+'_INDELs.vcf'\n}\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:metadata": {
                                "__inherit__": "normal_pileup"
                            },
                            "sbg:inheritMetadataFrom": "#tumor_pileup"
                        },
                        "sbg:fileTypes": "VCF",
                        "id": "#somatic_indels",
                        "description": "Output file for INDEL calls."
                    }
                ],
                "cwlVersion": "sbg:draft-2",
                "sbg:contributors": [
                    "marouf",
                    "bix-demo",
                    "markop"
                ],
                "sbg:sbgMaintained": false,
                "stdout": ""
            },
            "id": "#VarScan2_Somatic_1",
            "sbg:y": 362.2971600713027
        },
        {
            "scatter": "#VarScan2_Filter_1.indel_file",
            "inputs": [
                {
                    "id": "#VarScan2_Filter_1.snp_file",
                    "source": [
                        "#VarScan2_Somatic_1.somatic_snps"
                    ]
                },
                {
                    "id": "#VarScan2_Filter_1.p_value"
                },
                {
                    "id": "#VarScan2_Filter_1.min_var_freq"
                },
                {
                    "id": "#VarScan2_Filter_1.min_strands2"
                },
                {
                    "id": "#VarScan2_Filter_1.min_reads2"
                },
                {
                    "id": "#VarScan2_Filter_1.min_coverage"
                },
                {
                    "id": "#VarScan2_Filter_1.min_avg_qual"
                },
                {
                    "id": "#VarScan2_Filter_1.mem_mb"
                },
                {
                    "id": "#VarScan2_Filter_1.indel_file",
                    "source": [
                        "#VarScan2_Somatic_1.somatic_indels"
                    ]
                }
            ],
            "outputs": [
                {
                    "id": "#VarScan2_Filter_1.filtered_snps"
                }
            ],
            "sbg:x": 1122.4999314546587,
            "run": {
                "sbg:toolkit": "VarScan",
                "requirements": [
                    {
                        "requirements": [
                            {
                                "class": "DockerRequirement",
                                "dockerPull": "rabix/js-engine"
                            }
                        ],
                        "class": "ExpressionEngineRequirement",
                        "engineCommand": "cwl-engine.js",
                        "id": "#cwl-js-engine"
                    }
                ],
                "sbg:validationErrors": [],
                "sbg:categories": [
                    "VCF-Processing"
                ],
                "label": "VarScan2 Filter",
                "sbg:createdBy": "bix-demo",
                "sbg:cmdPreview": "java -jar -Xmx2048m /opt/VarScan2.3.9.jar somaticFilter  NORMAL.sorted.contig_MT.mpileup_SNPs.vcf --indel-file NORMAL.sorted.contig_MT.mpileup_INDELs.vcf --output-file NORMAL.sorted.contig_MT.mpileup_SNPs.filtered.vcf",
                "sbg:image_url": null,
                "sbg:links": [
                    {
                        "label": "Homepage",
                        "id": "http://varscan.sourceforge.net/using-varscan.html"
                    },
                    {
                        "label": "Source Code",
                        "id": "http://dkoboldt.github.io/varscan"
                    },
                    {
                        "label": "Wiki",
                        "id": "http://varscan.sourceforge.net/somatic-calling.html"
                    },
                    {
                        "label": "Download",
                        "id": "http://sourceforge.net/projects/varscan/files/VarScan.v2.3.9.jar/download"
                    },
                    {
                        "label": "Publication",
                        "id": "http://www.ncbi.nlm.nih.gov/pubmed/19542151"
                    },
                    {
                        "label": "Documentation",
                        "id": "http://varscan.sourceforge.net/doc/index.html"
                    }
                ],
                "sbg:revisionsInfo": [
                    {
                        "sbg:revision": 0,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911528
                    },
                    {
                        "sbg:revision": 1,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911528
                    },
                    {
                        "sbg:revision": 2,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "markop",
                        "sbg:modifiedOn": 1467637122
                    },
                    {
                        "sbg:revision": 3,
                        "sbg:revisionNotes": "new design of varscan filter. This tool can work now with and without scattering.",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1473926671
                    },
                    {
                        "sbg:revision": 4,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1473926971
                    },
                    {
                        "sbg:revision": 5,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1473927110
                    }
                ],
                "y": 293.5474334777928,
                "sbg:modifiedOn": 1473927110,
                "sbg:license": "Non-Profit Open Software License 3.0",
                "sbg:modifiedBy": "marouf",
                "class": "CommandLineTool",
                "inputs": [
                    {
                        "sbg:fileTypes": "VCF,TXT",
                        "label": "VCF file for somatic SNPs",
                        "required": true,
                        "description": "VCF file for somatic SNPs generated using VarScan2 Somatic tool. Array of files are accepted on this input just in case scattering is enabled which is the case of varscan pipeline from BAM, so please don't provide array on this input.",
                        "id": "#snp_file",
                        "sbg:category": "File type inputs",
                        "type": [
                            {
                                "name": "snp_file",
                                "type": "array",
                                "items": "File"
                            }
                        ]
                    },
                    {
                        "label": "Default p-value threshold",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--p-value",
                            "sbg:cmdInclude": true,
                            "position": 8,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "0.05",
                        "description": "Default p-value threshold for calling variants.",
                        "id": "#p_value",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "float"
                        ]
                    },
                    {
                        "label": "Minimum variant allele frequency threshold",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--min-var-freq",
                            "sbg:cmdInclude": true,
                            "position": 7,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "0.20",
                        "description": "Minimum variant allele frequency threshold.",
                        "id": "#min_var_freq",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "float"
                        ]
                    },
                    {
                        "label": "Minimum # of strands on which variant observed",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--min-strands2",
                            "sbg:cmdInclude": true,
                            "position": 5
                        },
                        "sbg:toolDefaultValue": "1",
                        "description": "Minimum # of strands on which variant observed.",
                        "id": "#min_strands2",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            {
                                "name": "min_strands2",
                                "symbols": [
                                    "1",
                                    "2"
                                ],
                                "type": "enum"
                            }
                        ]
                    },
                    {
                        "label": "Minimum supporting reads",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--min-reads2",
                            "sbg:cmdInclude": true,
                            "position": 4
                        },
                        "sbg:toolDefaultValue": "4",
                        "description": "Minimum supporting reads at a position to call variants.",
                        "id": "#min_reads2",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "label": "Minimum read depth",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--min-coverage",
                            "sbg:cmdInclude": true,
                            "position": 2,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "10",
                        "description": "Minimum read depth at a position to make a call.",
                        "id": "#min_coverage",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "label": "Minimum average base quality for variant-supporting reads",
                        "required": false,
                        "inputBinding": {
                            "prefix": "--min-avg-qual",
                            "sbg:cmdInclude": true,
                            "position": 6,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "20",
                        "description": "Minimum average base quality for variant-supporting reads.",
                        "id": "#min_avg_qual",
                        "sbg:category": "Other input types",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "description": "RAM requirement for the java process execution.",
                        "id": "#mem_mb",
                        "label": "Java Xmx%m requirement",
                        "required": false,
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "label": "File of INDELs for filtering nearby SNPs",
                        "required": false,
                        "description": "File of INDELs for filtering nearby SNPs.",
                        "id": "#indel_file",
                        "sbg:category": "File type inputs",
                        "type": [
                            "null",
                            "File"
                        ]
                    }
                ],
                "temporaryFailCodes": [],
                "sbg:revision": 5,
                "id": "bix-demo/varscan2-v2-3-9-demo/varscan2-filter-v2-3-9/5",
                "sbg:latestRevision": 5,
                "sbg:toolkitVersion": "2.3.9",
                "sbg:toolAuthor": "Genome Institute at Washington University",
                "successCodes": [],
                "outputs": [
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Filtered VCF file",
                        "outputBinding": {
                            "glob": {
                                "script": "\n{\n  \n  \n  // scatter mode using indels input\n  if($job.inputs.snp_file.length>1)\n  {\n    snp = \"0\"\n    indel = \"0\"\n\n    // one indel and list of SNP files\n    if([].concat($job.inputs.indel_file)[0])\n    {\n      indel_scatteredUsing = [].concat($job.inputs.indel_file)[0].metadata.ScatteredUsing\n      indel = [].concat($job.inputs.indel_file)[0].path\n\n      for (fn in $job.inputs.snp_file)\n      {\n        if ($job.inputs.snp_file[fn].metadata.ScatteredUsing == indel_scatteredUsing)\n        {\n          snp = $job.inputs.snp_file[fn].path\n        } \n      }\n\n    }\n    \n    filename = snp\n    basename = filename.split('.').slice(0, filename.split('.').length-1).join('.').replace(/^.*[\\\\\\/]/, '')\n    return basename+'.filtered.vcf'\n\n\n  }\n  else\n  {\n    \n    if([].concat($job.inputs.snp_file)[0]!=null)\n    {\n      filename = [].concat($job.inputs.snp_file)[0].path\n      basename = filename.split('.').slice(0, filename.split('.').length-1).join('.').replace(/^.*[\\\\\\/]/, '')\n      return basename+'.filtered.vcf'\n    }\n    \n  }\n  \n}\n\n\n\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:metadata": {
                                "__inherit__": "snp_file"
                            },
                            "sbg:inheritMetadataFrom": "#indel_file"
                        },
                        "sbg:fileTypes": "VCF,TXT",
                        "id": "#filtered_snps",
                        "description": "SNP VCF file filtered around the INDELs."
                    }
                ],
                "arguments": [
                    {
                        "prefix": "--output-file",
                        "position": 15,
                        "valueFrom": {
                            "script": "\n{\n  \n  \n  // scatter mode using indels input\n  if($job.inputs.snp_file.length>1)\n  {\n    snp = \"0\"\n    indel = \"0\"\n\n    // one indel and list of SNP files\n    if([].concat($job.inputs.indel_file)[0])\n    {\n      indel_scatteredUsing = [].concat($job.inputs.indel_file)[0].metadata.ScatteredUsing\n      indel = [].concat($job.inputs.indel_file)[0].path\n\n      for (fn in $job.inputs.snp_file)\n      {\n        if ($job.inputs.snp_file[fn].metadata.ScatteredUsing == indel_scatteredUsing)\n        {\n          snp = $job.inputs.snp_file[fn].path\n        } \n      }\n\n    }\n    \n    filename = snp\n    basename = filename.split('.').slice(0, filename.split('.').length-1).join('.').replace(/^.*[\\\\\\/]/, '')\n    return basename+'.filtered.vcf'\n\n\n  }\n  else\n  {\n    \n    if([].concat($job.inputs.snp_file)[0]!=null)\n    {\n      filename = [].concat($job.inputs.snp_file)[0].path\n      basename = filename.split('.').slice(0, filename.split('.').length-1).join('.').replace(/^.*[\\\\\\/]/, '')\n      return basename+'.filtered.vcf'\n    }\n    \n  }\n  \n}\n\n\n\n",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "separate": true
                    },
                    {
                        "position": 1,
                        "valueFrom": {
                            "script": "{\n  \n  \n  // scatter mode using indels input\n  if($job.inputs.snp_file.length>1)\n  {\n    snp = \"0\"\n    indel = \"0\"\n\n    // one indel and list of SNP files\n    if([].concat($job.inputs.indel_file)[0])\n    {\n      indel_scatteredUsing = [].concat($job.inputs.indel_file)[0].metadata.ScatteredUsing\n      indel = [].concat($job.inputs.indel_file)[0].path\n\n      for (fn in $job.inputs.snp_file)\n      {\n        if ($job.inputs.snp_file[fn].metadata.ScatteredUsing == indel_scatteredUsing)\n        {\n          snp = $job.inputs.snp_file[fn].path\n        } \n      }\n\n    }\n\n\tif(snp != '0' && indel != '0')\n  \t\treturn snp +\" \" +\"--indel-file \" + indel\n    else if(snp != '0')\n    \treturn snp \n\n\n  }\n  else\n  {\n    \n    if($job.inputs.indel_file.path)\n    {\n      return [].concat($job.inputs.snp_file)[0].path  +\" \" +\"--indel-file \" + $job.inputs.indel_file.path \n    }\n    else\n    {\n      return [].concat($job.inputs.snp_file)[0].path\n    }\n    \n  }\n  \n}\n\n",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "separate": true
                    }
                ],
                "hints": [
                    {
                        "dockerImageId": "",
                        "class": "DockerRequirement",
                        "dockerPull": "images.sbgenomics.com/djordje_klisic/varscan2:2.3.9"
                    },
                    {
                        "value": 1,
                        "class": "sbg:CPURequirement"
                    },
                    {
                        "value": {
                            "script": "{\n  if ($job.inputs.mem_mb)\n  {\n    return $job.inputs.mem_mb\n  }\n  \n  else\n  {\n    return 2048\n  }\n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "class": "sbg:MemRequirement"
                    }
                ],
                "appUrl": "/u/bix-demo/varscan2-v2-3-9-demo/apps/#bix-demo/varscan2-v2-3-9-demo/varscan2-filter-v2-3-9/5",
                "baseCommand": [
                    "java",
                    "-jar",
                    {
                        "script": "{\n  if ($job.inputs.mem_mb)\n  {\n    return \"-Xmx\" + ($job.inputs.mem_mb).toString() + \"m\"\n  }\n  else\n  {\n    return \"-Xmx2048m\"\n  }\n}",
                        "engine": "#cwl-js-engine",
                        "class": "Expression"
                    },
                    "/opt/VarScan2.3.9.jar",
                    "somaticFilter"
                ],
                "stdin": "",
                "sbg:project": "bix-demo/varscan2-v2-3-9-demo",
                "description": "VarScan2 Filter tool filters variants in a file by coverage, supporting reads, variant frequency, or average base quality. It is for use with output from pileup2snp or pileup2indel.\nMore advanced filtering strategies consider mapping quality, read mismatches, soft-trimming, and other factors when deciding whether or not to filter a variant. For more information consult publication or documentation from links section.\n\nAnother mode for scattering is already applied in CWL implementation of this tool. This mode is used for varscan pipeline from BAM and is designed only to work with this pipeline.",
                "x": 1122.4999314546587,
                "sbg:id": "admin/sbg-public-data/varscan2-filter-v2-3-9/5",
                "sbg:createdOn": 1450911528,
                "sbg:job": {
                    "allocatedResources": {
                        "cpu": 1,
                        "mem": 2048
                    },
                    "inputs": {
                        "indel_file": {
                            "path": "NORMAL.sorted.contig_MT.mpileup_INDELs.vcf",
                            "size": 0,
                            "class": "File",
                            "secondaryFiles": []
                        },
                        "min_avg_qual": 0,
                        "min_var_freq": 0,
                        "mem_mb": 0,
                        "snp_file": [
                            {
                                "path": "NORMAL.sorted.contig_MT.mpileup_SNPs.vcf",
                                "size": 0,
                                "class": "File",
                                "secondaryFiles": []
                            }
                        ],
                        "p_value": 0
                    }
                },
                "cwlVersion": "sbg:draft-2",
                "sbg:contributors": [
                    "marouf",
                    "bix-demo",
                    "markop"
                ],
                "sbg:sbgMaintained": false,
                "stdout": ""
            },
            "id": "#VarScan2_Filter_1",
            "sbg:y": 293.5474334777928
        },
        {
            "sbg:x": 318.66669718424487,
            "outputs": [
                {
                    "id": "#SAMtools_Index_FASTA.output_fasta_file"
                },
                {
                    "id": "#SAMtools_Index_FASTA.generated_index"
                }
            ],
            "run": {
                "sbg:toolkit": "SAMtools",
                "requirements": [
                    {
                        "requirements": [
                            {
                                "class": "DockerRequirement",
                                "dockerPull": "rabix/js-engine"
                            }
                        ],
                        "id": "#cwl-js-engine",
                        "class": "ExpressionEngineRequirement"
                    }
                ],
                "sbg:validationErrors": [],
                "sbg:categories": [
                    "FASTA-Processing",
                    "Indexing"
                ],
                "label": "SAMtools Index FASTA",
                "sbg:createdBy": "marouf",
                "sbg:cmdPreview": "/opt/samtools-1.3/samtools faidx fasta.fa",
                "sbg:image_url": null,
                "cwlVersion": "sbg:draft-2",
                "sbg:revisionsInfo": [
                    {
                        "sbg:revision": 0,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1458812964
                    },
                    {
                        "sbg:revision": 1,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1458822310
                    },
                    {
                        "sbg:revision": 2,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1458822367
                    },
                    {
                        "sbg:revision": 3,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1458822452
                    },
                    {
                        "sbg:revision": 4,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1459328843
                    },
                    {
                        "sbg:revision": 5,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1462801546
                    },
                    {
                        "sbg:revision": 6,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1462801641
                    },
                    {
                        "sbg:revision": 7,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1462801722
                    },
                    {
                        "sbg:revision": 8,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1462807323
                    },
                    {
                        "sbg:revision": 9,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1467297878
                    },
                    {
                        "sbg:revision": 10,
                        "sbg:revisionNotes": "peer review",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1473255676
                    },
                    {
                        "sbg:revision": 11,
                        "sbg:revisionNotes": "peer review",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1473255729
                    },
                    {
                        "sbg:revision": 12,
                        "sbg:revisionNotes": "remove required option for output fastq file",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1480425012
                    }
                ],
                "y": 333.8379720052084,
                "sbg:toolkitVersion": "v1.3",
                "sbg:license": "The MIT License",
                "sbg:modifiedBy": "marouf",
                "class": "CommandLineTool",
                "inputs": [
                    {
                        "sbg:fileTypes": "FAI",
                        "sbg:stageInput": "link",
                        "label": "Input indexed file",
                        "required": false,
                        "description": "Input index file (FAI).",
                        "id": "#input_index_file",
                        "sbg:category": "File input",
                        "type": [
                            "null",
                            "File"
                        ]
                    },
                    {
                        "sbg:fileTypes": "FASTA,FA,GZ",
                        "sbg:stageInput": "copy",
                        "label": "FASTA input file",
                        "required": true,
                        "description": "FASTA input file.",
                        "id": "#input_fasta_file",
                        "sbg:category": "File input",
                        "type": [
                            "File"
                        ]
                    },
                    {
                        "description": "Don't output indexed data file.",
                        "id": "#Dont_output_indexed_data",
                        "label": "Don't output indexed data file",
                        "sbg:category": "Configuration",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    }
                ],
                "hints": [
                    {
                        "dockerImageId": "2fb927277493",
                        "class": "DockerRequirement",
                        "dockerPull": "images.sbgenomics.com/marouf/samtools:1.3"
                    },
                    {
                        "value": 1,
                        "class": "sbg:CPURequirement"
                    },
                    {
                        "value": 1000,
                        "class": "sbg:MemRequirement"
                    }
                ],
                "sbg:revision": 12,
                "baseCommand": [
                    {
                        "script": "{\n  if ($job.inputs.input_index_file)\n  {\n \treturn\"echo Skipping index step because index file is provided on the input.\"\n  }\n  else\n  {\n    return \"/opt/samtools-1.3/samtools faidx \" + $job.inputs.input_fasta_file.path\n  }\n}\n",
                        "engine": "#cwl-js-engine",
                        "class": "Expression"
                    }
                ],
                "id": "marouf/samtools-1-3-demo/samtools-index-fasta-1-3/12",
                "sbg:latestRevision": 12,
                "temporaryFailCodes": [],
                "successCodes": [],
                "sbg:toolAuthor": "Heng Li, Sanger Institute",
                "sbg:revisionNotes": "remove required option for output fastq file",
                "arguments": [],
                "sbg:id": "admin/sbg-public-data/samtools-index-fasta-1-3/12",
                "sbg:job": {
                    "allocatedResources": {
                        "cpu": 1,
                        "mem": 1000
                    },
                    "inputs": {
                        "input_fasta_file": {
                            "path": "fasta.fa",
                            "size": 0,
                            "class": "File",
                            "secondaryFiles": []
                        },
                        "Dont_output_indexed_data": true
                    }
                },
                "stdin": "",
                "sbg:project": "marouf/samtools-1-3-demo",
                "description": "SAMtools Index FASTA indexes the reference sequence in a FASTA format or extracts subsequence from the indexed reference sequence. If no region is specified, faidx will index the file and create <ref.fasta>.fai on the disk. If regions are specified, the subsequences will be retrieved and printed to stdout in the FASTA format. The input file can be compressed in a RAZF format.",
                "x": 318.66669718424487,
                "sbg:createdOn": 1458812964,
                "outputs": [
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Output FASTA file",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  if ($job.inputs.Dont_output_indexed_data === true)\n  {\n    return ''\n  } \n  else \n    \n  {\n    return $job.inputs.input_fasta_file.path.split(\"/\").pop()\n\n  }\n}\n\n\n\n\n\n\n",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "secondaryFiles": [
                                ".fai"
                            ],
                            "sbg:inheritMetadataFrom": "#input_fasta_file"
                        },
                        "sbg:fileTypes": "FASTA",
                        "id": "#output_fasta_file",
                        "description": "Output FASTA file with index (FAI) file."
                    },
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Generated index file",
                        "outputBinding": {
                            "glob": "*.fai",
                            "sbg:inheritMetadataFrom": "#input_fasta_file"
                        },
                        "sbg:fileTypes": "FAI",
                        "id": "#generated_index",
                        "description": "Generated index file (without the indexed data)."
                    }
                ],
                "sbg:links": [
                    {
                        "label": "Homepage",
                        "id": "http://www.htslib.org"
                    },
                    {
                        "label": "Source code",
                        "id": "https://github.com/samtools/"
                    },
                    {
                        "label": "Wiki",
                        "id": "http://www.htslib.org"
                    },
                    {
                        "label": "Download",
                        "id": "https://sourceforge.net/projects/samtools/files/samtools/"
                    },
                    {
                        "label": "Publication",
                        "id": "http://www.ncbi.nlm.nih.gov/pubmed/19505943"
                    },
                    {
                        "label": "Documentation",
                        "id": "http://www.htslib.org/doc/samtools.html"
                    }
                ],
                "sbg:contributors": [
                    "marouf"
                ],
                "sbg:modifiedOn": 1480425012,
                "sbg:sbgMaintained": false,
                "stdout": ""
            },
            "inputs": [
                {
                    "id": "#SAMtools_Index_FASTA.input_index_file"
                },
                {
                    "id": "#SAMtools_Index_FASTA.input_fasta_file",
                    "source": [
                        "#input_fasta_file"
                    ]
                },
                {
                    "id": "#SAMtools_Index_FASTA.Dont_output_indexed_data"
                }
            ],
            "id": "#SAMtools_Index_FASTA",
            "sbg:y": 333.8379720052084
        },
        {
            "sbg:x": 1511.2076822916672,
            "outputs": [
                {
                    "id": "#VCFtools_Concat_SNPs.output_file"
                }
            ],
            "run": {
                "sbg:toolkit": "VCFtools",
                "requirements": [
                    {
                        "requirements": [
                            {
                                "class": "DockerRequirement",
                                "dockerPull": "rabix/js-engine"
                            }
                        ],
                        "class": "ExpressionEngineRequirement",
                        "engineCommand": "cwl-engine.js",
                        "id": "#cwl-js-engine"
                    }
                ],
                "sbg:validationErrors": [],
                "sbg:categories": [
                    "VCF-Processing"
                ],
                "label": "VCFtools Concat SNPs",
                "sbg:createdBy": "bix-demo",
                "sbg:cmdPreview": "vcf-concat sample1.vcf sample2.vcf  > sample1.and_more.concat.vcf  && bgzip -c -f sample1.and_more.concat.vcf > sample1.and_more.concat.vcf.gz",
                "sbg:image_url": null,
                "cwlVersion": "sbg:draft-2",
                "sbg:revisionsInfo": [
                    {
                        "sbg:revision": 0,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911615
                    },
                    {
                        "sbg:revision": 1,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911616
                    },
                    {
                        "sbg:revision": 2,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911617
                    },
                    {
                        "sbg:revision": 3,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911617
                    },
                    {
                        "sbg:revision": 4,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911618
                    },
                    {
                        "sbg:revision": 5,
                        "sbg:revisionNotes": "Added -c option",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1472826063
                    },
                    {
                        "sbg:revision": 6,
                        "sbg:revisionNotes": "Changed info",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1472826160
                    },
                    {
                        "sbg:revision": 7,
                        "sbg:revisionNotes": "Changed concat",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1474591521
                    },
                    {
                        "sbg:revision": 8,
                        "sbg:revisionNotes": "Overhaul done",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1480431144
                    },
                    {
                        "sbg:revision": 9,
                        "sbg:revisionNotes": "Added flattens",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1480431543
                    }
                ],
                "y": 284.8053181966147,
                "sbg:toolkitVersion": "0.1.14",
                "sbg:license": "GNU General Public License version 3.0 (GPLv3)",
                "sbg:modifiedBy": "ognjenm",
                "class": "CommandLineTool",
                "inputs": [
                    {
                        "label": "Small overlap",
                        "type": [
                            "null",
                            "int"
                        ],
                        "inputBinding": {
                            "prefix": "-s",
                            "sbg:cmdInclude": true,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "None",
                        "description": "Allow small overlaps in N consecutive files (when run with the this option, only one chromosome can be present).",
                        "id": "#small_overlap",
                        "sbg:category": "Execution",
                        "sbg:altPrefix": "--merge-sort"
                    },
                    {
                        "label": "Pad missing",
                        "inputBinding": {
                            "prefix": "-p",
                            "sbg:cmdInclude": true,
                            "separate": false
                        },
                        "sbg:toolDefaultValue": "FALSE",
                        "description": "Write '.' in place of missing columns. Useful for joining chrY with the rest.",
                        "id": "#pad_missing",
                        "sbg:category": "Execution",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "sbg:fileTypes": "VCF, VCF.GZ",
                        "label": "Input files",
                        "required": true,
                        "inputBinding": {
                            "sbg:cmdInclude": true,
                            "position": 20,
                            "itemSeparator": null,
                            "separate": false
                        },
                        "description": "Input array of files (vcf or vcf.gz).",
                        "id": "#input_files",
                        "sbg:category": "File Input",
                        "type": [
                            {
                                "type": "array",
                                "items": "File"
                            }
                        ]
                    },
                    {
                        "sbg:fileTypes": "TXT",
                        "label": "List of files",
                        "required": false,
                        "inputBinding": {
                            "prefix": "-f",
                            "sbg:cmdInclude": true,
                            "separate": true
                        },
                        "description": "Read the list of files from a file.",
                        "id": "#file_list",
                        "sbg:category": "File Input",
                        "type": [
                            "null",
                            "File"
                        ]
                    },
                    {
                        "sbg:stageInput": null,
                        "type": [
                            "null",
                            "boolean"
                        ],
                        "label": "Compressed output",
                        "sbg:toolDefaultValue": "FALSE",
                        "id": "#compressed",
                        "sbg:category": "Execution",
                        "description": "Check to make the output compressed (usually for further processing)."
                    },
                    {
                        "sbg:stageInput": null,
                        "type": [
                            "null",
                            "boolean"
                        ],
                        "label": "Check columns",
                        "sbg:toolDefaultValue": "FALSE",
                        "inputBinding": {
                            "prefix": "-c",
                            "sbg:cmdInclude": true,
                            "position": 5,
                            "separate": true
                        },
                        "description": "Do not concatenate, only check if the columns agree.",
                        "id": "#check_columns",
                        "sbg:category": "Execution",
                        "sbg:altPrefix": "--check_columns"
                    }
                ],
                "hints": [
                    {
                        "dockerImageId": "54cb823ba25a",
                        "class": "DockerRequirement",
                        "dockerPull": "images.sbgenomics.com/ognjenm/vcftools:0.1.14"
                    },
                    {
                        "value": 1,
                        "class": "sbg:CPURequirement"
                    },
                    {
                        "value": 1000,
                        "class": "sbg:MemRequirement"
                    }
                ],
                "sbg:revision": 9,
                "baseCommand": [
                    "vcf-concat"
                ],
                "id": "bix-demo/vcftools-0-1-14-demo/vcftools-concat-0-1-14/9",
                "sbg:latestRevision": 9,
                "temporaryFailCodes": [],
                "successCodes": [],
                "sbg:toolAuthor": "Adam Auton, Petr Danecek, Anthony Marcketta",
                "sbg:revisionNotes": "Added flattens",
                "arguments": [
                    {
                        "position": 50,
                        "valueFrom": {
                            "script": "{\n      function flatten(array){\n  \n\n  \n  for(i=0;i<array.length;i++) {\n    if (array[i] instanceof Array) {\n    \tarray = array.concat(array[i])\n        array.splice(i, 1)\n        i -= 1\n    }\n  }\n  \n  return array\n  \n}\n  \n  inp = $job.inputs.input_files\n  inp = flatten([].concat( inp ));\n  \n  if (inp.length > 1) {\n     sufix = \"and_more.concat\"\n  } else {\n     sufix = \"concat\"\n  }\n  \n  sufix_ext = \"vcf\";\n  filepath = inp[0].path\n  filename = filepath.split(\"/\").pop()\n  \n  if (filename.lastIndexOf(\".vcf.gz\") != -1) {\n    basename = filename.substr(0,filename.lastIndexOf(\".vcf.gz\"))\n  } else {\n    basename = filename.substr(0,filename.lastIndexOf(\".\"))\n  }\n\n  new_filename = basename + \".\" + sufix + \".\" + sufix_ext;\n\n  return '> ' + new_filename\n  \n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "separate": true
                    },
                    {
                        "position": 100,
                        "valueFrom": {
                            "script": "{\n      function flatten(array){\n  \n\n  \n  for(i=0;i<array.length;i++) {\n    if (array[i] instanceof Array) {\n    \tarray = array.concat(array[i])\n        array.splice(i, 1)\n        i -= 1\n    }\n  }\n  \n  return array\n  \n}\n  \n  inp = $job.inputs.input_files\n  inp = flatten([].concat( inp ));\n  \n  if (inp.length > 1) {\n     sufix = \"and_more.concat\"\n  } else {\n     sufix = \"concat\"\n  }\n  \n  sufix_ext = \"vcf\";\n  filepath = inp[0].path\n  filename = filepath.split(\"/\").pop()\n  \n  if (filename.lastIndexOf(\".vcf.gz\") != -1) {\n    basename = filename.substr(0,filename.lastIndexOf(\".vcf.gz\"))\n  } else {\n    basename = filename.substr(0,filename.lastIndexOf(\".\"))\n  }\n\n  new_filename = basename + \".\" + sufix + \".\" + sufix_ext;\n\n  if ($job.inputs.compressed) {\n\n  \treturn '&& bgzip -c -f ' + new_filename + ' > ' + new_filename + '.gz'\n      \n  } else {\n    \n    return ''\n  }\n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "separate": true
                    }
                ],
                "sbg:id": "admin/sbg-public-data/vcftools-concat-0-1-14/9",
                "sbg:job": {
                    "allocatedResources": {
                        "cpu": 1,
                        "mem": 1000
                    },
                    "inputs": {
                        "input_files": [
                            {
                                "path": "sample1.vcf"
                            },
                            {
                                "path": "sample2.vcf"
                            }
                        ],
                        "file_list": {
                            "path": "file_list.ext",
                            "size": 0,
                            "class": "File",
                            "secondaryFiles": []
                        },
                        "compressed": true,
                        "check_columns": true,
                        "small_overlap": 0,
                        "pad_missing": true
                    }
                },
                "stdin": "",
                "sbg:project": "bix-demo/vcftools-0-1-14-demo",
                "description": "VCFtools concat concatenates VCF files (for example, splitting by chromosome). Note that the input and output VCFs will have the same number of columns. The script does not merge VCFs by position (see also vcf-merge).\n\nIn the default mode, this tool only checks that all files have the same columns. When run with the -s option, it will perform a partial merge sort and look at a limited number of open files simultaneously.",
                "x": 1511.2076822916672,
                "sbg:createdOn": 1450911615,
                "outputs": [
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Concatenated vcf file",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n    function flatten(array){\n  \n\n  \n  for(i=0;i<array.length;i++) {\n    if (array[i] instanceof Array) {\n    \tarray = array.concat(array[i])\n        array.splice(i, 1)\n        i -= 1\n    }\n  }\n  \n  return array\n  \n}\n  \n  inp = $job.inputs.input_files\n  inp = flatten([].concat( inp ));\n  \n  if (inp.length > 1) {\n     sufix = \"and_more.concat\"\n  } else {\n     sufix = \"concat\"\n  }\n  \n  sufix_ext = \"vcf\";\n  filepath = inp[0].path\n  filename = filepath.split(\"/\").pop()\n  \n  if (filename.lastIndexOf(\".vcf.gz\") != -1) {\n    basename = filename.substr(0,filename.lastIndexOf(\".vcf.gz\"))\n  } else {\n    basename = filename.substr(0,filename.lastIndexOf(\".\"))\n  }\n\n  new_filename = basename + \".\" + sufix + \".\" + sufix_ext;\n\n  if ($job.inputs.compressed) {\n\n  \treturn new_filename + '.gz'\n      \n  } else {\n  \n  \treturn new_filename\n  }\n}",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:inheritMetadataFrom": "#input_files"
                        },
                        "sbg:fileTypes": "VCF,VCF.GZ",
                        "id": "#output_file",
                        "description": "Concatenated files from the input as a single vcf file."
                    }
                ],
                "sbg:links": [
                    {
                        "label": "Homepage",
                        "id": "https://vcftools.github.io"
                    },
                    {
                        "label": "Source code",
                        "id": "https://github.com/vcftools/vcftools"
                    },
                    {
                        "label": "Publications",
                        "id": "http://bioinformatics.oxfordjournals.org/content/27/15/2156"
                    },
                    {
                        "label": "Download",
                        "id": "https://github.com/vcftools/vcftools/zipball/master"
                    },
                    {
                        "label": "Wiki",
                        "id": "https://sourceforge.net/p/vcftools/wiki/Home/"
                    }
                ],
                "sbg:contributors": [
                    "bix-demo",
                    "ognjenm"
                ],
                "sbg:modifiedOn": 1480431543,
                "sbg:sbgMaintained": false,
                "stdout": ""
            },
            "inputs": [
                {
                    "id": "#VCFtools_Concat_SNPs.small_overlap"
                },
                {
                    "id": "#VCFtools_Concat_SNPs.pad_missing"
                },
                {
                    "id": "#VCFtools_Concat_SNPs.input_files",
                    "source": [
                        "#VarScan2_ProcessSomatic_SNPs.somatic_hc"
                    ]
                },
                {
                    "id": "#VCFtools_Concat_SNPs.file_list"
                },
                {
                    "id": "#VCFtools_Concat_SNPs.compressed"
                },
                {
                    "id": "#VCFtools_Concat_SNPs.check_columns"
                }
            ],
            "id": "#VCFtools_Concat_SNPs",
            "sbg:y": 284.8053181966147
        },
        {
            "sbg:x": 1663.7606430941382,
            "outputs": [
                {
                    "id": "#VCFtools_Sort.output_file"
                }
            ],
            "run": {
                "sbg:toolkit": "VCFtools",
                "requirements": [
                    {
                        "requirements": [
                            {
                                "class": "DockerRequirement",
                                "dockerPull": "rabix/js-engine"
                            }
                        ],
                        "class": "ExpressionEngineRequirement",
                        "engineCommand": "cwl-engine.js",
                        "id": "#cwl-js-engine"
                    }
                ],
                "sbg:validationErrors": [],
                "sbg:categories": [
                    "VCF-Processing"
                ],
                "label": "VCFtools Sort",
                "sbg:createdBy": "bix-demo",
                "sbg:cmdPreview": "vcf-sort sample1.vcf  > sample1.sorted.vcf  && bgzip -c -f sample1.sorted.vcf > sample1.sorted.vcf.gz",
                "sbg:image_url": null,
                "cwlVersion": "sbg:draft-2",
                "sbg:revisionsInfo": [
                    {
                        "sbg:revision": 0,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911603
                    },
                    {
                        "sbg:revision": 1,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911604
                    },
                    {
                        "sbg:revision": 2,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911604
                    },
                    {
                        "sbg:revision": 3,
                        "sbg:revisionNotes": "Changed memory, CPU and added parallel",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1472821172
                    },
                    {
                        "sbg:revision": 4,
                        "sbg:revisionNotes": "Compressable output",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1474285644
                    },
                    {
                        "sbg:revision": 5,
                        "sbg:revisionNotes": "Changed --parallel to -p",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1478007439
                    },
                    {
                        "sbg:revision": 6,
                        "sbg:revisionNotes": "Changed CPU",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1480089729
                    },
                    {
                        "sbg:revision": 7,
                        "sbg:revisionNotes": "Added inherit metadata from input.",
                        "sbg:modifiedBy": "milan.domazet.sudo",
                        "sbg:modifiedOn": 1482334199
                    }
                ],
                "y": 284.09669379313533,
                "sbg:toolkitVersion": "0.1.14",
                "sbg:license": "GNU General Public License version 3.0 (GPLv3)",
                "sbg:modifiedBy": "milan.domazet.sudo",
                "class": "CommandLineTool",
                "inputs": [
                    {
                        "label": "Parallel threads",
                        "inputBinding": {
                            "prefix": "-p",
                            "sbg:cmdInclude": true,
                            "position": 2,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "1",
                        "description": "Change the number of sorts run concurrently to <int>.",
                        "id": "#parallel",
                        "sbg:category": "Execution",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "sbg:stageInput": null,
                        "type": [
                            "null",
                            "int"
                        ],
                        "label": "Memory in MB",
                        "sbg:toolDefaultValue": "N/A",
                        "id": "#mem_mb",
                        "sbg:category": "Execution",
                        "description": "Memory in MB for execution."
                    },
                    {
                        "sbg:fileTypes": "VCF, VCF.GZ",
                        "label": "Input file",
                        "required": true,
                        "inputBinding": {
                            "sbg:cmdInclude": true,
                            "position": 1,
                            "separate": false
                        },
                        "description": "Input file (vcf or vcf.gz)",
                        "id": "#input_file",
                        "type": [
                            "File"
                        ]
                    },
                    {
                        "sbg:stageInput": null,
                        "type": [
                            "null",
                            "boolean"
                        ],
                        "label": "Compressed output",
                        "sbg:toolDefaultValue": "FALSE",
                        "id": "#compressed",
                        "sbg:category": "Execution",
                        "description": "Check to make the output compressed (usually for further processing)."
                    },
                    {
                        "inputBinding": {
                            "prefix": "-c",
                            "sbg:cmdInclude": true,
                            "separate": true
                        },
                        "description": "Use natural ordering (1,2,10,MT,X) rather then the default (1,10,2,MT,X). This requires                                      new version of the unix \"sort\" command which supports the --version-sort option.",
                        "label": "Chromosomal order",
                        "id": "#chromosomal_order",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    }
                ],
                "hints": [
                    {
                        "dockerImageId": "54cb823ba25a",
                        "class": "DockerRequirement",
                        "dockerPull": "images.sbgenomics.com/ognjenm/vcftools:0.1.14"
                    },
                    {
                        "value": {
                            "script": "{\n  if ($job.inputs.parallel) {\n  \treturn $job.inputs.parallel\n  } else {\n    return 1\n  }  \n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "class": "sbg:CPURequirement"
                    },
                    {
                        "value": {
                            "script": "{\n  if ($job.inputs.mem_mb) {\n    \n    return $job.inputs.mem_mb\n    \n   }  else {\n      \n      return 1000\n      \n      }\n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "class": "sbg:MemRequirement"
                    }
                ],
                "sbg:revision": 7,
                "baseCommand": [
                    "vcf-sort"
                ],
                "id": "bix-demo/vcftools-0-1-14-demo/vcftools-sort-0-1-14/7",
                "sbg:latestRevision": 7,
                "temporaryFailCodes": [],
                "successCodes": [],
                "sbg:toolAuthor": "Adam Auton, Petr Danecek, Anthony Marcketta",
                "sbg:revisionNotes": "Added inherit metadata from input.",
                "arguments": [
                    {
                        "position": 100,
                        "valueFrom": {
                            "script": "{\n\n\n  if ($job.inputs.compressed) {\n    \n  \tfilepath = $job.inputs.input_file.path\n\n  \tfilename = filepath.split(\"/\").pop();\n\n  \tfile_dot_sep = filename.split(\".\");\n  \tfile_ext = file_dot_sep[file_dot_sep.length-1];\n\n  \tnew_filename = filename.substr(0,filename.lastIndexOf(\".vcf\")) + \".sorted.vcf\";\n  \treturn '&& bgzip -c -f ' + new_filename + ' > ' + new_filename + '.gz'\n      \n  }\n  \n  \n  \n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "separate": true
                    },
                    {
                        "position": 50,
                        "valueFrom": {
                            "script": "{\n  filepath = $job.inputs.input_file.path\n\n  filename = filepath.split(\"/\").pop();\n\n  file_dot_sep = filename.split(\".\");\n  file_ext = file_dot_sep[file_dot_sep.length-1];\n\n  new_filename = filename.substr(0,filename.lastIndexOf(\".vcf\")) + \".sorted.vcf\";\n  \n  return '> '+new_filename;\n  \n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "separate": true
                    }
                ],
                "sbg:id": "admin/sbg-public-data/vcftools-sort-0-1-14/7",
                "sbg:job": {
                    "allocatedResources": {
                        "cpu": 2,
                        "mem": 100
                    },
                    "inputs": {
                        "mem_mb": 100,
                        "parallel": 2,
                        "compressed": true,
                        "input_file": {
                            "path": "sample1.vcf"
                        }
                    }
                },
                "stdin": "",
                "sbg:project": "bix-demo/vcftools-0-1-14-demo",
                "description": "VCFtools sort sorts a VCF file.",
                "x": 1663.7606430941382,
                "sbg:createdOn": 1450911603,
                "outputs": [
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  filepath = $job.inputs.input_file.path\n\n  filename = filepath.split(\"/\").pop();\n\n  file_dot_sep = filename.split(\".\");\n  file_ext = file_dot_sep[file_dot_sep.length-1];\n\n  new_filename = filename.substr(0,filename.lastIndexOf(\".vcf\")) + \".sorted.vcf\";\n\n  if ($job.inputs.compressed) {\n      new_filename += \".gz\";\n  }\n  \n  return new_filename;\n  \n}",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:metadata": {
                                "__inherit__": "input_file"
                            },
                            "sbg:inheritMetadataFrom": "#input_file"
                        },
                        "label": "Output file",
                        "id": "#output_file",
                        "sbg:fileTypes": "VCF, VCF.GZ"
                    }
                ],
                "sbg:links": [
                    {
                        "label": "Homepage",
                        "id": "https://vcftools.github.io"
                    },
                    {
                        "label": "Source code",
                        "id": "https://github.com/vcftools/vcftools"
                    },
                    {
                        "label": "Publications",
                        "id": "http://bioinformatics.oxfordjournals.org/content/27/15/2156"
                    }
                ],
                "sbg:contributors": [
                    "milan.domazet.sudo",
                    "bix-demo",
                    "ognjenm"
                ],
                "sbg:modifiedOn": 1482334199,
                "sbg:sbgMaintained": false,
                "stdout": ""
            },
            "inputs": [
                {
                    "id": "#VCFtools_Sort.parallel"
                },
                {
                    "id": "#VCFtools_Sort.mem_mb"
                },
                {
                    "id": "#VCFtools_Sort.input_file",
                    "source": [
                        "#VCFtools_Concat_SNPs.output_file"
                    ]
                },
                {
                    "id": "#VCFtools_Sort.compressed"
                },
                {
                    "id": "#VCFtools_Sort.chromosomal_order"
                }
            ],
            "id": "#VCFtools_Sort",
            "sbg:y": 284.09669379313533
        },
        {
            "sbg:x": 1375.0001424786312,
            "outputs": [
                {
                    "id": "#VCFtools_Concat_INDELs.output_file"
                }
            ],
            "run": {
                "sbg:toolkit": "VCFtools",
                "requirements": [
                    {
                        "requirements": [
                            {
                                "class": "DockerRequirement",
                                "dockerPull": "rabix/js-engine"
                            }
                        ],
                        "class": "ExpressionEngineRequirement",
                        "engineCommand": "cwl-engine.js",
                        "id": "#cwl-js-engine"
                    }
                ],
                "sbg:validationErrors": [],
                "sbg:categories": [
                    "VCF-Processing"
                ],
                "label": "VCFtools Concat INDELs",
                "sbg:createdBy": "bix-demo",
                "sbg:cmdPreview": "vcf-concat sample1.vcf sample2.vcf  > sample1.and_more.concat.vcf  && bgzip -c -f sample1.and_more.concat.vcf > sample1.and_more.concat.vcf.gz",
                "sbg:image_url": null,
                "cwlVersion": "sbg:draft-2",
                "sbg:revisionsInfo": [
                    {
                        "sbg:revision": 0,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911615
                    },
                    {
                        "sbg:revision": 1,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911616
                    },
                    {
                        "sbg:revision": 2,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911617
                    },
                    {
                        "sbg:revision": 3,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911617
                    },
                    {
                        "sbg:revision": 4,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911618
                    },
                    {
                        "sbg:revision": 5,
                        "sbg:revisionNotes": "Added -c option",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1472826063
                    },
                    {
                        "sbg:revision": 6,
                        "sbg:revisionNotes": "Changed info",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1472826160
                    },
                    {
                        "sbg:revision": 7,
                        "sbg:revisionNotes": "Changed concat",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1474591521
                    },
                    {
                        "sbg:revision": 8,
                        "sbg:revisionNotes": "Overhaul done",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1480431144
                    },
                    {
                        "sbg:revision": 9,
                        "sbg:revisionNotes": "Added flattens",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1480431543
                    }
                ],
                "y": 429.04766971237996,
                "sbg:toolkitVersion": "0.1.14",
                "sbg:license": "GNU General Public License version 3.0 (GPLv3)",
                "sbg:modifiedBy": "ognjenm",
                "class": "CommandLineTool",
                "inputs": [
                    {
                        "label": "Small overlap",
                        "type": [
                            "null",
                            "int"
                        ],
                        "inputBinding": {
                            "prefix": "-s",
                            "sbg:cmdInclude": true,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "None",
                        "description": "Allow small overlaps in N consecutive files (when run with the this option, only one chromosome can be present).",
                        "id": "#small_overlap",
                        "sbg:category": "Execution",
                        "sbg:altPrefix": "--merge-sort"
                    },
                    {
                        "label": "Pad missing",
                        "inputBinding": {
                            "prefix": "-p",
                            "sbg:cmdInclude": true,
                            "separate": false
                        },
                        "sbg:toolDefaultValue": "FALSE",
                        "description": "Write '.' in place of missing columns. Useful for joining chrY with the rest.",
                        "id": "#pad_missing",
                        "sbg:category": "Execution",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "sbg:fileTypes": "VCF, VCF.GZ",
                        "label": "Input files",
                        "required": true,
                        "inputBinding": {
                            "sbg:cmdInclude": true,
                            "position": 20,
                            "itemSeparator": null,
                            "separate": false
                        },
                        "description": "Input array of files (vcf or vcf.gz).",
                        "id": "#input_files",
                        "sbg:category": "File Input",
                        "type": [
                            {
                                "type": "array",
                                "items": "File"
                            }
                        ]
                    },
                    {
                        "sbg:fileTypes": "TXT",
                        "label": "List of files",
                        "required": false,
                        "inputBinding": {
                            "prefix": "-f",
                            "sbg:cmdInclude": true,
                            "separate": true
                        },
                        "description": "Read the list of files from a file.",
                        "id": "#file_list",
                        "sbg:category": "File Input",
                        "type": [
                            "null",
                            "File"
                        ]
                    },
                    {
                        "sbg:stageInput": null,
                        "type": [
                            "null",
                            "boolean"
                        ],
                        "label": "Compressed output",
                        "sbg:toolDefaultValue": "FALSE",
                        "id": "#compressed",
                        "sbg:category": "Execution",
                        "description": "Check to make the output compressed (usually for further processing)."
                    },
                    {
                        "sbg:stageInput": null,
                        "type": [
                            "null",
                            "boolean"
                        ],
                        "label": "Check columns",
                        "sbg:toolDefaultValue": "FALSE",
                        "inputBinding": {
                            "prefix": "-c",
                            "sbg:cmdInclude": true,
                            "position": 5,
                            "separate": true
                        },
                        "description": "Do not concatenate, only check if the columns agree.",
                        "id": "#check_columns",
                        "sbg:category": "Execution",
                        "sbg:altPrefix": "--check_columns"
                    }
                ],
                "hints": [
                    {
                        "dockerImageId": "54cb823ba25a",
                        "class": "DockerRequirement",
                        "dockerPull": "images.sbgenomics.com/ognjenm/vcftools:0.1.14"
                    },
                    {
                        "value": 1,
                        "class": "sbg:CPURequirement"
                    },
                    {
                        "value": 1000,
                        "class": "sbg:MemRequirement"
                    }
                ],
                "sbg:revision": 9,
                "baseCommand": [
                    "vcf-concat"
                ],
                "id": "bix-demo/vcftools-0-1-14-demo/vcftools-concat-0-1-14/9",
                "sbg:latestRevision": 9,
                "temporaryFailCodes": [],
                "successCodes": [],
                "sbg:toolAuthor": "Adam Auton, Petr Danecek, Anthony Marcketta",
                "sbg:revisionNotes": "Added flattens",
                "arguments": [
                    {
                        "position": 50,
                        "valueFrom": {
                            "script": "{\n      function flatten(array){\n  \n\n  \n  for(i=0;i<array.length;i++) {\n    if (array[i] instanceof Array) {\n    \tarray = array.concat(array[i])\n        array.splice(i, 1)\n        i -= 1\n    }\n  }\n  \n  return array\n  \n}\n  \n  inp = $job.inputs.input_files\n  inp = flatten([].concat( inp ));\n  \n  if (inp.length > 1) {\n     sufix = \"and_more.concat\"\n  } else {\n     sufix = \"concat\"\n  }\n  \n  sufix_ext = \"vcf\";\n  filepath = inp[0].path\n  filename = filepath.split(\"/\").pop()\n  \n  if (filename.lastIndexOf(\".vcf.gz\") != -1) {\n    basename = filename.substr(0,filename.lastIndexOf(\".vcf.gz\"))\n  } else {\n    basename = filename.substr(0,filename.lastIndexOf(\".\"))\n  }\n\n  new_filename = basename + \".\" + sufix + \".\" + sufix_ext;\n\n  return '> ' + new_filename\n  \n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "separate": true
                    },
                    {
                        "position": 100,
                        "valueFrom": {
                            "script": "{\n      function flatten(array){\n  \n\n  \n  for(i=0;i<array.length;i++) {\n    if (array[i] instanceof Array) {\n    \tarray = array.concat(array[i])\n        array.splice(i, 1)\n        i -= 1\n    }\n  }\n  \n  return array\n  \n}\n  \n  inp = $job.inputs.input_files\n  inp = flatten([].concat( inp ));\n  \n  if (inp.length > 1) {\n     sufix = \"and_more.concat\"\n  } else {\n     sufix = \"concat\"\n  }\n  \n  sufix_ext = \"vcf\";\n  filepath = inp[0].path\n  filename = filepath.split(\"/\").pop()\n  \n  if (filename.lastIndexOf(\".vcf.gz\") != -1) {\n    basename = filename.substr(0,filename.lastIndexOf(\".vcf.gz\"))\n  } else {\n    basename = filename.substr(0,filename.lastIndexOf(\".\"))\n  }\n\n  new_filename = basename + \".\" + sufix + \".\" + sufix_ext;\n\n  if ($job.inputs.compressed) {\n\n  \treturn '&& bgzip -c -f ' + new_filename + ' > ' + new_filename + '.gz'\n      \n  } else {\n    \n    return ''\n  }\n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "separate": true
                    }
                ],
                "sbg:id": "admin/sbg-public-data/vcftools-concat-0-1-14/9",
                "sbg:job": {
                    "allocatedResources": {
                        "cpu": 1,
                        "mem": 1000
                    },
                    "inputs": {
                        "input_files": [
                            {
                                "path": "sample1.vcf"
                            },
                            {
                                "path": "sample2.vcf"
                            }
                        ],
                        "file_list": {
                            "path": "file_list.ext",
                            "size": 0,
                            "class": "File",
                            "secondaryFiles": []
                        },
                        "compressed": true,
                        "check_columns": true,
                        "small_overlap": 0,
                        "pad_missing": true
                    }
                },
                "stdin": "",
                "sbg:project": "bix-demo/vcftools-0-1-14-demo",
                "description": "VCFtools concat concatenates VCF files (for example, splitting by chromosome). Note that the input and output VCFs will have the same number of columns. The script does not merge VCFs by position (see also vcf-merge).\n\nIn the default mode, this tool only checks that all files have the same columns. When run with the -s option, it will perform a partial merge sort and look at a limited number of open files simultaneously.",
                "x": 1375.0001424786312,
                "sbg:createdOn": 1450911615,
                "outputs": [
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "label": "Concatenated vcf file",
                        "outputBinding": {
                            "glob": {
                                "script": "{\n    function flatten(array){\n  \n\n  \n  for(i=0;i<array.length;i++) {\n    if (array[i] instanceof Array) {\n    \tarray = array.concat(array[i])\n        array.splice(i, 1)\n        i -= 1\n    }\n  }\n  \n  return array\n  \n}\n  \n  inp = $job.inputs.input_files\n  inp = flatten([].concat( inp ));\n  \n  if (inp.length > 1) {\n     sufix = \"and_more.concat\"\n  } else {\n     sufix = \"concat\"\n  }\n  \n  sufix_ext = \"vcf\";\n  filepath = inp[0].path\n  filename = filepath.split(\"/\").pop()\n  \n  if (filename.lastIndexOf(\".vcf.gz\") != -1) {\n    basename = filename.substr(0,filename.lastIndexOf(\".vcf.gz\"))\n  } else {\n    basename = filename.substr(0,filename.lastIndexOf(\".\"))\n  }\n\n  new_filename = basename + \".\" + sufix + \".\" + sufix_ext;\n\n  if ($job.inputs.compressed) {\n\n  \treturn new_filename + '.gz'\n      \n  } else {\n  \n  \treturn new_filename\n  }\n}",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:inheritMetadataFrom": "#input_files"
                        },
                        "sbg:fileTypes": "VCF,VCF.GZ",
                        "id": "#output_file",
                        "description": "Concatenated files from the input as a single vcf file."
                    }
                ],
                "sbg:links": [
                    {
                        "label": "Homepage",
                        "id": "https://vcftools.github.io"
                    },
                    {
                        "label": "Source code",
                        "id": "https://github.com/vcftools/vcftools"
                    },
                    {
                        "label": "Publications",
                        "id": "http://bioinformatics.oxfordjournals.org/content/27/15/2156"
                    },
                    {
                        "label": "Download",
                        "id": "https://github.com/vcftools/vcftools/zipball/master"
                    },
                    {
                        "label": "Wiki",
                        "id": "https://sourceforge.net/p/vcftools/wiki/Home/"
                    }
                ],
                "sbg:contributors": [
                    "bix-demo",
                    "ognjenm"
                ],
                "sbg:modifiedOn": 1480431543,
                "sbg:sbgMaintained": false,
                "stdout": ""
            },
            "inputs": [
                {
                    "id": "#VCFtools_Concat_INDELs.small_overlap"
                },
                {
                    "id": "#VCFtools_Concat_INDELs.pad_missing"
                },
                {
                    "id": "#VCFtools_Concat_INDELs.input_files",
                    "source": [
                        "#VarScan2_ProcessSomatic_INDELs.somatic_hc"
                    ]
                },
                {
                    "id": "#VCFtools_Concat_INDELs.file_list"
                },
                {
                    "id": "#VCFtools_Concat_INDELs.compressed"
                },
                {
                    "id": "#VCFtools_Concat_INDELs.check_columns"
                }
            ],
            "id": "#VCFtools_Concat_INDELs",
            "sbg:y": 429.04766971237996
        },
        {
            "sbg:x": 1528.613932751045,
            "outputs": [
                {
                    "id": "#VCFtools_Sort_1.output_file"
                }
            ],
            "run": {
                "sbg:toolkit": "VCFtools",
                "requirements": [
                    {
                        "requirements": [
                            {
                                "class": "DockerRequirement",
                                "dockerPull": "rabix/js-engine"
                            }
                        ],
                        "class": "ExpressionEngineRequirement",
                        "engineCommand": "cwl-engine.js",
                        "id": "#cwl-js-engine"
                    }
                ],
                "sbg:validationErrors": [],
                "sbg:categories": [
                    "VCF-Processing"
                ],
                "label": "VCFtools Sort",
                "sbg:createdBy": "bix-demo",
                "sbg:cmdPreview": "vcf-sort sample1.vcf  > sample1.sorted.vcf  && bgzip -c -f sample1.sorted.vcf > sample1.sorted.vcf.gz",
                "sbg:image_url": null,
                "cwlVersion": "sbg:draft-2",
                "sbg:revisionsInfo": [
                    {
                        "sbg:revision": 0,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911603
                    },
                    {
                        "sbg:revision": 1,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911604
                    },
                    {
                        "sbg:revision": 2,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "bix-demo",
                        "sbg:modifiedOn": 1450911604
                    },
                    {
                        "sbg:revision": 3,
                        "sbg:revisionNotes": "Changed memory, CPU and added parallel",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1472821172
                    },
                    {
                        "sbg:revision": 4,
                        "sbg:revisionNotes": "Compressable output",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1474285644
                    },
                    {
                        "sbg:revision": 5,
                        "sbg:revisionNotes": "Changed --parallel to -p",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1478007439
                    },
                    {
                        "sbg:revision": 6,
                        "sbg:revisionNotes": "Changed CPU",
                        "sbg:modifiedBy": "ognjenm",
                        "sbg:modifiedOn": 1480089729
                    },
                    {
                        "sbg:revision": 7,
                        "sbg:revisionNotes": "Added inherit metadata from input.",
                        "sbg:modifiedBy": "milan.domazet.sudo",
                        "sbg:modifiedOn": 1482334199
                    }
                ],
                "y": 428.7956309413542,
                "sbg:toolkitVersion": "0.1.14",
                "sbg:license": "GNU General Public License version 3.0 (GPLv3)",
                "sbg:modifiedBy": "milan.domazet.sudo",
                "class": "CommandLineTool",
                "inputs": [
                    {
                        "label": "Parallel threads",
                        "inputBinding": {
                            "prefix": "-p",
                            "sbg:cmdInclude": true,
                            "position": 2,
                            "separate": true
                        },
                        "sbg:toolDefaultValue": "1",
                        "description": "Change the number of sorts run concurrently to <int>.",
                        "id": "#parallel",
                        "sbg:category": "Execution",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "sbg:stageInput": null,
                        "type": [
                            "null",
                            "int"
                        ],
                        "label": "Memory in MB",
                        "sbg:toolDefaultValue": "N/A",
                        "id": "#mem_mb",
                        "sbg:category": "Execution",
                        "description": "Memory in MB for execution."
                    },
                    {
                        "sbg:fileTypes": "VCF, VCF.GZ",
                        "label": "Input file",
                        "required": true,
                        "inputBinding": {
                            "sbg:cmdInclude": true,
                            "position": 1,
                            "separate": false
                        },
                        "description": "Input file (vcf or vcf.gz)",
                        "id": "#input_file",
                        "type": [
                            "File"
                        ]
                    },
                    {
                        "sbg:stageInput": null,
                        "type": [
                            "null",
                            "boolean"
                        ],
                        "label": "Compressed output",
                        "sbg:toolDefaultValue": "FALSE",
                        "id": "#compressed",
                        "sbg:category": "Execution",
                        "description": "Check to make the output compressed (usually for further processing)."
                    },
                    {
                        "inputBinding": {
                            "prefix": "-c",
                            "sbg:cmdInclude": true,
                            "separate": true
                        },
                        "description": "Use natural ordering (1,2,10,MT,X) rather then the default (1,10,2,MT,X). This requires                                      new version of the unix \"sort\" command which supports the --version-sort option.",
                        "label": "Chromosomal order",
                        "id": "#chromosomal_order",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    }
                ],
                "hints": [
                    {
                        "dockerImageId": "54cb823ba25a",
                        "class": "DockerRequirement",
                        "dockerPull": "images.sbgenomics.com/ognjenm/vcftools:0.1.14"
                    },
                    {
                        "value": {
                            "script": "{\n  if ($job.inputs.parallel) {\n  \treturn $job.inputs.parallel\n  } else {\n    return 1\n  }  \n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "class": "sbg:CPURequirement"
                    },
                    {
                        "value": {
                            "script": "{\n  if ($job.inputs.mem_mb) {\n    \n    return $job.inputs.mem_mb\n    \n   }  else {\n      \n      return 1000\n      \n      }\n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "class": "sbg:MemRequirement"
                    }
                ],
                "sbg:revision": 7,
                "baseCommand": [
                    "vcf-sort"
                ],
                "id": "bix-demo/vcftools-0-1-14-demo/vcftools-sort-0-1-14/7",
                "sbg:latestRevision": 7,
                "temporaryFailCodes": [],
                "successCodes": [],
                "sbg:toolAuthor": "Adam Auton, Petr Danecek, Anthony Marcketta",
                "sbg:revisionNotes": "Added inherit metadata from input.",
                "arguments": [
                    {
                        "position": 100,
                        "valueFrom": {
                            "script": "{\n\n\n  if ($job.inputs.compressed) {\n    \n  \tfilepath = $job.inputs.input_file.path\n\n  \tfilename = filepath.split(\"/\").pop();\n\n  \tfile_dot_sep = filename.split(\".\");\n  \tfile_ext = file_dot_sep[file_dot_sep.length-1];\n\n  \tnew_filename = filename.substr(0,filename.lastIndexOf(\".vcf\")) + \".sorted.vcf\";\n  \treturn '&& bgzip -c -f ' + new_filename + ' > ' + new_filename + '.gz'\n      \n  }\n  \n  \n  \n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "separate": true
                    },
                    {
                        "position": 50,
                        "valueFrom": {
                            "script": "{\n  filepath = $job.inputs.input_file.path\n\n  filename = filepath.split(\"/\").pop();\n\n  file_dot_sep = filename.split(\".\");\n  file_ext = file_dot_sep[file_dot_sep.length-1];\n\n  new_filename = filename.substr(0,filename.lastIndexOf(\".vcf\")) + \".sorted.vcf\";\n  \n  return '> '+new_filename;\n  \n}",
                            "engine": "#cwl-js-engine",
                            "class": "Expression"
                        },
                        "separate": true
                    }
                ],
                "sbg:id": "admin/sbg-public-data/vcftools-sort-0-1-14/7",
                "sbg:job": {
                    "allocatedResources": {
                        "cpu": 2,
                        "mem": 100
                    },
                    "inputs": {
                        "mem_mb": 100,
                        "parallel": 2,
                        "compressed": true,
                        "input_file": {
                            "path": "sample1.vcf"
                        }
                    }
                },
                "stdin": "",
                "sbg:project": "bix-demo/vcftools-0-1-14-demo",
                "description": "VCFtools sort sorts a VCF file.",
                "x": 1528.613932751045,
                "sbg:createdOn": 1450911603,
                "outputs": [
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "outputBinding": {
                            "glob": {
                                "script": "{\n  filepath = $job.inputs.input_file.path\n\n  filename = filepath.split(\"/\").pop();\n\n  file_dot_sep = filename.split(\".\");\n  file_ext = file_dot_sep[file_dot_sep.length-1];\n\n  new_filename = filename.substr(0,filename.lastIndexOf(\".vcf\")) + \".sorted.vcf\";\n\n  if ($job.inputs.compressed) {\n      new_filename += \".gz\";\n  }\n  \n  return new_filename;\n  \n}",
                                "engine": "#cwl-js-engine",
                                "class": "Expression"
                            },
                            "sbg:metadata": {
                                "__inherit__": "input_file"
                            },
                            "sbg:inheritMetadataFrom": "#input_file"
                        },
                        "label": "Output file",
                        "id": "#output_file",
                        "sbg:fileTypes": "VCF, VCF.GZ"
                    }
                ],
                "sbg:links": [
                    {
                        "label": "Homepage",
                        "id": "https://vcftools.github.io"
                    },
                    {
                        "label": "Source code",
                        "id": "https://github.com/vcftools/vcftools"
                    },
                    {
                        "label": "Publications",
                        "id": "http://bioinformatics.oxfordjournals.org/content/27/15/2156"
                    }
                ],
                "sbg:contributors": [
                    "milan.domazet.sudo",
                    "bix-demo",
                    "ognjenm"
                ],
                "sbg:modifiedOn": 1482334199,
                "sbg:sbgMaintained": false,
                "stdout": ""
            },
            "inputs": [
                {
                    "id": "#VCFtools_Sort_1.parallel"
                },
                {
                    "id": "#VCFtools_Sort_1.mem_mb"
                },
                {
                    "id": "#VCFtools_Sort_1.input_file",
                    "source": [
                        "#VCFtools_Concat_INDELs.output_file"
                    ]
                },
                {
                    "id": "#VCFtools_Sort_1.compressed"
                },
                {
                    "id": "#VCFtools_Sort_1.chromosomal_order"
                }
            ],
            "id": "#VCFtools_Sort_1",
            "sbg:y": 428.7956309413542
        },
        {
            "sbg:x": 682.0830918351933,
            "outputs": [
                {
                    "id": "#SAMtools_Mpileup_parallel.output_file"
                }
            ],
            "run": {
                "sbg:toolkit": "SAMtools",
                "requirements": [],
                "sbg:validationErrors": [],
                "sbg:categories": [
                    "SAM/BAM-Processing"
                ],
                "class": "Workflow",
                "label": "SAMtools Mpileup parallel",
                "sbg:createdBy": "marouf",
                "sbg:image_url": "https://brood.sbgenomics.com/static/marouf/samtools-mpileup-parallel-demo/samtools-mpileup-parallel/22.png",
                "sbg:links": [
                    {
                        "label": "Homepage",
                        "id": "http://samtools.sourceforge.net/"
                    },
                    {
                        "label": "Source Code",
                        "id": "https://github.com/samtools/samtools"
                    },
                    {
                        "label": "Wiki",
                        "id": "http://sourceforge.net/p/samtools/wiki/Home/"
                    },
                    {
                        "label": "Download",
                        "id": "http://sourceforge.net/projects/samtools/files/samtools/0.1.18/"
                    },
                    {
                        "label": "Publication",
                        "id": "http://www.ncbi.nlm.nih.gov/pubmed/19505943"
                    },
                    {
                        "label": "Documentation",
                        "id": "http://www.htslib.org/doc/samtools.html"
                    }
                ],
                "sbg:revisionsInfo": [
                    {
                        "sbg:revision": 0,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1474020918
                    },
                    {
                        "sbg:revision": 1,
                        "sbg:revisionNotes": "Update workflow from dev project",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1474020953
                    },
                    {
                        "sbg:revision": 2,
                        "sbg:revisionNotes": "update tools in the pipeline",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1474972698
                    },
                    {
                        "sbg:revision": 3,
                        "sbg:revisionNotes": "Update merge pileup tool",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1474974158
                    },
                    {
                        "sbg:revision": 4,
                        "sbg:revisionNotes": "Change the description after peer review",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1474978330
                    },
                    {
                        "sbg:revision": 5,
                        "sbg:revisionNotes": "uncheck the required checkbox for the fai file input",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1474987689
                    },
                    {
                        "sbg:revision": 6,
                        "sbg:revisionNotes": "Add suggested files",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1474988027
                    },
                    {
                        "sbg:revision": 7,
                        "sbg:revisionNotes": "update the merge mpileup tool",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1475052533
                    },
                    {
                        "sbg:revision": 8,
                        "sbg:revisionNotes": "Change the description",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1475055294
                    },
                    {
                        "sbg:revision": 9,
                        "sbg:revisionNotes": "update merge_pileup tool",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1475058452
                    },
                    {
                        "sbg:revision": 10,
                        "sbg:revisionNotes": "Update the merge pileup and change the description",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1475083788
                    },
                    {
                        "sbg:revision": 11,
                        "sbg:revisionNotes": "change parameters settings",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1475237926
                    },
                    {
                        "sbg:revision": 12,
                        "sbg:revisionNotes": "provide better description and remove bed file",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1475238571
                    },
                    {
                        "sbg:revision": 13,
                        "sbg:revisionNotes": "more description",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1475655549
                    },
                    {
                        "sbg:revision": 14,
                        "sbg:revisionNotes": "change merge pileup",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1476440562
                    },
                    {
                        "sbg:revision": 15,
                        "sbg:revisionNotes": "change description",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1476794923
                    },
                    {
                        "sbg:revision": 16,
                        "sbg:revisionNotes": "Exposed mapq_threshold.",
                        "sbg:modifiedBy": "milan.domazet.sudo",
                        "sbg:modifiedOn": 1477483754
                    },
                    {
                        "sbg:revision": 17,
                        "sbg:revisionNotes": "Exposed mapq_threshold.",
                        "sbg:modifiedBy": "milan.domazet.sudo",
                        "sbg:modifiedOn": 1477483870
                    },
                    {
                        "sbg:revision": 18,
                        "sbg:revisionNotes": "Add filter to mpileu tool",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1478526969
                    },
                    {
                        "sbg:revision": 19,
                        "sbg:revisionNotes": "expose filter",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1478527215
                    },
                    {
                        "sbg:revision": 20,
                        "sbg:revisionNotes": "Updated all tools (mostly inherit metadata from input)",
                        "sbg:modifiedBy": "milan.domazet.sudo",
                        "sbg:modifiedOn": 1482334741
                    },
                    {
                        "sbg:revision": 21,
                        "sbg:revisionNotes": "fix metadata",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1482417615
                    },
                    {
                        "sbg:revision": 22,
                        "sbg:revisionNotes": "fix metadata issue when there is no metadata at inputs",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1482421868
                    }
                ],
                "sbg:toolkitVersion": "1.3",
                "sbg:license": "The MIT License",
                "sbg:canvas_x": -61,
                "inputs": [
                    {
                        "type": [
                            {
                                "type": "array",
                                "items": "File"
                            }
                        ],
                        "sbg:x": 627,
                        "label": "bams",
                        "sbg:fileTypes": "BAM",
                        "id": "#bams",
                        "sbg:y": 348,
                        "sbg:includeInPorts": true
                    },
                    {
                        "type": [
                            "File"
                        ],
                        "sbg:x": 624,
                        "sbg:includeInPorts": true,
                        "sbg:suggestedValue": {
                            "path": "5772b6d8507c1752674486e6",
                            "name": "human_g1k_v37_decoy.fasta",
                            "class": "File"
                        },
                        "sbg:fileTypes": "FASTA,FA,GZ",
                        "id": "#reference_fasta",
                        "sbg:y": 221,
                        "label": "reference_fasta"
                    },
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "sbg:x": 629,
                        "sbg:includeInPorts": true,
                        "sbg:suggestedValue": {
                            "path": "578cf949507c17681a3117e2",
                            "name": "human_g1k_v37_decoy.fasta.fai",
                            "class": "File"
                        },
                        "sbg:fileTypes": "FAI",
                        "id": "#fai_file",
                        "sbg:y": 469,
                        "label": "fai_file"
                    },
                    {
                        "sbg:includeInPorts": false,
                        "label": "Skip bases with baseQ/BAQ smaller than",
                        "required": false,
                        "sbg:toolDefaultValue": "13",
                        "description": "Minimum base quality for a base to be considered.",
                        "id": "#skip_bases_with_baq_smaller_than_defined",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "type": [
                            "null",
                            "int"
                        ],
                        "label": "Required flags",
                        "required": false,
                        "description": "Required flags: skip reads with mask bits unset.",
                        "id": "#required_flags",
                        "sbg:category": "Input options",
                        "sbg:includeInPorts": false
                    },
                    {
                        "type": [
                            "null",
                            "string"
                        ],
                        "label": "Region in which pileup is generated",
                        "required": false,
                        "description": "Region in which pileup is generated.",
                        "id": "#region_pileup_generation",
                        "sbg:category": "Input options",
                        "sbg:includeInPorts": false
                    },
                    {
                        "type": [
                            "null",
                            "boolean"
                        ],
                        "label": "Recalculate BAQ on the fly",
                        "required": false,
                        "description": "Recalculate BAQ on the fly, ignore existing BQ tags.",
                        "id": "#recalc_BAQ_fly",
                        "sbg:category": "Input options",
                        "sbg:includeInPorts": false
                    },
                    {
                        "sbg:stageInput": null,
                        "label": "Apply -m and -F thresholds per sample",
                        "description": "Apply -m and -F thresholds per sample to increase sensitivity of calling. By default both options are applied to reads pooled from all samples.",
                        "id": "#per_sample_mF",
                        "sbg:category": "configuration",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "sbg:includeInPorts": false,
                        "label": "Parameter for adjusting mapQ",
                        "required": false,
                        "sbg:toolDefaultValue": "0",
                        "description": "Coefficient for downgrading mapping quality for reads containing excessive mismatches. Given a read with a phred-scaled probability q of being generated from the mapped position, the new mapping quality is about sqrt((INT-q)/INT)*INT. A zero value disables this functionality; if enabled, the recommended value for BWA is 50.",
                        "id": "#parameter_for_adjusting_mapq",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "description": "Similar to -g except that the output is uncompressed BCF, which is preferred for piping.",
                        "id": "#output_uncompressed_bcf_or_vcf",
                        "label": "Generate uncompress BCF/VCF output",
                        "sbg:category": "Output options",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "description": "Output mapping quality (disabled by -g/-u).",
                        "id": "#output_mapping_quality",
                        "label": "Output mapping quality",
                        "sbg:category": "Output options",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "description": "Output file format.",
                        "id": "#output_format",
                        "label": "Output file format",
                        "sbg:category": "configuration",
                        "type": [
                            {
                                "name": "output_format",
                                "symbols": [
                                    "PILEUP",
                                    "BCF",
                                    "VCF"
                                ],
                                "type": "enum"
                            }
                        ]
                    },
                    {
                        "description": "Output base positions on reads (disabled by -g/-u).",
                        "id": "#output_base_positions_on_reads",
                        "label": "Output base positions on reads",
                        "sbg:category": "Output options",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "description": "Do not perform indel calling.",
                        "id": "#no_indel_calling",
                        "label": "Do not perform indel calling",
                        "sbg:category": "SNP/INDEL genotype likelihoods options",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "description": "Comma-separated list of FORMAT and INFO tags (DP,AD,ADF,ADR,SP,INFO/AD,INFO/ADF,INFO/ADR []\") to output.",
                        "id": "#more_info_to_output",
                        "label": "Comma-separated list of FORMAT and INFO tags to output",
                        "sbg:category": "configuration",
                        "type": [
                            "null",
                            "string"
                        ]
                    },
                    {
                        "label": "Minimum fraction of gapped reads for candidates",
                        "sbg:toolDefaultValue": "0.002",
                        "description": "Minimum fraction of gapped reads for candidates.",
                        "id": "#minimum_fraction_of_gapped_reads",
                        "sbg:category": "SNP/INDEL genotype likelihoods options",
                        "type": [
                            "null",
                            "float"
                        ]
                    },
                    {
                        "label": "Minimum gapped reads for indel candidates",
                        "sbg:toolDefaultValue": "1",
                        "description": "Minimum gapped reads for indel candidates.",
                        "id": "#min_gapped_reads_for_indel",
                        "sbg:category": "SNP/INDEL genotype likelihoods options",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "label": "Max per-BAM depth",
                        "sbg:toolDefaultValue": "250",
                        "description": "At a position, read maximally INT reads per input BAM.",
                        "id": "#max_per_bam_depth",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "label": "Skip INDEL calling if the average per-sample depth is above",
                        "type": [
                            "null",
                            "int"
                        ],
                        "sbg:toolDefaultValue": "250",
                        "schema": [
                            "null",
                            "int"
                        ],
                        "id": "#max_idepth",
                        "sbg:category": "configuration",
                        "description": "Skip INDEL calling if the average per-sample depth is above."
                    },
                    {
                        "description": "Minimum mapping quality for an alignment to be used.",
                        "id": "#mapq_threshold",
                        "label": "Skip alignments with mapQ  smaller than",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "description": "Ignore rg tags.",
                        "id": "#ignore_rg_tags",
                        "label": "Ignore RG tags",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "description": "Disable read-pair overlap detection.",
                        "id": "#ignore_overlaps",
                        "label": "Disable read-pair overlap detection",
                        "sbg:category": "configuration",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "label": "Coefficient for homopolymer errors",
                        "sbg:toolDefaultValue": "100",
                        "description": "Coefficient for modeling homopolymer errors. Given an l-long homopolymer run, the sequencing error of an indel of size s is modeled as INT*s/l.",
                        "id": "#homopolymer_err_coeficient",
                        "sbg:category": "SNP/INDEL genotype likelihoods options",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "label": "Phred-scaled gap open sequencing error probability",
                        "sbg:toolDefaultValue": "40",
                        "description": "Phred-scaled gap open sequencing error probability. Reducing INT leads to more indel calls.",
                        "id": "#gap_open_sequencing_error_probability",
                        "sbg:category": "SNP/INDEL genotype likelihoods options",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "label": "Phred-scaled gap extension seq error probability",
                        "sbg:toolDefaultValue": "20",
                        "description": "Phred-scaled gap extension sequencing error probability. Reducing INT leads to longer indels.",
                        "id": "#gap_extension_seq_error_probability",
                        "sbg:category": "SNP/INDEL genotype likelihoods options",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "sbg:stageInput": null,
                        "label": "Filter zero coverage lines.",
                        "description": "Filter zero coverage lines from pileup files. This option is valid only when output pileup files.",
                        "id": "#filter_zero_coverage_lines",
                        "sbg:category": "output configuration",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "sbg:stageInput": null,
                        "label": "Filter flags",
                        "description": "Filter flags: skip reads with mask bits set [UNMAP,SECONDARY,QCFAIL,DUP].",
                        "id": "#filter_flags",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "string"
                        ]
                    },
                    {
                        "description": "Disable probabilistic realignment for the computation of base alignment quality (BAQ). BAQ is the Phred-scaled probability of a read base being misaligned. Applying this option greatly helps to reduce false SNPs caused by misalignments.",
                        "id": "#disable_baq_computation",
                        "label": "Disable BAQ computation",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "description": "Do not skip anomalous read pairs in variant calling.",
                        "id": "#count_anomalous_read_pairs",
                        "label": "Count anomalous read pairs",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "label": "Comma separated list of platforms for indels",
                        "sbg:toolDefaultValue": "\"all\"",
                        "description": "Comma dilimited list of platforms (determined by @RG-PL) from which indel candidates are obtained. It is recommended to collect indel candidates from sequencing technologies that have low indel error rate such as ILLUMINA.",
                        "id": "#comma_separated_list_of_platforms_for_indels",
                        "sbg:category": "SNP/INDEL genotype likelihoods options",
                        "type": [
                            "null",
                            "string"
                        ]
                    },
                    {
                        "description": "Assume the quality is in the Illumina-1.3+ encoding.",
                        "id": "#assume_the_quality_Illumina_encoding",
                        "label": "Assume the quality is in the Illumina-1.3+ encoding",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "description": "Mode to tell how to merge mpileup files or to pass them through.",
                        "label": "Output state",
                        "id": "#output_state",
                        "type": [
                            {
                                "name": "output_state",
                                "symbols": [
                                    "Merge",
                                    "Pass nonempty files",
                                    "Pass all files"
                                ],
                                "type": "enum"
                            }
                        ]
                    }
                ],
                "hints": [
                    {
                        "value": "cc2.8xlarge",
                        "class": "sbg:forceAmazonInstance"
                    },
                    {
                        "value": "n1-standard-32",
                        "class": "sbg:forceGoogleInstance"
                    }
                ],
                "sbg:revision": 22,
                "sbg:modifiedOn": 1482421868,
                "id": "https://api.sbgenomics.com/v2/apps/marouf/samtools-mpileup-parallel-demo/samtools-mpileup-parallel/22/raw/",
                "sbg:latestRevision": 22,
                "steps": [
                    {
                        "scatter": "#SAMtools_Mpileup.chr_pos_list_file",
                        "inputs": [
                            {
                                "id": "#SAMtools_Mpileup.skip_bases_with_baq_smaller_than_defined",
                                "source": [
                                    "#skip_bases_with_baq_smaller_than_defined"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.required_flags",
                                "source": [
                                    "#required_flags"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.region_pileup_generation",
                                "source": [
                                    "#region_pileup_generation"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.reference_fasta_file",
                                "source": [
                                    "#reference_fasta"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.recalc_BAQ_fly",
                                "source": [
                                    "#recalc_BAQ_fly"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.per_sample_mF",
                                "source": [
                                    "#per_sample_mF"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.parameter_for_adjusting_mapq",
                                "source": [
                                    "#parameter_for_adjusting_mapq"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.output_uncompressed_bcf_or_vcf",
                                "source": [
                                    "#output_uncompressed_bcf_or_vcf"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.output_mapping_quality",
                                "source": [
                                    "#output_mapping_quality"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.output_format",
                                "source": [
                                    "#output_format"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.output_base_positions_on_reads",
                                "source": [
                                    "#output_base_positions_on_reads"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.no_indel_calling",
                                "source": [
                                    "#no_indel_calling"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.more_info_to_output",
                                "source": [
                                    "#more_info_to_output"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.minimum_fraction_of_gapped_reads",
                                "source": [
                                    "#minimum_fraction_of_gapped_reads"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.min_gapped_reads_for_indel",
                                "source": [
                                    "#min_gapped_reads_for_indel"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.max_per_bam_depth",
                                "source": [
                                    "#max_per_bam_depth"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.max_idepth",
                                "source": [
                                    "#max_idepth"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.mapq_threshold",
                                "source": [
                                    "#mapq_threshold"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.input_bam_files",
                                "source": [
                                    "#bams"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.ignore_rg_tags",
                                "source": [
                                    "#ignore_rg_tags"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.ignore_overlaps",
                                "source": [
                                    "#ignore_overlaps"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.homopolymer_err_coeficient",
                                "source": [
                                    "#homopolymer_err_coeficient"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.gap_open_sequencing_error_probability",
                                "source": [
                                    "#gap_open_sequencing_error_probability"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.gap_extension_seq_error_probability",
                                "source": [
                                    "#gap_extension_seq_error_probability"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.filter_zero_coverage_lines",
                                "source": [
                                    "#filter_zero_coverage_lines"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.filter_flags",
                                "source": [
                                    "#filter_flags"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.exclude_read_groups_list_file"
                            },
                            {
                                "id": "#SAMtools_Mpileup.disable_baq_computation",
                                "source": [
                                    "#disable_baq_computation"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.count_anomalous_read_pairs",
                                "source": [
                                    "#count_anomalous_read_pairs"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.comma_separated_list_of_platforms_for_indels",
                                "source": [
                                    "#comma_separated_list_of_platforms_for_indels"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.chr_pos_list_file",
                                "source": [
                                    "#SBG_Prepare_Intervals.intervals"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.assume_the_quality_Illumina_encoding",
                                "source": [
                                    "#assume_the_quality_Illumina_encoding"
                                ]
                            }
                        ],
                        "outputs": [
                            {
                                "id": "#SAMtools_Mpileup.output_pileup_vcf_or_bcf_file"
                            }
                        ],
                        "sbg:x": 924,
                        "run": {
                            "sbg:toolkit": "SAMtools",
                            "requirements": [
                                {
                                    "requirements": [
                                        {
                                            "class": "DockerRequirement",
                                            "dockerPull": "rabix/js-engine"
                                        }
                                    ],
                                    "id": "#cwl-js-engine",
                                    "class": "ExpressionEngineRequirement"
                                }
                            ],
                            "sbg:validationErrors": [],
                            "sbg:categories": [
                                "SAM/BAM-Processing"
                            ],
                            "label": "SAMtools Mpileup",
                            "sbg:createdBy": "marouf",
                            "sbg:cmdPreview": "/opt/samtools-1.3/samtools mpileup    input_bam_file1.bam  input_bam_file2.bam   | awk '{if($4 != 0) print $0}'  > input_bam_file1.pileup",
                            "sbg:image_url": null,
                            "cwlVersion": "sbg:draft-2",
                            "sbg:revisionsInfo": [
                                {
                                    "sbg:revision": 0,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1458812970
                                },
                                {
                                    "sbg:revision": 1,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1458825811
                                },
                                {
                                    "sbg:revision": 2,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1462805320
                                },
                                {
                                    "sbg:revision": 3,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1468578873
                                },
                                {
                                    "sbg:revision": 4,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1468580666
                                },
                                {
                                    "sbg:revision": 5,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1468586077
                                },
                                {
                                    "sbg:revision": 6,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1468924014
                                },
                                {
                                    "sbg:revision": 7,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1473160925
                                },
                                {
                                    "sbg:revision": 8,
                                    "sbg:revisionNotes": "Peer review",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1473433741
                                },
                                {
                                    "sbg:revision": 9,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1473434123
                                },
                                {
                                    "sbg:revision": 10,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1473771444
                                },
                                {
                                    "sbg:revision": 11,
                                    "sbg:revisionNotes": "change positional argument for BCF output",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1473779051
                                },
                                {
                                    "sbg:revision": 12,
                                    "sbg:revisionNotes": "Delete default value inserted mistakenly for some parameters.",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1473849264
                                },
                                {
                                    "sbg:revision": 13,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1473931426
                                },
                                {
                                    "sbg:revision": 14,
                                    "sbg:revisionNotes": "add new common issue",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1474901461
                                },
                                {
                                    "sbg:revision": 15,
                                    "sbg:revisionNotes": "Add filter to pileup files",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1478526836
                                },
                                {
                                    "sbg:revision": 16,
                                    "sbg:revisionNotes": "change the test case",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1478526922
                                },
                                {
                                    "sbg:revision": 17,
                                    "sbg:revisionNotes": "update description",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1480075403
                                },
                                {
                                    "sbg:revision": 18,
                                    "sbg:revisionNotes": "output is not required field anymore",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1480440302
                                },
                                {
                                    "sbg:revision": 19,
                                    "sbg:revisionNotes": "Added inherit metadata from input.",
                                    "sbg:modifiedBy": "milan.domazet.sudo",
                                    "sbg:modifiedOn": 1482334695
                                }
                            ],
                            "y": 361,
                            "sbg:toolkitVersion": "v1.3",
                            "sbg:license": "The MIT License",
                            "sbg:modifiedBy": "milan.domazet.sudo",
                            "class": "CommandLineTool",
                            "inputs": [
                                {
                                    "label": "Skip bases with baseQ/BAQ smaller than",
                                    "inputBinding": {
                                        "prefix": "-Q",
                                        "sbg:cmdInclude": true,
                                        "separate": true,
                                        "separator": " ",
                                        "position": 0
                                    },
                                    "sbg:toolDefaultValue": "13",
                                    "description": "Minimum base quality for a base to be considered.",
                                    "id": "#skip_bases_with_baq_smaller_than_defined",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "label": "Required flags",
                                    "inputBinding": {
                                        "prefix": "--rf",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Required flags: skip reads with mask bits unset.",
                                    "id": "#required_flags",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "label": "Region in which pileup is generated",
                                    "inputBinding": {
                                        "prefix": "-r",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Region in which pileup is generated.",
                                    "id": "#region_pileup_generation",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "string"
                                    ]
                                },
                                {
                                    "sbg:fileTypes": "FASTA,FA,GZ",
                                    "label": "Reference FASTA file",
                                    "type": [
                                        "null",
                                        "File"
                                    ],
                                    "required": false,
                                    "inputBinding": {
                                        "prefix": "-f",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "secondaryFiles": [
                                            ".fai"
                                        ],
                                        "separate": true
                                    },
                                    "description": "Faidx indexed reference sequence (FASTA) file.",
                                    "id": "#reference_fasta_file",
                                    "sbg:category": "File input",
                                    "schema": [
                                        "null",
                                        "File"
                                    ]
                                },
                                {
                                    "label": "Recalculate BAQ on the fly",
                                    "inputBinding": {
                                        "prefix": "-E",
                                        "sbg:cmdInclude": true,
                                        "separate": true,
                                        "separator": " ",
                                        "position": 0
                                    },
                                    "description": "Recalculate BAQ on the fly, ignore existing BQ tags.",
                                    "id": "#recalc_BAQ_fly",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "sbg:stageInput": null,
                                    "type": [
                                        "null",
                                        "boolean"
                                    ],
                                    "inputBinding": {
                                        "prefix": "-p",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Apply -m and -F thresholds per sample to increase sensitivity of calling. By default both options are applied to reads pooled from all samples.",
                                    "id": "#per_sample_mF",
                                    "sbg:category": "configuration",
                                    "label": "Apply -m and -F thresholds per sample"
                                },
                                {
                                    "label": "Parameter for adjusting mapQ",
                                    "inputBinding": {
                                        "prefix": "-C",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Coefficient for downgrading mapping quality for reads containing excessive mismatches. Given a read with a phred-scaled probability q of being generated from the mapped position, the new mapping quality is about sqrt((INT-q)/INT)*INT. A zero value disables this functionality; if enabled, the recommended value for BWA is 50.",
                                    "id": "#parameter_for_adjusting_mapq",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "label": "Generate uncompress BCF/VCF output",
                                    "inputBinding": {
                                        "prefix": "-u",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Similar to -g except that the output is uncompressed BCF, which is preferred for piping.",
                                    "id": "#output_uncompressed_bcf_or_vcf",
                                    "sbg:category": "Output options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "label": "Output mapping quality",
                                    "inputBinding": {
                                        "prefix": "-s",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Output mapping quality (disabled by -g/-u).",
                                    "id": "#output_mapping_quality",
                                    "sbg:category": "Output options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "description": "Output file format.",
                                    "id": "#output_format",
                                    "label": "Output file format",
                                    "sbg:category": "configuration",
                                    "type": [
                                        {
                                            "name": "output_format",
                                            "symbols": [
                                                "PILEUP",
                                                "BCF",
                                                "VCF"
                                            ],
                                            "type": "enum"
                                        }
                                    ]
                                },
                                {
                                    "label": "Output base positions on reads",
                                    "inputBinding": {
                                        "prefix": "-O",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Output base positions on reads (disabled by -g/-u).",
                                    "id": "#output_base_positions_on_reads",
                                    "sbg:category": "Output options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "label": "Do not perform indel calling",
                                    "inputBinding": {
                                        "prefix": "-I",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Do not perform indel calling.",
                                    "id": "#no_indel_calling",
                                    "sbg:category": "SNP/INDEL genotype likelihoods options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "label": "Comma-separated list of FORMAT and INFO tags to output",
                                    "inputBinding": {
                                        "prefix": "-t",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "itemSeparator": null,
                                        "separate": true
                                    },
                                    "description": "Comma-separated list of FORMAT and INFO tags (DP,AD,ADF,ADR,SP,INFO/AD,INFO/ADF,INFO/ADR []\") to output.",
                                    "id": "#more_info_to_output",
                                    "sbg:category": "configuration",
                                    "type": [
                                        "null",
                                        "string"
                                    ]
                                },
                                {
                                    "label": "Minimum fraction of gapped reads for candidates",
                                    "inputBinding": {
                                        "prefix": "-F",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "sbg:toolDefaultValue": "0.002",
                                    "description": "Minimum fraction of gapped reads for candidates.",
                                    "id": "#minimum_fraction_of_gapped_reads",
                                    "sbg:category": "SNP/INDEL genotype likelihoods options",
                                    "type": [
                                        "null",
                                        "float"
                                    ]
                                },
                                {
                                    "label": "Minimum gapped reads for indel candidates",
                                    "inputBinding": {
                                        "prefix": "-m",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "sbg:toolDefaultValue": "1",
                                    "description": "Minimum gapped reads for indel candidates.",
                                    "id": "#min_gapped_reads_for_indel",
                                    "sbg:category": "SNP/INDEL genotype likelihoods options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "label": "Max per-BAM depth",
                                    "inputBinding": {
                                        "prefix": "-d",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "sbg:toolDefaultValue": "250",
                                    "description": "At a position, read maximally INT reads per input BAM.",
                                    "id": "#max_per_bam_depth",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "label": "Skip INDEL calling if the average per-sample depth is above",
                                    "type": [
                                        "null",
                                        "int"
                                    ],
                                    "inputBinding": {
                                        "prefix": "-L",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "sbg:toolDefaultValue": "250",
                                    "schema": [
                                        "null",
                                        "int"
                                    ],
                                    "id": "#max_idepth",
                                    "sbg:category": "configuration",
                                    "description": "Skip INDEL calling if the average per-sample depth is above."
                                },
                                {
                                    "label": "Skip alignments with mapQ  smaller than",
                                    "inputBinding": {
                                        "prefix": "-q",
                                        "sbg:cmdInclude": true,
                                        "separate": true,
                                        "separator": " ",
                                        "position": 0
                                    },
                                    "description": "Minimum mapping quality for an alignment to be used.",
                                    "id": "#mapq_threshold",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "sbg:fileTypes": "BAM",
                                    "sbg:stageInput": "link",
                                    "type": [
                                        {
                                            "name": "input_bam_files",
                                            "type": "array",
                                            "items": "File"
                                        }
                                    ],
                                    "required": true,
                                    "label": "List of input BAM files",
                                    "inputBinding": {
                                        "separator": " ",
                                        "position": 1,
                                        "sbg:cmdInclude": true,
                                        "streamable": false,
                                        "itemSeparator": null,
                                        "separate": true
                                    },
                                    "description": "List of input BAM files.",
                                    "id": "#input_bam_files",
                                    "sbg:category": "Input options",
                                    "schema": [
                                        "File"
                                    ]
                                },
                                {
                                    "label": "Ignore RG tags",
                                    "inputBinding": {
                                        "prefix": "-R",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Ignore rg tags.",
                                    "id": "#ignore_rg_tags",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "label": "Disable read-pair overlap detection",
                                    "inputBinding": {
                                        "prefix": "-x",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Disable read-pair overlap detection.",
                                    "id": "#ignore_overlaps",
                                    "sbg:category": "configuration",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "label": "Coefficient for homopolymer errors",
                                    "inputBinding": {
                                        "prefix": "-h",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "sbg:toolDefaultValue": "100",
                                    "description": "Coefficient for modeling homopolymer errors. Given an l-long homopolymer run, the sequencing error of an indel of size s is modeled as INT*s/l.",
                                    "id": "#homopolymer_err_coeficient",
                                    "sbg:category": "SNP/INDEL genotype likelihoods options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "label": "Phred-scaled gap open sequencing error probability",
                                    "inputBinding": {
                                        "prefix": "-o",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "sbg:toolDefaultValue": "40",
                                    "description": "Phred-scaled gap open sequencing error probability. Reducing INT leads to more indel calls.",
                                    "id": "#gap_open_sequencing_error_probability",
                                    "sbg:category": "SNP/INDEL genotype likelihoods options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "label": "Phred-scaled gap extension seq error probability",
                                    "inputBinding": {
                                        "prefix": "-e",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "sbg:toolDefaultValue": "20",
                                    "description": "Phred-scaled gap extension sequencing error probability. Reducing INT leads to longer indels.",
                                    "id": "#gap_extension_seq_error_probability",
                                    "sbg:category": "SNP/INDEL genotype likelihoods options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "sbg:stageInput": null,
                                    "label": "Filter zero coverage lines.",
                                    "description": "Filter zero coverage lines from pileup files. This option is valid only when output pileup files.",
                                    "id": "#filter_zero_coverage_lines",
                                    "sbg:category": "output configuration",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "sbg:stageInput": null,
                                    "type": [
                                        "null",
                                        "string"
                                    ],
                                    "inputBinding": {
                                        "prefix": "--ff",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Filter flags: skip reads with mask bits set [UNMAP,SECONDARY,QCFAIL,DUP].",
                                    "id": "#filter_flags",
                                    "sbg:category": "Input options",
                                    "label": "Filter flags"
                                },
                                {
                                    "sbg:fileTypes": "TXT",
                                    "label": "Exclude read groups listed in file",
                                    "required": false,
                                    "inputBinding": {
                                        "prefix": "-G",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Exclude read groups listed in file.",
                                    "id": "#exclude_read_groups_list_file",
                                    "sbg:category": "File input",
                                    "type": [
                                        "null",
                                        "File"
                                    ]
                                },
                                {
                                    "label": "Disable BAQ computation",
                                    "inputBinding": {
                                        "prefix": "-B",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Disable probabilistic realignment for the computation of base alignment quality (BAQ). BAQ is the Phred-scaled probability of a read base being misaligned. Applying this option greatly helps to reduce false SNPs caused by misalignments.",
                                    "id": "#disable_baq_computation",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "label": "Count anomalous read pairs",
                                    "inputBinding": {
                                        "prefix": "-A",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Do not skip anomalous read pairs in variant calling.",
                                    "id": "#count_anomalous_read_pairs",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "label": "Comma separated list of platforms for indels",
                                    "inputBinding": {
                                        "prefix": "-P",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "sbg:toolDefaultValue": "\"all\"",
                                    "description": "Comma dilimited list of platforms (determined by @RG-PL) from which indel candidates are obtained. It is recommended to collect indel candidates from sequencing technologies that have low indel error rate such as ILLUMINA.",
                                    "id": "#comma_separated_list_of_platforms_for_indels",
                                    "sbg:category": "SNP/INDEL genotype likelihoods options",
                                    "type": [
                                        "null",
                                        "string"
                                    ]
                                },
                                {
                                    "sbg:fileTypes": "BED",
                                    "label": "List of positions (chr pos) or regions in BED file",
                                    "required": false,
                                    "inputBinding": {
                                        "prefix": "-l",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "BED or position list file containing a list of regions or sites where pileup or BCF should be generated.",
                                    "id": "#chr_pos_list_file",
                                    "sbg:category": "File input",
                                    "type": [
                                        "null",
                                        "File"
                                    ]
                                },
                                {
                                    "label": "Assume the quality is in the Illumina-1.3+ encoding",
                                    "inputBinding": {
                                        "prefix": "-6",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Assume the quality is in the Illumina-1.3+ encoding.",
                                    "id": "#assume_the_quality_Illumina_encoding",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                }
                            ],
                            "hints": [
                                {
                                    "dockerImageId": "2fb927277493",
                                    "class": "DockerRequirement",
                                    "dockerPull": "images.sbgenomics.com/marouf/samtools:1.3"
                                },
                                {
                                    "value": 1,
                                    "class": "sbg:CPURequirement"
                                },
                                {
                                    "value": 1000,
                                    "class": "sbg:MemRequirement"
                                }
                            ],
                            "sbg:revision": 19,
                            "baseCommand": [
                                "/opt/samtools-1.3/samtools",
                                "mpileup"
                            ],
                            "id": "marouf/samtools-1-3-demo/mpileup-1-3/19",
                            "sbg:latestRevision": 19,
                            "temporaryFailCodes": [],
                            "successCodes": [],
                            "sbg:toolAuthor": "Heng Li, Sanger Institute",
                            "sbg:revisionNotes": "Added inherit metadata from input.",
                            "arguments": [
                                {
                                    "separate": true,
                                    "valueFrom": {
                                        "script": "{\n\n if($job.inputs.output_format)\n {\n    if($job.inputs.output_format == \"PILEUP\")\n        return\n    else if($job.inputs.output_format == \"VCF\")\n    \treturn \"-v\"\n    else if($job.inputs.output_format == \"BCF\")\n    \treturn \"-g\"\n    else\n      return\n  \n \n }\n\n\n\n}",
                                        "engine": "#cwl-js-engine",
                                        "class": "Expression"
                                    }
                                },
                                {
                                    "position": 1000,
                                    "valueFrom": {
                                        "script": "{\n\n // Check if input files are delivered in an array or not\n if($job.inputs.input_bam_files.constructor == Array)\n {\n   // Select the first entry of the array\n\u2002\u2002\u2002\u2002filepath = \"/\"+$job.inputs.input_bam_files[0].path \n\u2002\u2002} \n  \n  else\n    \n  {\n    // Use the file name directly\n\u2002\u2002\u2002\u2002filepath = \"/\"+$job.inputs.input_bam_files.path\n \u2002}\n   \n \n filename = filepath.split(\"/\").pop()\n new_filename = filename.substr(0,filename.lastIndexOf(\".\"))\n    \n \n if ($job.inputs.chr_pos_list_file)\n   new_filename = new_filename+'_'+ $job.inputs.chr_pos_list_file.path.split('/')[$job.inputs.chr_pos_list_file.path.split('/').length-1].split('.')[0]\n\n \n extension = '.pileup'\n  \n if ( (typeof $job.inputs.output_format !== \"undefined\" && $job.inputs.output_format === \"BCF\") )\n   extension = '.bcf'\n             \n else if ((typeof $job.inputs.output_format !== \"undefined\" && $job.inputs.output_format === \"VCF\") )  \n   extension = '.vcf' \n  \n if($job.inputs.filter_zero_coverage_lines && extension== '.pileup')   \n \treturn  \" | awk '{if($4 != 0) print $0}'  > \"+  new_filename + extension\n else\n   return  \" > \"+  new_filename + extension\n}",
                                        "engine": "#cwl-js-engine",
                                        "class": "Expression"
                                    },
                                    "separate": true
                                }
                            ],
                            "sbg:id": "admin/sbg-public-data/mpileup-1-3/19",
                            "sbg:job": {
                                "allocatedResources": {
                                    "cpu": 1,
                                    "mem": 1000
                                },
                                "inputs": {
                                    "filter_zero_coverage_lines": true,
                                    "output_uncompressed_bcf_or_vcf": true,
                                    "region_pileup_generation": "region_pileup_generation",
                                    "max_idepth": 11,
                                    "recalc_BAQ_fly": true,
                                    "no_indel_calling": true,
                                    "per_sample_mF": true,
                                    "homopolymer_err_coeficient": 11,
                                    "required_flags": 0,
                                    "ignore_rg_tags": true,
                                    "minimum_fraction_of_gapped_reads": 11,
                                    "reference_fasta_file": {
                                        "path": "input/input.fasta",
                                        "secondaryFiles": [
                                            {
                                                "path": ".fai"
                                            }
                                        ]
                                    },
                                    "exclude_read_groups_list_file": {
                                        "path": "exclude_read_groups_list_file.txt",
                                        "size": 0,
                                        "class": "File",
                                        "secondaryFiles": []
                                    },
                                    "output_mapping_quality": true,
                                    "assume_the_quality_Illumina_encoding": true,
                                    "parameter_for_adjusting_mapq": 11,
                                    "min_gapped_reads_for_indel": 11,
                                    "gap_open_sequencing_error_probability": 11,
                                    "input_bam_files": [
                                        {
                                            "path": "input_bam_file1.bam",
                                            "size": 0,
                                            "class": "File",
                                            "secondaryFiles": []
                                        },
                                        {
                                            "path": "input_bam_file2.bam"
                                        }
                                    ],
                                    "output_format": "PILEUP",
                                    "ignore_overlaps": true,
                                    "disable_baq_computation": true,
                                    "count_anomalous_read_pairs": true,
                                    "gap_extension_seq_error_probability": 11,
                                    "more_info_to_output": "more_info_to_output-string-value",
                                    "max_per_bam_depth": 11,
                                    "skip_bases_with_baq_smaller_than_defined": 11,
                                    "comma_separated_list_of_platforms_for_indels": "all",
                                    "filter_flags": "filter_flags-string-value",
                                    "mapq_threshold": 11,
                                    "output_base_positions_on_reads": true
                                }
                            },
                            "stdin": "",
                            "sbg:project": "marouf/samtools-1-3-demo",
                            "description": "SAMtools Mpileup generates BCF or PILEUP for one or multiple BAM files. Alignment records are grouped by sample identifiers in @RG header lines. If sample identifiers are absent, each input file is regarded as one sample.\n\nIn the pileup format (without -uor-g), each line represents a genomic position consisting of chromosome name, coordinate, reference base, read bases, read qualities, and alignment mapping qualities. Information on match, mismatch, indel, strand, mapping quality, and the start and end of a read are all encoded at the read base column. In this column, a dot stands for a match to the reference base on the forward strand, a comma for a match on the reverse strand, a '>' or '<' for a reference skip, 'ACGTN' for a mismatch on the forward strand, and 'acgtn' for a mismatch on the reverse strand. A pattern `\\\\+[0-9]+[ACGTNacgtn]+' indicates there is an insertion between this reference position and the next reference position. The length of the insertion is given by the integer in the pattern followed by the inserted sequence. Similarly, a pattern '-[0-9]+[ACGTNacgtn]+' represents a deletion from the reference. The deleted bases will be presented as `*' in the following lines. Also at the read base column, a symbol '^', marks the start of a read. The ASCII of the character following '^' minus 33 gives the mapping quality. The symbol '$' marks the end of a read segment.\n\n#### Common Issue:\n- Please use the public pileup parallel pipeline for large input files.\n- Please sort BAM files by coordinates before using mpileup.",
                            "x": 924,
                            "sbg:createdOn": 1458812970,
                            "outputs": [
                                {
                                    "type": [
                                        "null",
                                        "File"
                                    ],
                                    "label": "Output PILEUP, VCF, or BCF file",
                                    "outputBinding": {
                                        "glob": "{*.pileup,*.bcf,*.vcf}",
                                        "sbg:metadata": {
                                            "ScatteredUsing": {
                                                "script": "{\n\tif ($job.inputs.chr_pos_list_file)\n    {\n\t    len =  $job.inputs.chr_pos_list_file.path.split('/').length\n        \n        return $job.inputs.chr_pos_list_file.path.split('/')[len-1].split('.')[0]\n    }\n  \telse\n    {\n\t    return \n    }\n}",
                                                "engine": "#cwl-js-engine",
                                                "class": "Expression"
                                            }
                                        },
                                        "streamable": false,
                                        "sbg:inheritMetadataFrom": "#input_bam_files"
                                    },
                                    "sbg:fileTypes": "PILEUP,BCF,VCF",
                                    "id": "#output_pileup_vcf_or_bcf_file",
                                    "description": "Output PILEUP, VCF, or BCF file."
                                }
                            ],
                            "sbg:links": [
                                {
                                    "label": "Homepage",
                                    "id": "http://samtools.sourceforge.net/"
                                },
                                {
                                    "label": "Source code",
                                    "id": "https://github.com/samtools/samtools"
                                },
                                {
                                    "label": "Wiki",
                                    "id": "http://sourceforge.net/p/samtools/wiki/Home/"
                                },
                                {
                                    "label": "Download",
                                    "id": "http://sourceforge.net/projects/samtools/files/"
                                },
                                {
                                    "label": "Publication",
                                    "id": "http://www.ncbi.nlm.nih.gov/pubmed/19505943"
                                },
                                {
                                    "label": "Documentation",
                                    "id": "http://www.htslib.org/doc/samtools.html"
                                }
                            ],
                            "sbg:contributors": [
                                "milan.domazet.sudo",
                                "marouf"
                            ],
                            "sbg:modifiedOn": 1482334695,
                            "sbg:sbgMaintained": false,
                            "stdout": ""
                        },
                        "id": "#SAMtools_Mpileup",
                        "sbg:y": 361
                    },
                    {
                        "sbg:x": 782,
                        "outputs": [
                            {
                                "id": "#SBG_Prepare_Intervals.names"
                            },
                            {
                                "id": "#SBG_Prepare_Intervals.intervals"
                            }
                        ],
                        "run": {
                            "sbg:toolkit": "SBGTools",
                            "requirements": [
                                {
                                    "fileDef": [
                                        {
                                            "filename": "sbg_prepare_intervals.py",
                                            "fileContent": "\"\"\"\nUsage:\n    sbg_prepare_intervals.py [options] [--fastq FILE --bed FILE --mode INT --format STR --others STR]\n\nDescription:\n    Purpose of this tool is to split BED file into files based on the selected mode.\n    If bed file is not provided fai(fasta index) file is converted to bed.\n\nOptions:\n\n    -h, --help            Show this message.\n\n    -v, -V, --version     Tool version.\n\n    -b, -B, --bed FILE    Path to input bed file.\n\n    --fai FILE            Path to input fai file.\n\n    --format STR          Output file format.\n\n    --mode INT            Select input mode.\n\n\"\"\"\n\n\nfrom docopt import docopt\nimport os\nimport shutil\nimport glob\n\ndefault_extension = '.bed'  # for output files\n\n\n\ndef create_file(contents, contig_name, extension=default_extension):\n    \"\"\"function for creating a file for all intervals in a contig\"\"\"\n\n    new_file = open(\"Intervals/\" + contig_name + extension, \"w\")\n    new_file.write(contents)\n    new_file.close()\n\n\ndef add_to_file(line, name, extension=default_extension):\n    \"\"\"function for adding a line to a file\"\"\"\n\n    new_file = open(\"Intervals/\" + name + extension, \"a\")\n    if lformat == formats[1]:\n        sep = line.split(\"\\t\")\n        line = sep[0] + \":\" + sep[1] + \"-\" + sep[2]\n    new_file.write(line)\n    new_file.close()\n\n\ndef fai2bed(fai):\n    \"\"\"function to create a bed file from fai file\"\"\"\n\n    region_thr = 10000000  # threshold used to determine starting point accounting for telomeres in chromosomes\n    if not fai.rfind(\".fasta.fai\") == -1:\n        basename = fai[0:fai.rfind(\".fasta.fai\")]\n    else:\n        basename = fai[0:fai.rfind(\".\")]\n    with open(fai, \"r\") as ins:\n        new_array = []\n        for line in ins:\n            len_reg = int(line.split()[1])\n            cutoff = 0 if (len_reg < region_thr) else 0  # sd\\\\telomeres or start with 1\n            new_line = line.split()[0] + '\\t' + str(cutoff) + '\\t' + str(len_reg + cutoff)\n            new_array.append(new_line)\n    new_file = open(basename + \".bed\", \"w\")\n    new_file.write(\"\\n\".join(new_array))\n    return basename + \".bed\"\n\ndef chr_intervals(no_of_chrms = 23):\n    \"\"\"returns all possible designations for chromosome intervals\"\"\"\n    \n    chrms = []\n    for i in range(1, no_of_chrms):\n        chrms.append(\"chr\" + str(i))\n        chrms.append(str(i))\n    chrms.extend([\"x\", \"y\", \"chrx\", \"chry\"])\n    return chrms\n\n\ndef mode_1(orig_file):\n    \"\"\"mode 1: every line is a new file\"\"\"\n\n    with open(orig_file, \"r\") as ins:\n        prev = \"\"\n        counter = 0\n        names = []\n        for line in ins:\n            if line.split()[0] == prev:\n                counter += 1\n            else:\n                counter = 0\n            suffix = \"\" if (counter == 0) else \"_\" + str(counter)\n            create_file(line, line.split()[0] + suffix)\n            names.append(line.split()[0] + suffix)\n            prev = line.split()[0]\n\n        create_file(str(names), \"names\", extension=\".txt\")\n\ndef mode_2(orig_file, others_name):\n    \"\"\"mode 2: separate file is created for each chromosome, and one file is created for other intervals\"\"\"\n\n    chrms = chr_intervals()\n    names = []\n\n    with open(orig_file, 'r') as ins:\n        for line in ins:\n            name = line.split()[0]\n            if name.lower() in chrms:\n                name = name.lower()\n            else:\n                name = others_name\n            try:\n                add_to_file(line, name)\n                if not name in names:\n                    names.append(name)\n            except:\n                raise Exception(\"Couldn't create or write in the file in mode 2\")\n\n        create_file(str(names), \"names\", extension = \".txt\")\n\n\ndef mode_3(orig_file, extension=default_extension):\n    \"\"\"mode 3: input file is staged to output\"\"\"\n\n    orig_name = orig_file.split(\"/\")[len(orig_file.split(\"/\")) - 1]\n    output_file = r\"./Intervals/\" + orig_name[0:orig_name.rfind('.')] + extension\n\n    shutil.copyfile(orig_file, output_file)\n\n    names = [orig_name[0:orig_name.rfind('.')]]\n    create_file(str(names), \"names\", extension=\".txt\")\n\n\ndef mode_4(orig_file, others_name):\n    \"\"\"mode 4: every interval in chromosomes is in a separate file. Other intervals are in a single file\"\"\"\n\n    chrms = chr_intervals()\n    names = []\n\n    with open(orig_file, \"r\") as ins:\n        counter = {}\n        for line in ins:\n            name = line.split()[0].lower()\n            if name in chrms:\n                if name in counter:\n                    counter[name] += 1\n                else:\n                    counter[name] = 0\n                suffix = \"\" if (counter[name] == 0) else \"_\" + str(counter[name])\n                create_file(line, name + suffix)\n                names.append(name + suffix)\n                prev = name\n            else:\n                name = others_name\n                if not name in names:\n                    names.append(name)\n                try:\n                    add_to_file(line, name)\n                except:\n                    raise Exception(\"Couldn't create or write in the file in mode 4\")\n\n        create_file(str(names), \"names\", extension=\".txt\")\n\n\ndef prepare_intervals():\n    # reading input files and split mode from command line\n    args = docopt(__doc__, version='1.0')\n\n    bed_file = args['--bed']\n    fai_file = args['--fai']\n    split_mode = int(args['--mode'])\n\n    \n    # define file name for non-chromosomal contigs\n    others_name = 'others' \n\n    global formats, lformat\n    formats = [\"chr start end\", \"chr:start-end\"]\n    lformat = args['--format']\n    if lformat == None:\n        lformat = formats[0]\n    if not lformat in formats:\n        raise Exception('Unsuported interval format')\n\n    if not os.path.exists(r\"./Intervals\"):\n        os.mkdir(r\"./Intervals\")\n    else:\n        files = glob.glob(r\"./Intervals/*\")\n        for f in files:\n            os.remove(f)\n\n    # create variable input_file taking bed_file as priority\n    if bed_file:\n        input_file = bed_file\n    elif fai_file:\n        input_file = fai2bed(fai_file)\n    else:\n        raise Exception('No input files are provided')\n\n    # calling adequate split mode function\n    if split_mode == 1:\n        mode_1(input_file)\n    elif split_mode == 2:\n        mode_2(input_file, others_name)\n    elif split_mode == 3:\n        if bed_file:\n            mode_3(input_file)\n        else:\n            raise Exception('Bed file is required for mode 3')\n    elif split_mode == 4:\n        mode_4(input_file, others_name)\n    else:\n        raise Exception('Split mode value is not set')\n\n\nif __name__ == '__main__':\n    prepare_intervals()"
                                        }
                                    ],
                                    "class": "CreateFileRequirement"
                                },
                                {
                                    "requirements": [
                                        {
                                            "class": "DockerRequirement",
                                            "dockerPull": "rabix/js-engine"
                                        }
                                    ],
                                    "id": "#cwl-js-engine",
                                    "class": "ExpressionEngineRequirement"
                                }
                            ],
                            "sbg:validationErrors": [],
                            "sbg:categories": [
                                "Converters"
                            ],
                            "label": "SBG Prepare Intervals",
                            "sbg:createdBy": "vladimirk",
                            "temporaryFailCodes": [],
                            "sbg:image_url": null,
                            "cwlVersion": "sbg:draft-2",
                            "sbg:revisionsInfo": [
                                {
                                    "sbg:revision": 0,
                                    "sbg:revisionNotes": "Copy of medjo/sbg-prepare-intervals/sbg-prepare-intervals/75",
                                    "sbg:modifiedBy": "vladimirk",
                                    "sbg:modifiedOn": 1473083821
                                },
                                {
                                    "sbg:revision": 1,
                                    "sbg:revisionNotes": "Copy of medjo/sbg-prepare-intervals/sbg-prepare-intervals/76",
                                    "sbg:modifiedBy": "bogdang",
                                    "sbg:modifiedOn": 1473084447
                                },
                                {
                                    "sbg:revision": 2,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "medjo",
                                    "sbg:modifiedOn": 1473928444
                                },
                                {
                                    "sbg:revision": 3,
                                    "sbg:revisionNotes": "split_mode set to required",
                                    "sbg:modifiedBy": "medjo",
                                    "sbg:modifiedOn": 1474970272
                                },
                                {
                                    "sbg:revision": 4,
                                    "sbg:revisionNotes": "Fixed Toolkit name.",
                                    "sbg:modifiedBy": "bix-demo",
                                    "sbg:modifiedOn": 1478525360
                                }
                            ],
                            "y": 494,
                            "sbg:modifiedOn": 1478525360,
                            "sbg:license": "Apache License 2.0",
                            "sbg:modifiedBy": "bix-demo",
                            "class": "CommandLineTool",
                            "inputs": [
                                {
                                    "label": "Split mode",
                                    "inputBinding": {
                                        "prefix": "--mode",
                                        "sbg:cmdInclude": true,
                                        "position": 3,
                                        "valueFrom": {
                                            "script": "{\n  mode = $job.inputs.split_mode\n  switch (mode) \n  {\n    case \"File per interval\": \n      return 1\n    case \"File per chr with alt contig in a single file\": \n      return 2\n    case \"Output original BED\": \n      return 3\n    case \"File per interval with alt contig in a single file\": \n      return 4  \n  }\n  return 3\n}",
                                            "engine": "#cwl-js-engine",
                                            "class": "Expression"
                                        },
                                        "separate": true
                                    },
                                    "description": "Depending on selected Split Mode value, output files are generated in accordance with description below:  1. Whole Genome - The tool creates one interval file per line of the input BED(FAI) file. Each interval file contains a single line (one of the lines of BED(FAI) input file).  2. Whole Genome with reduced number of jobs - For each contig(chromosome) a single file is created containing all the intervals corresponding to it . All the intervals (lines) other than (chr1, chr2 ... chrY or 1, 2 ... Y) are saved as (\"others.bed\").  3. Whole Exome - BED file is required for execution of this mode. If mode  3 is applied input is passed to the output.  4. Whole Exome Parallel - BED file is required for execution of this mode. For each contig(chromosome) a single file is created containing all the intervals corresponding to it. All the intervals (lines) other than (chr1, chr2 ... chrY or 1, 2 ... Y) are saved as (\"others.bed\"). If split mode is not selected, the tool will output original BED file.",
                                    "id": "#split_mode",
                                    "sbg:category": "Input",
                                    "type": [
                                        {
                                            "name": "split_mode",
                                            "symbols": [
                                                "File per interval",
                                                "File per chr with alt contig in a single file",
                                                "Output original BED",
                                                "File per interval with alt contig in a single file"
                                            ],
                                            "type": "enum"
                                        }
                                    ]
                                },
                                {
                                    "description": "Format of the intervals in the generated files.",
                                    "id": "#format",
                                    "label": "Interval format",
                                    "sbg:category": "Input",
                                    "type": [
                                        "null",
                                        {
                                            "name": "format",
                                            "symbols": [
                                                "chr start end",
                                                "chr:start-end"
                                            ],
                                            "type": "enum"
                                        }
                                    ]
                                },
                                {
                                    "sbg:fileTypes": "FAI",
                                    "label": "Input FAI file",
                                    "required": false,
                                    "inputBinding": {
                                        "prefix": "--fai",
                                        "sbg:cmdInclude": true,
                                        "position": 2,
                                        "separate": true
                                    },
                                    "description": "FAI file is converted to BED format if BED file is not provided.",
                                    "id": "#fai_file",
                                    "sbg:category": "File Input",
                                    "type": [
                                        "null",
                                        "File"
                                    ]
                                },
                                {
                                    "sbg:fileTypes": "BED",
                                    "sbg:stageInput": "link",
                                    "type": [
                                        "null",
                                        "File"
                                    ],
                                    "required": false,
                                    "inputBinding": {
                                        "prefix": "--bed",
                                        "sbg:cmdInclude": true,
                                        "position": 1,
                                        "separate": true
                                    },
                                    "description": "Input BED file containing intervals. Required for modes 3 and 4.",
                                    "id": "#bed_file",
                                    "sbg:category": "File Input",
                                    "label": "Input BED file"
                                }
                            ],
                            "hints": [
                                {
                                    "value": 1000,
                                    "class": "sbg:MemRequirement"
                                },
                                {
                                    "dockerImageId": "",
                                    "class": "DockerRequirement",
                                    "dockerPull": "images.sbgenomics.com/bogdang/sbg_prepare_intervals:1.0"
                                },
                                {
                                    "value": 1,
                                    "class": "sbg:CPURequirement"
                                }
                            ],
                            "sbg:revision": 4,
                            "baseCommand": [
                                "python",
                                "sbg_prepare_intervals.py"
                            ],
                            "id": "bix-demo/sbgtools-demo/sbg-prepare-intervals/4",
                            "sbg:latestRevision": 4,
                            "outputs": [
                                {
                                    "type": [
                                        "null",
                                        "string"
                                    ],
                                    "label": "Output file names",
                                    "outputBinding": {
                                        "glob": "Intervals/names.txt",
                                        "outputEval": {
                                            "script": "{   \n content =  $self[0].contents.replace(/\\0/g, '')\n content = content.replace('[','')\n content = content.replace(']','')\n content = content.replace(/\\'/g, \"\")\n content = content.replace(/\\s/g, '')\n content_arr = content.split(\",\")\n\n return content_arr\n \n\n} ",
                                            "engine": "#cwl-js-engine",
                                            "class": "Expression"
                                        },
                                        "loadContents": true
                                    },
                                    "sbg:fileTypes": "TXT",
                                    "id": "#names",
                                    "description": "File containing the names of created files."
                                },
                                {
                                    "type": [
                                        "null",
                                        {
                                            "name": "intervals",
                                            "type": "array",
                                            "items": "File"
                                        }
                                    ],
                                    "label": "Intervals",
                                    "outputBinding": {
                                        "glob": "Intervals/*.bed",
                                        "sbg:metadata": {
                                            "sbg_scatter": "true"
                                        }
                                    },
                                    "sbg:fileTypes": "BED",
                                    "id": "#intervals",
                                    "description": "Array of BED files genereted as per selected Split Mode."
                                }
                            ],
                            "sbg:toolkitVersion": "1.0",
                            "sbg:toolAuthor": "Seven Bridges Genomics",
                            "successCodes": [],
                            "sbg:revisionNotes": "Fixed Toolkit name.",
                            "arguments": [
                                {
                                    "separate": true,
                                    "valueFrom": {
                                        "script": "{\t\n  if (typeof($job.inputs.format) !== \"undefined\")\n  \treturn \"--format \" + \"\\\"\" + $job.inputs.format + \"\\\"\"\n}",
                                        "engine": "#cwl-js-engine",
                                        "class": "Expression"
                                    }
                                }
                            ],
                            "sbg:id": "admin/sbg-public-data/sbg-prepare-intervals/82",
                            "sbg:job": {
                                "allocatedResources": {
                                    "cpu": 1,
                                    "mem": 1000
                                },
                                "inputs": {
                                    "format": "chr start end",
                                    "fai_file": {
                                        "path": "/path/to/fai_file.ext",
                                        "size": 0,
                                        "class": "File",
                                        "secondaryFiles": []
                                    },
                                    "bed_file": {
                                        "path": "/path/to/bed_file.ext",
                                        "size": 0,
                                        "class": "File",
                                        "secondaryFiles": []
                                    },
                                    "split_mode": null
                                }
                            },
                            "stdin": "",
                            "sbg:project": "bix-demo/sbgtools-demo",
                            "description": "Depending on selected Split Mode value, output files are generated in accordance with description below:\n\n1. Whole Genome - The tool creates one interval file per line of the input BED(FAI) file.\nEach interval file contains a single line (one of the lines of BED(FAI) input file).\n\n2. Whole Genome with reduced number of jobs - For each contig(chromosome) a single file\nis created containing all the intervals corresponding to it .\nAll the intervals (lines) other than (chr1, chr2 ... chrY or 1, 2 ... Y) are saved as\n(\"others.bed\").\n\n3. Whole Exome - BED file is required for execution of this mode. If mode  3 is applied input is passed to the output.\n\n4. Whole Genome Hybrid - For each chromosome a single file is created for each interval.\nAll the intervals (lines) other than (chr1, chr2 ... chrY or 1, 2 ... Y) are saved as\n(\"others.bed\").",
                            "x": 782,
                            "sbg:createdOn": 1473083821,
                            "sbg:cmdPreview": "python sbg_prepare_intervals.py  --format \"chr start end\" --mode 3",
                            "sbg:contributors": [
                                "bix-demo",
                                "medjo",
                                "bogdang",
                                "vladimirk"
                            ],
                            "sbg:sbgMaintained": false,
                            "stdout": ""
                        },
                        "inputs": [
                            {
                                "default": "File per chr with alt contig in a single file",
                                "id": "#SBG_Prepare_Intervals.split_mode"
                            },
                            {
                                "default": "chr start end",
                                "id": "#SBG_Prepare_Intervals.format"
                            },
                            {
                                "id": "#SBG_Prepare_Intervals.fai_file",
                                "source": [
                                    "#fai_file"
                                ]
                            },
                            {
                                "id": "#SBG_Prepare_Intervals.bed_file"
                            }
                        ],
                        "id": "#SBG_Prepare_Intervals",
                        "sbg:y": 494
                    },
                    {
                        "sbg:x": 1109,
                        "outputs": [
                            {
                                "id": "#SBG_SAMtools_Merge_Mpileup.output_file"
                            }
                        ],
                        "run": {
                            "sbg:toolkit": "SBGTools",
                            "requirements": [
                                {
                                    "fileDef": [
                                        {
                                            "filename": "merge.py",
                                            "fileContent": "\"\"\"\nUsage: merge.py     --file-list FILE --output-format FRMT  \n\nDescription:\n  Main function of this tool is to provide option to the user weather to merge list of files from SAMtools Mpileup\n  tool or to pass list of files for each contig to the output. If --Merge option is selected, regardless of the type,\n  files will be merged following the contig order from the BAM header.\n\nOptions:\n\n    --help                              This message.\n\n    --version                           Tool version.\n    \n    --output-format FRMT                Output_format.\n\n    --file-list FILE                    List of contigs in BCF or PILEUP format produced by SAMtools Mpileup tool.\n\n\"\"\"\n\nimport docopt\nfrom subprocess import Popen\n\nargs = docopt.docopt(__doc__, version=\"1.0\") \n\nBCFTOOLS_ROOT = \"bcftools\"\n\nsorted_file_list = list()\nfile_list = args['--file-list'].split(',')\n\noutput_format =  args['--output-format']\n\n\nfor fp in file_list:\n    sorted_file_list.append(fp)\n\n\n\nif output_format ==\"pileup\":\n    with open(sorted_file_list[0].split('/')[-1].replace('.pileup', '.merged.pileup'), 'w')  as rr:\n        cmd = ['cat'] + sorted_file_list\n        p = Popen(cmd, stdout=rr)\n        p.wait()\nelif output_format ==\"bcf\":\n    with open(sorted_file_list[0].split('/')[-1].replace('.bcf', '.merged.bcf'), 'w')  as rr:\n        cmd = [BCFTOOLS_ROOT, 'concat'] + sorted_file_list\n        p = Popen(cmd, stdout=rr)\n        p.wait()\nelif output_format ==\"vcf\":\n    with open(sorted_file_list[0].split('/')[-1].replace('.vcf', '.merged.vcf'), 'w')  as rr:\n        cmd = [BCFTOOLS_ROOT, 'concat'] + sorted_file_list\n        p = Popen(cmd, stdout=rr)\n        p.wait()"
                                        },
                                        {
                                            "filename": "delete_empty.py",
                                            "fileContent": "\"\"\"\nUsage: delete_empty.py     --file-list FILE  \n\nDescription:\n  Main function of this tool is to remove empty pileup files.\n\nOptions:\n\n    --help                              This message.\n\n    --version                           Tool version.\n\n    --file-list FILE                    List of contigs in BCF or PILEUP format produced by SAMtools Mpileup tool.\n\n\"\"\"\n\nimport os\nimport docopt\nfrom subprocess import Popen\n\nargs = docopt.docopt(__doc__, version=\"1.0\") \n\n\nsorted_file_list = list()\nfile_list = args['--file-list'].split(',')\n\nfor fp in file_list:\n    os.remove(fp)"
                                        }
                                    ],
                                    "class": "CreateFileRequirement"
                                },
                                {
                                    "requirements": [
                                        {
                                            "class": "DockerRequirement",
                                            "dockerPull": "rabix/js-engine"
                                        }
                                    ],
                                    "id": "#cwl-js-engine",
                                    "class": "ExpressionEngineRequirement"
                                }
                            ],
                            "sbg:validationErrors": [],
                            "sbg:categories": [
                                "Text-Processing",
                                "SAM/BAM-Processing"
                            ],
                            "label": "SBG SAMtools Merge Mpileup",
                            "sbg:createdBy": "bix-demo",
                            "temporaryFailCodes": [],
                            "sbg:image_url": null,
                            "sbg:revisionsInfo": [
                                {
                                    "sbg:revision": 0,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "bix-demo",
                                    "sbg:modifiedOn": 1450911287
                                },
                                {
                                    "sbg:revision": 1,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "bix-demo",
                                    "sbg:modifiedOn": 1450911287
                                },
                                {
                                    "sbg:revision": 2,
                                    "sbg:revisionNotes": "import from dev project",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1473929938
                                },
                                {
                                    "sbg:revision": 3,
                                    "sbg:revisionNotes": "add some verification to avoid deleting null in case there are not empty files to delete.",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1474019719
                                },
                                {
                                    "sbg:revision": 4,
                                    "sbg:revisionNotes": "Update the merge.py script and fix bug with the output file name",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1474973977
                                },
                                {
                                    "sbg:revision": 5,
                                    "sbg:revisionNotes": "change the python code in merge.py",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1475052403
                                },
                                {
                                    "sbg:revision": 6,
                                    "sbg:revisionNotes": "change the merge.py",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1475058095
                                },
                                {
                                    "sbg:revision": 7,
                                    "sbg:revisionNotes": "add . to the output file name",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1475083435
                                },
                                {
                                    "sbg:revision": 8,
                                    "sbg:revisionNotes": "update parameters ID, and label",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1475237770
                                },
                                {
                                    "sbg:revision": 9,
                                    "sbg:revisionNotes": "link instead of copy inputs",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1476440477
                                },
                                {
                                    "sbg:revision": 10,
                                    "sbg:revisionNotes": "Added inherit metadata from input.",
                                    "sbg:modifiedBy": "milan.domazet.sudo",
                                    "sbg:modifiedOn": 1482334679
                                },
                                {
                                    "sbg:revision": 11,
                                    "sbg:revisionNotes": "fix metadata issue",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1482417569
                                },
                                {
                                    "sbg:revision": 12,
                                    "sbg:revisionNotes": "fix metadata issue when there is not metadata at input files",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1482421824
                                }
                            ],
                            "sbg:toolkitVersion": "0.2",
                            "sbg:license": "Apache License 2.0",
                            "sbg:cmdPreview": "",
                            "sbg:modifiedBy": "marouf",
                            "class": "CommandLineTool",
                            "inputs": [
                                {
                                    "type": [
                                        {
                                            "name": "file_list",
                                            "type": "array",
                                            "items": "File"
                                        }
                                    ],
                                    "sbg:stageInput": "link",
                                    "label": "Input file list",
                                    "sbg:fileTypes": "PILEUP,VCF,BCF",
                                    "id": "#file_list",
                                    "description": "Input files for merging."
                                },
                                {
                                    "type": [
                                        {
                                            "name": "contig_order_names",
                                            "type": "array",
                                            "items": "string"
                                        }
                                    ],
                                    "description": "Ordered names of contigs that are used for merging of file list loaded via file_list input.",
                                    "label": "merging_order",
                                    "id": "#contig_order_names",
                                    "sbg:includeInPorts": true
                                },
                                {
                                    "description": "Mode to tell how to merge mpileup files or to pass them through.",
                                    "label": "Output state",
                                    "id": "#output_state",
                                    "type": [
                                        {
                                            "name": "output_state",
                                            "symbols": [
                                                "Merge",
                                                "Pass nonempty files",
                                                "Pass all files"
                                            ],
                                            "type": "enum"
                                        }
                                    ]
                                }
                            ],
                            "hints": [
                                {
                                    "value": 1,
                                    "class": "sbg:CPURequirement"
                                },
                                {
                                    "value": 1000,
                                    "class": "sbg:MemRequirement"
                                },
                                {
                                    "dockerImageId": "",
                                    "class": "DockerRequirement",
                                    "dockerPull": "images.sbgenomics.com/marouf/sbg-pileup-merge-out:0.1"
                                }
                            ],
                            "sbg:revision": 12,
                            "baseCommand": [
                                ""
                            ],
                            "id": "https://api.sbgenomics.com/v2/apps/bix-demo/sbgtools-demo/sbg-samtools-mpileup-merge-out/12/raw/",
                            "sbg:latestRevision": 12,
                            "outputs": [
                                {
                                    "type": [
                                        "null",
                                        {
                                            "type": "array",
                                            "items": "File"
                                        }
                                    ],
                                    "label": "output_file",
                                    "outputBinding": {
                                        "glob": {
                                            "script": "{\n  \n  \n   if($job.inputs.output_state.includes(\"Merge\"))\n   {\n    \n    format = $job.inputs.file_list[0].path.split(\"/\").pop()\n    format = format.split('.').pop()\n\n  \treturn \"*.merged.\"+format.toLowerCase()\n   }\n  else\n  {\n    return \"{*.pileup,*.vcf,*.bcf}\"\n  }\n \n}",
                                            "engine": "#cwl-js-engine",
                                            "class": "Expression"
                                        },
                                        "outputEval": {
                                            "script": "{\n  \n  var inputs = {};\n\n  to_ret = \"\"\n\n  for ( fn in $job.inputs.file_list)\n  {\n    if($job.inputs.file_list[fn].size!==0)\n    {\n\n      name = $job.inputs.file_list[fn].path.split(\"/\").pop()  \n       \n\n\t if(\"metadata\" in $job.inputs.file_list[fn])\n     { \n     \n      inputs[name] = $job.inputs.file_list[fn].metadata\n      \n     }\n\n    }\n  }\n             \n  content = $self\n\n  if($job.inputs.output_state.includes(\"Merge\"))\n  {  \n    \n    // set metadata from any input to this tool\n    name =  content[0][\"name\"]\n    \n    if (typeof name !== 'undefined')\n    {\n      metadata = inputs[name]\n      \n\t if(typeof metadata !== 'undefined')\n     {\n      if (\"ScatteredUsing\" in metadata)\n\t\tdelete metadata[\"ScatteredUsing\"]\n        \n      for ( var fp in content )\n       {\n\n          content[fp][\"metadata\"] = metadata\n\n       }\n     }\n    }\n  \n  }\n   else\n   {\n     for ( var fp in content )\n     {\n\n       name =  content[fp][\"name\"]\n       \n       if (name in inputs)\n       {\n         metadata = inputs[name]\n\n         content[fp][\"metadata\"] = metadata\n       }  \n      \n     }\n\n     return content\n   }\n\n  return content\n}\n\n\n\n\n\n\n\n",
                                            "engine": "#cwl-js-engine",
                                            "class": "Expression"
                                        }
                                    },
                                    "sbg:fileTypes": "PILEUP,BCF,VCF",
                                    "id": "#output_file",
                                    "description": "File obtained by merging files form #file_list."
                                }
                            ],
                            "successCodes": [],
                            "sbg:toolAuthor": "Mohamed Marouf, Seven Bridges Genomics, <mohamed.marouf@sbgenomics.com>",
                            "sbg:revisionNotes": "fix metadata issue when there is not metadata at input files",
                            "arguments": [
                                {
                                    "prefix": "",
                                    "position": 3,
                                    "valueFrom": {
                                        "script": "{\n\n  if($job.inputs.output_state.includes(\"Merge\"))\n  {\n    to_ret = \"\"\n    for(f in $job.inputs.contig_order_names)\n    {\n\n      for ( fn in $job.inputs.file_list)\n      {\n        if($job.inputs.file_list[fn].metadata)\n        {\n\n\n          if($job.inputs.contig_order_names[f] == $job.inputs.file_list[fn].metadata.ScatteredUsing)\n          {\n            if(to_ret == \"\")\n              to_ret =  $job.inputs.file_list[fn].path\n            else\n    \t\t  to_ret =  to_ret + \",\"+ $job.inputs.file_list[fn].path\n\n          }\n        }\n      }\n\n    }\n    \n    format = $job.inputs.file_list[0].path.split(\"/\").pop()\n    format = format.split('.').pop()\n    \n    \n    if(to_ret==\"\")\n    \treturn \n    else\n      \treturn \"python merge.py --output-format \" + format.toLowerCase()+ \" --file-list \"+ to_ret\n  \n  }\n  else if($job.inputs.output_state == \"Pass nonempty files\")\n  {\n    to_ret = \"\"\n          \n    for ( fn in $job.inputs.file_list)\n      {\n        if($job.inputs.file_list[fn].size==0)\n        {\n\n          if(to_ret == \"\")\n            to_ret =  $job.inputs.file_list[fn].path\n          else\n            to_ret =  to_ret + \",\"+ $job.inputs.file_list[fn].path\n\n        }\n      }\n    \n    if(to_ret==\"\")\n    \treturn \n    else\n  \t\treturn \"python delete_empty.py --file-list \"+ to_ret\n          \n  }\n\n\n\n          \n}",
                                        "engine": "#cwl-js-engine",
                                        "class": "Expression"
                                    },
                                    "separate": true
                                }
                            ],
                            "sbg:id": "admin/sbg-public-data/sbg-samtools-mpileup-merge-out/12",
                            "sbg:job": {
                                "allocatedResources": {
                                    "cpu": 1,
                                    "mem": 1000
                                },
                                "inputs": {
                                    "contig_order_names": [
                                        "1",
                                        "2",
                                        "3"
                                    ],
                                    "file_list": [
                                        {
                                            "metadata": {
                                                "ScatteredUsing": "3"
                                            },
                                            "path": "/ovde/1994060146_RNASeq_R.Aligned.toTranscriptome.out_22.vcf",
                                            "size": 0,
                                            "class": "File",
                                            "secondaryFiles": []
                                        },
                                        {
                                            "metadata": {
                                                "SID": "111",
                                                "ScatteredUsing": "1"
                                            },
                                            "path": "1994060146_RNASeq_R.Aligned.toTranscriptome.out_11.vcf",
                                            "size": 1,
                                            "class": "File",
                                            "secondaryFiles": []
                                        },
                                        {
                                            "path": "file.txt",
                                            "size": 1,
                                            "class": "File",
                                            "secondaryFiles": []
                                        }
                                    ],
                                    "output_state": "Pass all files"
                                }
                            },
                            "stdin": "",
                            "sbg:project": "bix-demo/sbgtools-demo",
                            "description": "Main function of this tool is to provide option to the user whether to merge list of files from SAMtools Mpileup tool or to pass list of files for each contig to the output. \n\nIf MERGE option is selected, regardless of the type, files will be merged following the contig order from the reference file index file or BED file. This order is provided using input contig_order_names.  So, this tool will receive list of pileup, vcf or BC files and list of files names of BED files used in creating input files. Pairing and finding the correct order will be done using metadata field  ScatteredUsing.\n\n####Common issue:\nThis tools is built specially for mpileup parallel and Varscan pipelines, so please don't use it outside these pipeline unless you are sure how to use it. Contact our support team for information.",
                            "sbg:createdOn": 1450911287,
                            "cwlVersion": "sbg:draft-2",
                            "sbg:contributors": [
                                "milan.domazet.sudo",
                                "bix-demo",
                                "marouf"
                            ],
                            "sbg:modifiedOn": 1482421824,
                            "sbg:sbgMaintained": false,
                            "stdout": ""
                        },
                        "inputs": [
                            {
                                "id": "#SBG_SAMtools_Merge_Mpileup.file_list",
                                "source": [
                                    "#SAMtools_Mpileup.output_pileup_vcf_or_bcf_file"
                                ]
                            },
                            {
                                "id": "#SBG_SAMtools_Merge_Mpileup.contig_order_names",
                                "source": [
                                    "#SBG_Prepare_Intervals.names"
                                ]
                            },
                            {
                                "id": "#SBG_SAMtools_Merge_Mpileup.output_state",
                                "source": [
                                    "#output_state"
                                ]
                            }
                        ],
                        "id": "#SBG_SAMtools_Merge_Mpileup",
                        "sbg:y": 447
                    }
                ],
                "sbg:toolAuthor": "Heng Li, Sanger Institute",
                "sbg:revisionNotes": "fix metadata issue when there is no metadata at inputs",
                "sbg:modifiedBy": "marouf",
                "sbg:id": "admin/sbg-public-data/samtools-mpileup-parallel/22",
                "sbg:canvas_y": -51,
                "sbg:project": "marouf/samtools-mpileup-parallel-demo",
                "description": "The purpose of this workflow is to allow user to execute the SAMtools Mpileup tool in parallel using all of the instance resources. \n\nInstead of analyzing whole BAM at once, this workflow obtains contigs information from FASTA reference index (FAI). This file contains a list of contigs where pileup or BCF should be generated. This is done by using parameter (-I or --positions FILE) of the SAMtools Mpileup tool. Jobs will be then executed per chromosome on each of the available core of the instance, while all small contigs will be executed together in a single job on one core.\n\nAs a result, execution time will be order of magnitude faster compared to the execution of the whole BAM at once. Also, output of the workflow can be an integral file for the whole BAM, or list of files per chromosome which is defined by the parameter \"output_state\". Finally, user can choose format of the output file or list of files by the SAMtools mpileup parameter output_type. Available output formats are PILEUP, VCF and BCF.\n\n######Required inputs:\n\n- Reference file-The faidx-indexed reference file in the FASTA format.\n\n- BAM file-One or a list of input BAM files.\n\n- FAI file-One of these two is required input. FAI file is used to extract intervals that will be used to scatter the SAMtools mpileup job. FAI file can be generated using SAMtools faidx. \n\n- Merging mode-This defines the way this pipeline outputs its results. Options are:\n      - \"Merge\": to produce VCF, PILEUP, or BCF file acquired after merging results of the scattered pileup.\n      - \"Pass nonempty files\"  to pass through all nonempty PILEUP, VF, BCF files without merging.\n      - \"Pass all files\": to output all PILEUP, VF, BCF Separately.\n\n- Output file format -This parameter can be used to change the output format (available choices are VCF/BCF and pileup)\n\n###### Outputs:\nThis pipeline has one output port and the output depends on the selected input parameters (merge mode and output file format). Please refer to these parameters for more details.\n\n###### Common Issues\nBAM files should be sorted by coordinates before using SAMmpileup.",
                "sbg:createdOn": 1474020918,
                "sbg:canvas_zoom": 1,
                "outputs": [
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "sbg:x": 1257,
                        "label": "output_pileup",
                        "source": [
                            "#SBG_SAMtools_Merge_Mpileup.output_file"
                        ],
                        "required": false,
                        "id": "#output_file",
                        "sbg:y": 447,
                        "sbg:includeInPorts": true
                    }
                ],
                "cwlVersion": "sbg:draft-2",
                "sbg:contributors": [
                    "milan.domazet.sudo",
                    "marouf"
                ],
                "sbg:sbgMaintained": false
            },
            "inputs": [
                {
                    "id": "#SAMtools_Mpileup_parallel.bams",
                    "source": [
                        "#BamTools_Index_Tumor.output_bam_file"
                    ]
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.reference_fasta",
                    "source": [
                        "#SAMtools_Index_FASTA.output_fasta_file"
                    ]
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.fai_file",
                    "source": [
                        "#SAMtools_Index_FASTA.generated_index"
                    ]
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.skip_bases_with_baq_smaller_than_defined"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.required_flags"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.region_pileup_generation"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.recalc_BAQ_fly"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.per_sample_mF"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.parameter_for_adjusting_mapq"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.output_uncompressed_bcf_or_vcf"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.output_mapping_quality"
                },
                {
                    "default": "PILEUP",
                    "id": "#SAMtools_Mpileup_parallel.output_format"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.output_base_positions_on_reads"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.no_indel_calling"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.more_info_to_output"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.minimum_fraction_of_gapped_reads"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.min_gapped_reads_for_indel"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.max_per_bam_depth"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.max_idepth"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.mapq_threshold"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.ignore_rg_tags"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.ignore_overlaps"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.homopolymer_err_coeficient"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.gap_open_sequencing_error_probability"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.gap_extension_seq_error_probability"
                },
                {
                    "default": true,
                    "id": "#SAMtools_Mpileup_parallel.filter_zero_coverage_lines"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.filter_flags"
                },
                {
                    "default": true,
                    "id": "#SAMtools_Mpileup_parallel.disable_baq_computation"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.count_anomalous_read_pairs"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.comma_separated_list_of_platforms_for_indels"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel.assume_the_quality_Illumina_encoding"
                },
                {
                    "default": "Pass nonempty files",
                    "id": "#SAMtools_Mpileup_parallel.output_state"
                }
            ],
            "id": "#SAMtools_Mpileup_parallel",
            "sbg:y": 251.04741114095313
        },
        {
            "sbg:x": 685.4164649148879,
            "outputs": [
                {
                    "id": "#SAMtools_Mpileup_parallel_1.output_file"
                }
            ],
            "run": {
                "sbg:toolkit": "SAMtools",
                "requirements": [],
                "sbg:validationErrors": [],
                "sbg:categories": [
                    "SAM/BAM-Processing"
                ],
                "class": "Workflow",
                "label": "SAMtools Mpileup parallel",
                "sbg:createdBy": "marouf",
                "sbg:image_url": "https://brood.sbgenomics.com/static/marouf/samtools-mpileup-parallel-demo/samtools-mpileup-parallel/22.png",
                "sbg:links": [
                    {
                        "label": "Homepage",
                        "id": "http://samtools.sourceforge.net/"
                    },
                    {
                        "label": "Source Code",
                        "id": "https://github.com/samtools/samtools"
                    },
                    {
                        "label": "Wiki",
                        "id": "http://sourceforge.net/p/samtools/wiki/Home/"
                    },
                    {
                        "label": "Download",
                        "id": "http://sourceforge.net/projects/samtools/files/samtools/0.1.18/"
                    },
                    {
                        "label": "Publication",
                        "id": "http://www.ncbi.nlm.nih.gov/pubmed/19505943"
                    },
                    {
                        "label": "Documentation",
                        "id": "http://www.htslib.org/doc/samtools.html"
                    }
                ],
                "sbg:revisionsInfo": [
                    {
                        "sbg:revision": 0,
                        "sbg:revisionNotes": null,
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1474020918
                    },
                    {
                        "sbg:revision": 1,
                        "sbg:revisionNotes": "Update workflow from dev project",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1474020953
                    },
                    {
                        "sbg:revision": 2,
                        "sbg:revisionNotes": "update tools in the pipeline",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1474972698
                    },
                    {
                        "sbg:revision": 3,
                        "sbg:revisionNotes": "Update merge pileup tool",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1474974158
                    },
                    {
                        "sbg:revision": 4,
                        "sbg:revisionNotes": "Change the description after peer review",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1474978330
                    },
                    {
                        "sbg:revision": 5,
                        "sbg:revisionNotes": "uncheck the required checkbox for the fai file input",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1474987689
                    },
                    {
                        "sbg:revision": 6,
                        "sbg:revisionNotes": "Add suggested files",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1474988027
                    },
                    {
                        "sbg:revision": 7,
                        "sbg:revisionNotes": "update the merge mpileup tool",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1475052533
                    },
                    {
                        "sbg:revision": 8,
                        "sbg:revisionNotes": "Change the description",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1475055294
                    },
                    {
                        "sbg:revision": 9,
                        "sbg:revisionNotes": "update merge_pileup tool",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1475058452
                    },
                    {
                        "sbg:revision": 10,
                        "sbg:revisionNotes": "Update the merge pileup and change the description",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1475083788
                    },
                    {
                        "sbg:revision": 11,
                        "sbg:revisionNotes": "change parameters settings",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1475237926
                    },
                    {
                        "sbg:revision": 12,
                        "sbg:revisionNotes": "provide better description and remove bed file",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1475238571
                    },
                    {
                        "sbg:revision": 13,
                        "sbg:revisionNotes": "more description",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1475655549
                    },
                    {
                        "sbg:revision": 14,
                        "sbg:revisionNotes": "change merge pileup",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1476440562
                    },
                    {
                        "sbg:revision": 15,
                        "sbg:revisionNotes": "change description",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1476794923
                    },
                    {
                        "sbg:revision": 16,
                        "sbg:revisionNotes": "Exposed mapq_threshold.",
                        "sbg:modifiedBy": "milan.domazet.sudo",
                        "sbg:modifiedOn": 1477483754
                    },
                    {
                        "sbg:revision": 17,
                        "sbg:revisionNotes": "Exposed mapq_threshold.",
                        "sbg:modifiedBy": "milan.domazet.sudo",
                        "sbg:modifiedOn": 1477483870
                    },
                    {
                        "sbg:revision": 18,
                        "sbg:revisionNotes": "Add filter to mpileu tool",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1478526969
                    },
                    {
                        "sbg:revision": 19,
                        "sbg:revisionNotes": "expose filter",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1478527215
                    },
                    {
                        "sbg:revision": 20,
                        "sbg:revisionNotes": "Updated all tools (mostly inherit metadata from input)",
                        "sbg:modifiedBy": "milan.domazet.sudo",
                        "sbg:modifiedOn": 1482334741
                    },
                    {
                        "sbg:revision": 21,
                        "sbg:revisionNotes": "fix metadata",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1482417615
                    },
                    {
                        "sbg:revision": 22,
                        "sbg:revisionNotes": "fix metadata issue when there is no metadata at inputs",
                        "sbg:modifiedBy": "marouf",
                        "sbg:modifiedOn": 1482421868
                    }
                ],
                "sbg:toolkitVersion": "1.3",
                "sbg:license": "The MIT License",
                "sbg:canvas_x": -61,
                "inputs": [
                    {
                        "type": [
                            {
                                "type": "array",
                                "items": "File"
                            }
                        ],
                        "sbg:x": 627,
                        "label": "bams",
                        "sbg:fileTypes": "BAM",
                        "id": "#bams",
                        "sbg:y": 348,
                        "sbg:includeInPorts": true
                    },
                    {
                        "type": [
                            "File"
                        ],
                        "sbg:x": 624,
                        "sbg:includeInPorts": true,
                        "sbg:suggestedValue": {
                            "path": "5772b6d8507c1752674486e6",
                            "name": "human_g1k_v37_decoy.fasta",
                            "class": "File"
                        },
                        "sbg:fileTypes": "FASTA,FA,GZ",
                        "id": "#reference_fasta",
                        "sbg:y": 221,
                        "label": "reference_fasta"
                    },
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "sbg:x": 629,
                        "sbg:includeInPorts": true,
                        "sbg:suggestedValue": {
                            "path": "578cf949507c17681a3117e2",
                            "name": "human_g1k_v37_decoy.fasta.fai",
                            "class": "File"
                        },
                        "sbg:fileTypes": "FAI",
                        "id": "#fai_file",
                        "sbg:y": 469,
                        "label": "fai_file"
                    },
                    {
                        "sbg:includeInPorts": false,
                        "label": "Skip bases with baseQ/BAQ smaller than",
                        "required": false,
                        "sbg:toolDefaultValue": "13",
                        "description": "Minimum base quality for a base to be considered.",
                        "id": "#skip_bases_with_baq_smaller_than_defined",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "type": [
                            "null",
                            "int"
                        ],
                        "label": "Required flags",
                        "required": false,
                        "description": "Required flags: skip reads with mask bits unset.",
                        "id": "#required_flags",
                        "sbg:category": "Input options",
                        "sbg:includeInPorts": false
                    },
                    {
                        "type": [
                            "null",
                            "string"
                        ],
                        "label": "Region in which pileup is generated",
                        "required": false,
                        "description": "Region in which pileup is generated.",
                        "id": "#region_pileup_generation",
                        "sbg:category": "Input options",
                        "sbg:includeInPorts": false
                    },
                    {
                        "type": [
                            "null",
                            "boolean"
                        ],
                        "label": "Recalculate BAQ on the fly",
                        "required": false,
                        "description": "Recalculate BAQ on the fly, ignore existing BQ tags.",
                        "id": "#recalc_BAQ_fly",
                        "sbg:category": "Input options",
                        "sbg:includeInPorts": false
                    },
                    {
                        "sbg:stageInput": null,
                        "label": "Apply -m and -F thresholds per sample",
                        "description": "Apply -m and -F thresholds per sample to increase sensitivity of calling. By default both options are applied to reads pooled from all samples.",
                        "id": "#per_sample_mF",
                        "sbg:category": "configuration",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "sbg:includeInPorts": false,
                        "label": "Parameter for adjusting mapQ",
                        "required": false,
                        "sbg:toolDefaultValue": "0",
                        "description": "Coefficient for downgrading mapping quality for reads containing excessive mismatches. Given a read with a phred-scaled probability q of being generated from the mapped position, the new mapping quality is about sqrt((INT-q)/INT)*INT. A zero value disables this functionality; if enabled, the recommended value for BWA is 50.",
                        "id": "#parameter_for_adjusting_mapq",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "description": "Similar to -g except that the output is uncompressed BCF, which is preferred for piping.",
                        "id": "#output_uncompressed_bcf_or_vcf",
                        "label": "Generate uncompress BCF/VCF output",
                        "sbg:category": "Output options",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "description": "Output mapping quality (disabled by -g/-u).",
                        "id": "#output_mapping_quality",
                        "label": "Output mapping quality",
                        "sbg:category": "Output options",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "description": "Output file format.",
                        "id": "#output_format",
                        "label": "Output file format",
                        "sbg:category": "configuration",
                        "type": [
                            {
                                "name": "output_format",
                                "symbols": [
                                    "PILEUP",
                                    "BCF",
                                    "VCF"
                                ],
                                "type": "enum"
                            }
                        ]
                    },
                    {
                        "description": "Output base positions on reads (disabled by -g/-u).",
                        "id": "#output_base_positions_on_reads",
                        "label": "Output base positions on reads",
                        "sbg:category": "Output options",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "description": "Do not perform indel calling.",
                        "id": "#no_indel_calling",
                        "label": "Do not perform indel calling",
                        "sbg:category": "SNP/INDEL genotype likelihoods options",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "description": "Comma-separated list of FORMAT and INFO tags (DP,AD,ADF,ADR,SP,INFO/AD,INFO/ADF,INFO/ADR []\") to output.",
                        "id": "#more_info_to_output",
                        "label": "Comma-separated list of FORMAT and INFO tags to output",
                        "sbg:category": "configuration",
                        "type": [
                            "null",
                            "string"
                        ]
                    },
                    {
                        "label": "Minimum fraction of gapped reads for candidates",
                        "sbg:toolDefaultValue": "0.002",
                        "description": "Minimum fraction of gapped reads for candidates.",
                        "id": "#minimum_fraction_of_gapped_reads",
                        "sbg:category": "SNP/INDEL genotype likelihoods options",
                        "type": [
                            "null",
                            "float"
                        ]
                    },
                    {
                        "label": "Minimum gapped reads for indel candidates",
                        "sbg:toolDefaultValue": "1",
                        "description": "Minimum gapped reads for indel candidates.",
                        "id": "#min_gapped_reads_for_indel",
                        "sbg:category": "SNP/INDEL genotype likelihoods options",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "label": "Max per-BAM depth",
                        "sbg:toolDefaultValue": "250",
                        "description": "At a position, read maximally INT reads per input BAM.",
                        "id": "#max_per_bam_depth",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "label": "Skip INDEL calling if the average per-sample depth is above",
                        "type": [
                            "null",
                            "int"
                        ],
                        "sbg:toolDefaultValue": "250",
                        "schema": [
                            "null",
                            "int"
                        ],
                        "id": "#max_idepth",
                        "sbg:category": "configuration",
                        "description": "Skip INDEL calling if the average per-sample depth is above."
                    },
                    {
                        "description": "Minimum mapping quality for an alignment to be used.",
                        "id": "#mapq_threshold",
                        "label": "Skip alignments with mapQ  smaller than",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "description": "Ignore rg tags.",
                        "id": "#ignore_rg_tags",
                        "label": "Ignore RG tags",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "description": "Disable read-pair overlap detection.",
                        "id": "#ignore_overlaps",
                        "label": "Disable read-pair overlap detection",
                        "sbg:category": "configuration",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "label": "Coefficient for homopolymer errors",
                        "sbg:toolDefaultValue": "100",
                        "description": "Coefficient for modeling homopolymer errors. Given an l-long homopolymer run, the sequencing error of an indel of size s is modeled as INT*s/l.",
                        "id": "#homopolymer_err_coeficient",
                        "sbg:category": "SNP/INDEL genotype likelihoods options",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "label": "Phred-scaled gap open sequencing error probability",
                        "sbg:toolDefaultValue": "40",
                        "description": "Phred-scaled gap open sequencing error probability. Reducing INT leads to more indel calls.",
                        "id": "#gap_open_sequencing_error_probability",
                        "sbg:category": "SNP/INDEL genotype likelihoods options",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "label": "Phred-scaled gap extension seq error probability",
                        "sbg:toolDefaultValue": "20",
                        "description": "Phred-scaled gap extension sequencing error probability. Reducing INT leads to longer indels.",
                        "id": "#gap_extension_seq_error_probability",
                        "sbg:category": "SNP/INDEL genotype likelihoods options",
                        "type": [
                            "null",
                            "int"
                        ]
                    },
                    {
                        "sbg:stageInput": null,
                        "label": "Filter zero coverage lines.",
                        "description": "Filter zero coverage lines from pileup files. This option is valid only when output pileup files.",
                        "id": "#filter_zero_coverage_lines",
                        "sbg:category": "output configuration",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "sbg:stageInput": null,
                        "label": "Filter flags",
                        "description": "Filter flags: skip reads with mask bits set [UNMAP,SECONDARY,QCFAIL,DUP].",
                        "id": "#filter_flags",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "string"
                        ]
                    },
                    {
                        "description": "Disable probabilistic realignment for the computation of base alignment quality (BAQ). BAQ is the Phred-scaled probability of a read base being misaligned. Applying this option greatly helps to reduce false SNPs caused by misalignments.",
                        "id": "#disable_baq_computation",
                        "label": "Disable BAQ computation",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "description": "Do not skip anomalous read pairs in variant calling.",
                        "id": "#count_anomalous_read_pairs",
                        "label": "Count anomalous read pairs",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "label": "Comma separated list of platforms for indels",
                        "sbg:toolDefaultValue": "\"all\"",
                        "description": "Comma dilimited list of platforms (determined by @RG-PL) from which indel candidates are obtained. It is recommended to collect indel candidates from sequencing technologies that have low indel error rate such as ILLUMINA.",
                        "id": "#comma_separated_list_of_platforms_for_indels",
                        "sbg:category": "SNP/INDEL genotype likelihoods options",
                        "type": [
                            "null",
                            "string"
                        ]
                    },
                    {
                        "description": "Assume the quality is in the Illumina-1.3+ encoding.",
                        "id": "#assume_the_quality_Illumina_encoding",
                        "label": "Assume the quality is in the Illumina-1.3+ encoding",
                        "sbg:category": "Input options",
                        "type": [
                            "null",
                            "boolean"
                        ]
                    },
                    {
                        "description": "Mode to tell how to merge mpileup files or to pass them through.",
                        "label": "Output state",
                        "id": "#output_state",
                        "type": [
                            {
                                "name": "output_state",
                                "symbols": [
                                    "Merge",
                                    "Pass nonempty files",
                                    "Pass all files"
                                ],
                                "type": "enum"
                            }
                        ]
                    }
                ],
                "hints": [
                    {
                        "value": "cc2.8xlarge",
                        "class": "sbg:forceAmazonInstance"
                    },
                    {
                        "value": "n1-standard-32",
                        "class": "sbg:forceGoogleInstance"
                    }
                ],
                "sbg:revision": 22,
                "sbg:modifiedOn": 1482421868,
                "id": "https://api.sbgenomics.com/v2/apps/marouf/samtools-mpileup-parallel-demo/samtools-mpileup-parallel/22/raw/",
                "sbg:latestRevision": 22,
                "steps": [
                    {
                        "scatter": "#SAMtools_Mpileup.chr_pos_list_file",
                        "inputs": [
                            {
                                "id": "#SAMtools_Mpileup.skip_bases_with_baq_smaller_than_defined",
                                "source": [
                                    "#skip_bases_with_baq_smaller_than_defined"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.required_flags",
                                "source": [
                                    "#required_flags"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.region_pileup_generation",
                                "source": [
                                    "#region_pileup_generation"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.reference_fasta_file",
                                "source": [
                                    "#reference_fasta"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.recalc_BAQ_fly",
                                "source": [
                                    "#recalc_BAQ_fly"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.per_sample_mF",
                                "source": [
                                    "#per_sample_mF"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.parameter_for_adjusting_mapq",
                                "source": [
                                    "#parameter_for_adjusting_mapq"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.output_uncompressed_bcf_or_vcf",
                                "source": [
                                    "#output_uncompressed_bcf_or_vcf"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.output_mapping_quality",
                                "source": [
                                    "#output_mapping_quality"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.output_format",
                                "source": [
                                    "#output_format"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.output_base_positions_on_reads",
                                "source": [
                                    "#output_base_positions_on_reads"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.no_indel_calling",
                                "source": [
                                    "#no_indel_calling"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.more_info_to_output",
                                "source": [
                                    "#more_info_to_output"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.minimum_fraction_of_gapped_reads",
                                "source": [
                                    "#minimum_fraction_of_gapped_reads"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.min_gapped_reads_for_indel",
                                "source": [
                                    "#min_gapped_reads_for_indel"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.max_per_bam_depth",
                                "source": [
                                    "#max_per_bam_depth"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.max_idepth",
                                "source": [
                                    "#max_idepth"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.mapq_threshold",
                                "source": [
                                    "#mapq_threshold"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.input_bam_files",
                                "source": [
                                    "#bams"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.ignore_rg_tags",
                                "source": [
                                    "#ignore_rg_tags"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.ignore_overlaps",
                                "source": [
                                    "#ignore_overlaps"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.homopolymer_err_coeficient",
                                "source": [
                                    "#homopolymer_err_coeficient"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.gap_open_sequencing_error_probability",
                                "source": [
                                    "#gap_open_sequencing_error_probability"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.gap_extension_seq_error_probability",
                                "source": [
                                    "#gap_extension_seq_error_probability"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.filter_zero_coverage_lines",
                                "source": [
                                    "#filter_zero_coverage_lines"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.filter_flags",
                                "source": [
                                    "#filter_flags"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.exclude_read_groups_list_file"
                            },
                            {
                                "id": "#SAMtools_Mpileup.disable_baq_computation",
                                "source": [
                                    "#disable_baq_computation"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.count_anomalous_read_pairs",
                                "source": [
                                    "#count_anomalous_read_pairs"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.comma_separated_list_of_platforms_for_indels",
                                "source": [
                                    "#comma_separated_list_of_platforms_for_indels"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.chr_pos_list_file",
                                "source": [
                                    "#SBG_Prepare_Intervals.intervals"
                                ]
                            },
                            {
                                "id": "#SAMtools_Mpileup.assume_the_quality_Illumina_encoding",
                                "source": [
                                    "#assume_the_quality_Illumina_encoding"
                                ]
                            }
                        ],
                        "outputs": [
                            {
                                "id": "#SAMtools_Mpileup.output_pileup_vcf_or_bcf_file"
                            }
                        ],
                        "sbg:x": 924,
                        "run": {
                            "sbg:toolkit": "SAMtools",
                            "requirements": [
                                {
                                    "requirements": [
                                        {
                                            "class": "DockerRequirement",
                                            "dockerPull": "rabix/js-engine"
                                        }
                                    ],
                                    "id": "#cwl-js-engine",
                                    "class": "ExpressionEngineRequirement"
                                }
                            ],
                            "sbg:validationErrors": [],
                            "sbg:categories": [
                                "SAM/BAM-Processing"
                            ],
                            "label": "SAMtools Mpileup",
                            "sbg:createdBy": "marouf",
                            "sbg:cmdPreview": "/opt/samtools-1.3/samtools mpileup    input_bam_file1.bam  input_bam_file2.bam   | awk '{if($4 != 0) print $0}'  > input_bam_file1.pileup",
                            "sbg:image_url": null,
                            "cwlVersion": "sbg:draft-2",
                            "sbg:revisionsInfo": [
                                {
                                    "sbg:revision": 0,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1458812970
                                },
                                {
                                    "sbg:revision": 1,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1458825811
                                },
                                {
                                    "sbg:revision": 2,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1462805320
                                },
                                {
                                    "sbg:revision": 3,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1468578873
                                },
                                {
                                    "sbg:revision": 4,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1468580666
                                },
                                {
                                    "sbg:revision": 5,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1468586077
                                },
                                {
                                    "sbg:revision": 6,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1468924014
                                },
                                {
                                    "sbg:revision": 7,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1473160925
                                },
                                {
                                    "sbg:revision": 8,
                                    "sbg:revisionNotes": "Peer review",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1473433741
                                },
                                {
                                    "sbg:revision": 9,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1473434123
                                },
                                {
                                    "sbg:revision": 10,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1473771444
                                },
                                {
                                    "sbg:revision": 11,
                                    "sbg:revisionNotes": "change positional argument for BCF output",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1473779051
                                },
                                {
                                    "sbg:revision": 12,
                                    "sbg:revisionNotes": "Delete default value inserted mistakenly for some parameters.",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1473849264
                                },
                                {
                                    "sbg:revision": 13,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1473931426
                                },
                                {
                                    "sbg:revision": 14,
                                    "sbg:revisionNotes": "add new common issue",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1474901461
                                },
                                {
                                    "sbg:revision": 15,
                                    "sbg:revisionNotes": "Add filter to pileup files",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1478526836
                                },
                                {
                                    "sbg:revision": 16,
                                    "sbg:revisionNotes": "change the test case",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1478526922
                                },
                                {
                                    "sbg:revision": 17,
                                    "sbg:revisionNotes": "update description",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1480075403
                                },
                                {
                                    "sbg:revision": 18,
                                    "sbg:revisionNotes": "output is not required field anymore",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1480440302
                                },
                                {
                                    "sbg:revision": 19,
                                    "sbg:revisionNotes": "Added inherit metadata from input.",
                                    "sbg:modifiedBy": "milan.domazet.sudo",
                                    "sbg:modifiedOn": 1482334695
                                }
                            ],
                            "y": 361,
                            "sbg:toolkitVersion": "v1.3",
                            "sbg:license": "The MIT License",
                            "sbg:modifiedBy": "milan.domazet.sudo",
                            "class": "CommandLineTool",
                            "inputs": [
                                {
                                    "label": "Skip bases with baseQ/BAQ smaller than",
                                    "inputBinding": {
                                        "prefix": "-Q",
                                        "sbg:cmdInclude": true,
                                        "separate": true,
                                        "separator": " ",
                                        "position": 0
                                    },
                                    "sbg:toolDefaultValue": "13",
                                    "description": "Minimum base quality for a base to be considered.",
                                    "id": "#skip_bases_with_baq_smaller_than_defined",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "label": "Required flags",
                                    "inputBinding": {
                                        "prefix": "--rf",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Required flags: skip reads with mask bits unset.",
                                    "id": "#required_flags",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "label": "Region in which pileup is generated",
                                    "inputBinding": {
                                        "prefix": "-r",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Region in which pileup is generated.",
                                    "id": "#region_pileup_generation",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "string"
                                    ]
                                },
                                {
                                    "sbg:fileTypes": "FASTA,FA,GZ",
                                    "label": "Reference FASTA file",
                                    "type": [
                                        "null",
                                        "File"
                                    ],
                                    "required": false,
                                    "inputBinding": {
                                        "prefix": "-f",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "secondaryFiles": [
                                            ".fai"
                                        ],
                                        "separate": true
                                    },
                                    "description": "Faidx indexed reference sequence (FASTA) file.",
                                    "id": "#reference_fasta_file",
                                    "sbg:category": "File input",
                                    "schema": [
                                        "null",
                                        "File"
                                    ]
                                },
                                {
                                    "label": "Recalculate BAQ on the fly",
                                    "inputBinding": {
                                        "prefix": "-E",
                                        "sbg:cmdInclude": true,
                                        "separate": true,
                                        "separator": " ",
                                        "position": 0
                                    },
                                    "description": "Recalculate BAQ on the fly, ignore existing BQ tags.",
                                    "id": "#recalc_BAQ_fly",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "sbg:stageInput": null,
                                    "type": [
                                        "null",
                                        "boolean"
                                    ],
                                    "inputBinding": {
                                        "prefix": "-p",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Apply -m and -F thresholds per sample to increase sensitivity of calling. By default both options are applied to reads pooled from all samples.",
                                    "id": "#per_sample_mF",
                                    "sbg:category": "configuration",
                                    "label": "Apply -m and -F thresholds per sample"
                                },
                                {
                                    "label": "Parameter for adjusting mapQ",
                                    "inputBinding": {
                                        "prefix": "-C",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Coefficient for downgrading mapping quality for reads containing excessive mismatches. Given a read with a phred-scaled probability q of being generated from the mapped position, the new mapping quality is about sqrt((INT-q)/INT)*INT. A zero value disables this functionality; if enabled, the recommended value for BWA is 50.",
                                    "id": "#parameter_for_adjusting_mapq",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "label": "Generate uncompress BCF/VCF output",
                                    "inputBinding": {
                                        "prefix": "-u",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Similar to -g except that the output is uncompressed BCF, which is preferred for piping.",
                                    "id": "#output_uncompressed_bcf_or_vcf",
                                    "sbg:category": "Output options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "label": "Output mapping quality",
                                    "inputBinding": {
                                        "prefix": "-s",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Output mapping quality (disabled by -g/-u).",
                                    "id": "#output_mapping_quality",
                                    "sbg:category": "Output options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "description": "Output file format.",
                                    "id": "#output_format",
                                    "label": "Output file format",
                                    "sbg:category": "configuration",
                                    "type": [
                                        {
                                            "name": "output_format",
                                            "symbols": [
                                                "PILEUP",
                                                "BCF",
                                                "VCF"
                                            ],
                                            "type": "enum"
                                        }
                                    ]
                                },
                                {
                                    "label": "Output base positions on reads",
                                    "inputBinding": {
                                        "prefix": "-O",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Output base positions on reads (disabled by -g/-u).",
                                    "id": "#output_base_positions_on_reads",
                                    "sbg:category": "Output options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "label": "Do not perform indel calling",
                                    "inputBinding": {
                                        "prefix": "-I",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Do not perform indel calling.",
                                    "id": "#no_indel_calling",
                                    "sbg:category": "SNP/INDEL genotype likelihoods options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "label": "Comma-separated list of FORMAT and INFO tags to output",
                                    "inputBinding": {
                                        "prefix": "-t",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "itemSeparator": null,
                                        "separate": true
                                    },
                                    "description": "Comma-separated list of FORMAT and INFO tags (DP,AD,ADF,ADR,SP,INFO/AD,INFO/ADF,INFO/ADR []\") to output.",
                                    "id": "#more_info_to_output",
                                    "sbg:category": "configuration",
                                    "type": [
                                        "null",
                                        "string"
                                    ]
                                },
                                {
                                    "label": "Minimum fraction of gapped reads for candidates",
                                    "inputBinding": {
                                        "prefix": "-F",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "sbg:toolDefaultValue": "0.002",
                                    "description": "Minimum fraction of gapped reads for candidates.",
                                    "id": "#minimum_fraction_of_gapped_reads",
                                    "sbg:category": "SNP/INDEL genotype likelihoods options",
                                    "type": [
                                        "null",
                                        "float"
                                    ]
                                },
                                {
                                    "label": "Minimum gapped reads for indel candidates",
                                    "inputBinding": {
                                        "prefix": "-m",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "sbg:toolDefaultValue": "1",
                                    "description": "Minimum gapped reads for indel candidates.",
                                    "id": "#min_gapped_reads_for_indel",
                                    "sbg:category": "SNP/INDEL genotype likelihoods options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "label": "Max per-BAM depth",
                                    "inputBinding": {
                                        "prefix": "-d",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "sbg:toolDefaultValue": "250",
                                    "description": "At a position, read maximally INT reads per input BAM.",
                                    "id": "#max_per_bam_depth",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "label": "Skip INDEL calling if the average per-sample depth is above",
                                    "type": [
                                        "null",
                                        "int"
                                    ],
                                    "inputBinding": {
                                        "prefix": "-L",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "sbg:toolDefaultValue": "250",
                                    "schema": [
                                        "null",
                                        "int"
                                    ],
                                    "id": "#max_idepth",
                                    "sbg:category": "configuration",
                                    "description": "Skip INDEL calling if the average per-sample depth is above."
                                },
                                {
                                    "label": "Skip alignments with mapQ  smaller than",
                                    "inputBinding": {
                                        "prefix": "-q",
                                        "sbg:cmdInclude": true,
                                        "separate": true,
                                        "separator": " ",
                                        "position": 0
                                    },
                                    "description": "Minimum mapping quality for an alignment to be used.",
                                    "id": "#mapq_threshold",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "sbg:fileTypes": "BAM",
                                    "sbg:stageInput": "link",
                                    "type": [
                                        {
                                            "name": "input_bam_files",
                                            "type": "array",
                                            "items": "File"
                                        }
                                    ],
                                    "required": true,
                                    "label": "List of input BAM files",
                                    "inputBinding": {
                                        "separator": " ",
                                        "position": 1,
                                        "sbg:cmdInclude": true,
                                        "streamable": false,
                                        "itemSeparator": null,
                                        "separate": true
                                    },
                                    "description": "List of input BAM files.",
                                    "id": "#input_bam_files",
                                    "sbg:category": "Input options",
                                    "schema": [
                                        "File"
                                    ]
                                },
                                {
                                    "label": "Ignore RG tags",
                                    "inputBinding": {
                                        "prefix": "-R",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Ignore rg tags.",
                                    "id": "#ignore_rg_tags",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "label": "Disable read-pair overlap detection",
                                    "inputBinding": {
                                        "prefix": "-x",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Disable read-pair overlap detection.",
                                    "id": "#ignore_overlaps",
                                    "sbg:category": "configuration",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "label": "Coefficient for homopolymer errors",
                                    "inputBinding": {
                                        "prefix": "-h",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "sbg:toolDefaultValue": "100",
                                    "description": "Coefficient for modeling homopolymer errors. Given an l-long homopolymer run, the sequencing error of an indel of size s is modeled as INT*s/l.",
                                    "id": "#homopolymer_err_coeficient",
                                    "sbg:category": "SNP/INDEL genotype likelihoods options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "label": "Phred-scaled gap open sequencing error probability",
                                    "inputBinding": {
                                        "prefix": "-o",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "sbg:toolDefaultValue": "40",
                                    "description": "Phred-scaled gap open sequencing error probability. Reducing INT leads to more indel calls.",
                                    "id": "#gap_open_sequencing_error_probability",
                                    "sbg:category": "SNP/INDEL genotype likelihoods options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "label": "Phred-scaled gap extension seq error probability",
                                    "inputBinding": {
                                        "prefix": "-e",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "sbg:toolDefaultValue": "20",
                                    "description": "Phred-scaled gap extension sequencing error probability. Reducing INT leads to longer indels.",
                                    "id": "#gap_extension_seq_error_probability",
                                    "sbg:category": "SNP/INDEL genotype likelihoods options",
                                    "type": [
                                        "null",
                                        "int"
                                    ]
                                },
                                {
                                    "sbg:stageInput": null,
                                    "label": "Filter zero coverage lines.",
                                    "description": "Filter zero coverage lines from pileup files. This option is valid only when output pileup files.",
                                    "id": "#filter_zero_coverage_lines",
                                    "sbg:category": "output configuration",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "sbg:stageInput": null,
                                    "type": [
                                        "null",
                                        "string"
                                    ],
                                    "inputBinding": {
                                        "prefix": "--ff",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Filter flags: skip reads with mask bits set [UNMAP,SECONDARY,QCFAIL,DUP].",
                                    "id": "#filter_flags",
                                    "sbg:category": "Input options",
                                    "label": "Filter flags"
                                },
                                {
                                    "sbg:fileTypes": "TXT",
                                    "label": "Exclude read groups listed in file",
                                    "required": false,
                                    "inputBinding": {
                                        "prefix": "-G",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Exclude read groups listed in file.",
                                    "id": "#exclude_read_groups_list_file",
                                    "sbg:category": "File input",
                                    "type": [
                                        "null",
                                        "File"
                                    ]
                                },
                                {
                                    "label": "Disable BAQ computation",
                                    "inputBinding": {
                                        "prefix": "-B",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Disable probabilistic realignment for the computation of base alignment quality (BAQ). BAQ is the Phred-scaled probability of a read base being misaligned. Applying this option greatly helps to reduce false SNPs caused by misalignments.",
                                    "id": "#disable_baq_computation",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "label": "Count anomalous read pairs",
                                    "inputBinding": {
                                        "prefix": "-A",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Do not skip anomalous read pairs in variant calling.",
                                    "id": "#count_anomalous_read_pairs",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                },
                                {
                                    "label": "Comma separated list of platforms for indels",
                                    "inputBinding": {
                                        "prefix": "-P",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "sbg:toolDefaultValue": "\"all\"",
                                    "description": "Comma dilimited list of platforms (determined by @RG-PL) from which indel candidates are obtained. It is recommended to collect indel candidates from sequencing technologies that have low indel error rate such as ILLUMINA.",
                                    "id": "#comma_separated_list_of_platforms_for_indels",
                                    "sbg:category": "SNP/INDEL genotype likelihoods options",
                                    "type": [
                                        "null",
                                        "string"
                                    ]
                                },
                                {
                                    "sbg:fileTypes": "BED",
                                    "label": "List of positions (chr pos) or regions in BED file",
                                    "required": false,
                                    "inputBinding": {
                                        "prefix": "-l",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "BED or position list file containing a list of regions or sites where pileup or BCF should be generated.",
                                    "id": "#chr_pos_list_file",
                                    "sbg:category": "File input",
                                    "type": [
                                        "null",
                                        "File"
                                    ]
                                },
                                {
                                    "label": "Assume the quality is in the Illumina-1.3+ encoding",
                                    "inputBinding": {
                                        "prefix": "-6",
                                        "sbg:cmdInclude": true,
                                        "position": 0,
                                        "separate": true
                                    },
                                    "description": "Assume the quality is in the Illumina-1.3+ encoding.",
                                    "id": "#assume_the_quality_Illumina_encoding",
                                    "sbg:category": "Input options",
                                    "type": [
                                        "null",
                                        "boolean"
                                    ]
                                }
                            ],
                            "hints": [
                                {
                                    "dockerImageId": "2fb927277493",
                                    "class": "DockerRequirement",
                                    "dockerPull": "images.sbgenomics.com/marouf/samtools:1.3"
                                },
                                {
                                    "value": 1,
                                    "class": "sbg:CPURequirement"
                                },
                                {
                                    "value": 1000,
                                    "class": "sbg:MemRequirement"
                                }
                            ],
                            "sbg:revision": 19,
                            "baseCommand": [
                                "/opt/samtools-1.3/samtools",
                                "mpileup"
                            ],
                            "id": "marouf/samtools-1-3-demo/mpileup-1-3/19",
                            "sbg:latestRevision": 19,
                            "temporaryFailCodes": [],
                            "successCodes": [],
                            "sbg:toolAuthor": "Heng Li, Sanger Institute",
                            "sbg:revisionNotes": "Added inherit metadata from input.",
                            "arguments": [
                                {
                                    "separate": true,
                                    "valueFrom": {
                                        "script": "{\n\n if($job.inputs.output_format)\n {\n    if($job.inputs.output_format == \"PILEUP\")\n        return\n    else if($job.inputs.output_format == \"VCF\")\n    \treturn \"-v\"\n    else if($job.inputs.output_format == \"BCF\")\n    \treturn \"-g\"\n    else\n      return\n  \n \n }\n\n\n\n}",
                                        "engine": "#cwl-js-engine",
                                        "class": "Expression"
                                    }
                                },
                                {
                                    "position": 1000,
                                    "valueFrom": {
                                        "script": "{\n\n // Check if input files are delivered in an array or not\n if($job.inputs.input_bam_files.constructor == Array)\n {\n   // Select the first entry of the array\n\u2002\u2002\u2002\u2002filepath = \"/\"+$job.inputs.input_bam_files[0].path \n\u2002\u2002} \n  \n  else\n    \n  {\n    // Use the file name directly\n\u2002\u2002\u2002\u2002filepath = \"/\"+$job.inputs.input_bam_files.path\n \u2002}\n   \n \n filename = filepath.split(\"/\").pop()\n new_filename = filename.substr(0,filename.lastIndexOf(\".\"))\n    \n \n if ($job.inputs.chr_pos_list_file)\n   new_filename = new_filename+'_'+ $job.inputs.chr_pos_list_file.path.split('/')[$job.inputs.chr_pos_list_file.path.split('/').length-1].split('.')[0]\n\n \n extension = '.pileup'\n  \n if ( (typeof $job.inputs.output_format !== \"undefined\" && $job.inputs.output_format === \"BCF\") )\n   extension = '.bcf'\n             \n else if ((typeof $job.inputs.output_format !== \"undefined\" && $job.inputs.output_format === \"VCF\") )  \n   extension = '.vcf' \n  \n if($job.inputs.filter_zero_coverage_lines && extension== '.pileup')   \n \treturn  \" | awk '{if($4 != 0) print $0}'  > \"+  new_filename + extension\n else\n   return  \" > \"+  new_filename + extension\n}",
                                        "engine": "#cwl-js-engine",
                                        "class": "Expression"
                                    },
                                    "separate": true
                                }
                            ],
                            "sbg:id": "admin/sbg-public-data/mpileup-1-3/19",
                            "sbg:job": {
                                "allocatedResources": {
                                    "cpu": 1,
                                    "mem": 1000
                                },
                                "inputs": {
                                    "filter_zero_coverage_lines": true,
                                    "output_uncompressed_bcf_or_vcf": true,
                                    "region_pileup_generation": "region_pileup_generation",
                                    "max_idepth": 11,
                                    "recalc_BAQ_fly": true,
                                    "no_indel_calling": true,
                                    "per_sample_mF": true,
                                    "homopolymer_err_coeficient": 11,
                                    "required_flags": 0,
                                    "ignore_rg_tags": true,
                                    "minimum_fraction_of_gapped_reads": 11,
                                    "reference_fasta_file": {
                                        "path": "input/input.fasta",
                                        "secondaryFiles": [
                                            {
                                                "path": ".fai"
                                            }
                                        ]
                                    },
                                    "exclude_read_groups_list_file": {
                                        "path": "exclude_read_groups_list_file.txt",
                                        "size": 0,
                                        "class": "File",
                                        "secondaryFiles": []
                                    },
                                    "output_mapping_quality": true,
                                    "assume_the_quality_Illumina_encoding": true,
                                    "parameter_for_adjusting_mapq": 11,
                                    "min_gapped_reads_for_indel": 11,
                                    "gap_open_sequencing_error_probability": 11,
                                    "input_bam_files": [
                                        {
                                            "path": "input_bam_file1.bam",
                                            "size": 0,
                                            "class": "File",
                                            "secondaryFiles": []
                                        },
                                        {
                                            "path": "input_bam_file2.bam"
                                        }
                                    ],
                                    "output_format": "PILEUP",
                                    "ignore_overlaps": true,
                                    "disable_baq_computation": true,
                                    "count_anomalous_read_pairs": true,
                                    "gap_extension_seq_error_probability": 11,
                                    "more_info_to_output": "more_info_to_output-string-value",
                                    "max_per_bam_depth": 11,
                                    "skip_bases_with_baq_smaller_than_defined": 11,
                                    "comma_separated_list_of_platforms_for_indels": "all",
                                    "filter_flags": "filter_flags-string-value",
                                    "mapq_threshold": 11,
                                    "output_base_positions_on_reads": true
                                }
                            },
                            "stdin": "",
                            "sbg:project": "marouf/samtools-1-3-demo",
                            "description": "SAMtools Mpileup generates BCF or PILEUP for one or multiple BAM files. Alignment records are grouped by sample identifiers in @RG header lines. If sample identifiers are absent, each input file is regarded as one sample.\n\nIn the pileup format (without -uor-g), each line represents a genomic position consisting of chromosome name, coordinate, reference base, read bases, read qualities, and alignment mapping qualities. Information on match, mismatch, indel, strand, mapping quality, and the start and end of a read are all encoded at the read base column. In this column, a dot stands for a match to the reference base on the forward strand, a comma for a match on the reverse strand, a '>' or '<' for a reference skip, 'ACGTN' for a mismatch on the forward strand, and 'acgtn' for a mismatch on the reverse strand. A pattern `\\\\+[0-9]+[ACGTNacgtn]+' indicates there is an insertion between this reference position and the next reference position. The length of the insertion is given by the integer in the pattern followed by the inserted sequence. Similarly, a pattern '-[0-9]+[ACGTNacgtn]+' represents a deletion from the reference. The deleted bases will be presented as `*' in the following lines. Also at the read base column, a symbol '^', marks the start of a read. The ASCII of the character following '^' minus 33 gives the mapping quality. The symbol '$' marks the end of a read segment.\n\n#### Common Issue:\n- Please use the public pileup parallel pipeline for large input files.\n- Please sort BAM files by coordinates before using mpileup.",
                            "x": 924,
                            "sbg:createdOn": 1458812970,
                            "outputs": [
                                {
                                    "type": [
                                        "null",
                                        "File"
                                    ],
                                    "label": "Output PILEUP, VCF, or BCF file",
                                    "outputBinding": {
                                        "glob": "{*.pileup,*.bcf,*.vcf}",
                                        "sbg:metadata": {
                                            "ScatteredUsing": {
                                                "script": "{\n\tif ($job.inputs.chr_pos_list_file)\n    {\n\t    len =  $job.inputs.chr_pos_list_file.path.split('/').length\n        \n        return $job.inputs.chr_pos_list_file.path.split('/')[len-1].split('.')[0]\n    }\n  \telse\n    {\n\t    return \n    }\n}",
                                                "engine": "#cwl-js-engine",
                                                "class": "Expression"
                                            }
                                        },
                                        "streamable": false,
                                        "sbg:inheritMetadataFrom": "#input_bam_files"
                                    },
                                    "sbg:fileTypes": "PILEUP,BCF,VCF",
                                    "id": "#output_pileup_vcf_or_bcf_file",
                                    "description": "Output PILEUP, VCF, or BCF file."
                                }
                            ],
                            "sbg:links": [
                                {
                                    "label": "Homepage",
                                    "id": "http://samtools.sourceforge.net/"
                                },
                                {
                                    "label": "Source code",
                                    "id": "https://github.com/samtools/samtools"
                                },
                                {
                                    "label": "Wiki",
                                    "id": "http://sourceforge.net/p/samtools/wiki/Home/"
                                },
                                {
                                    "label": "Download",
                                    "id": "http://sourceforge.net/projects/samtools/files/"
                                },
                                {
                                    "label": "Publication",
                                    "id": "http://www.ncbi.nlm.nih.gov/pubmed/19505943"
                                },
                                {
                                    "label": "Documentation",
                                    "id": "http://www.htslib.org/doc/samtools.html"
                                }
                            ],
                            "sbg:contributors": [
                                "milan.domazet.sudo",
                                "marouf"
                            ],
                            "sbg:modifiedOn": 1482334695,
                            "sbg:sbgMaintained": false,
                            "stdout": ""
                        },
                        "id": "#SAMtools_Mpileup",
                        "sbg:y": 361
                    },
                    {
                        "sbg:x": 782,
                        "outputs": [
                            {
                                "id": "#SBG_Prepare_Intervals.names"
                            },
                            {
                                "id": "#SBG_Prepare_Intervals.intervals"
                            }
                        ],
                        "run": {
                            "sbg:toolkit": "SBGTools",
                            "requirements": [
                                {
                                    "fileDef": [
                                        {
                                            "filename": "sbg_prepare_intervals.py",
                                            "fileContent": "\"\"\"\nUsage:\n    sbg_prepare_intervals.py [options] [--fastq FILE --bed FILE --mode INT --format STR --others STR]\n\nDescription:\n    Purpose of this tool is to split BED file into files based on the selected mode.\n    If bed file is not provided fai(fasta index) file is converted to bed.\n\nOptions:\n\n    -h, --help            Show this message.\n\n    -v, -V, --version     Tool version.\n\n    -b, -B, --bed FILE    Path to input bed file.\n\n    --fai FILE            Path to input fai file.\n\n    --format STR          Output file format.\n\n    --mode INT            Select input mode.\n\n\"\"\"\n\n\nfrom docopt import docopt\nimport os\nimport shutil\nimport glob\n\ndefault_extension = '.bed'  # for output files\n\n\n\ndef create_file(contents, contig_name, extension=default_extension):\n    \"\"\"function for creating a file for all intervals in a contig\"\"\"\n\n    new_file = open(\"Intervals/\" + contig_name + extension, \"w\")\n    new_file.write(contents)\n    new_file.close()\n\n\ndef add_to_file(line, name, extension=default_extension):\n    \"\"\"function for adding a line to a file\"\"\"\n\n    new_file = open(\"Intervals/\" + name + extension, \"a\")\n    if lformat == formats[1]:\n        sep = line.split(\"\\t\")\n        line = sep[0] + \":\" + sep[1] + \"-\" + sep[2]\n    new_file.write(line)\n    new_file.close()\n\n\ndef fai2bed(fai):\n    \"\"\"function to create a bed file from fai file\"\"\"\n\n    region_thr = 10000000  # threshold used to determine starting point accounting for telomeres in chromosomes\n    if not fai.rfind(\".fasta.fai\") == -1:\n        basename = fai[0:fai.rfind(\".fasta.fai\")]\n    else:\n        basename = fai[0:fai.rfind(\".\")]\n    with open(fai, \"r\") as ins:\n        new_array = []\n        for line in ins:\n            len_reg = int(line.split()[1])\n            cutoff = 0 if (len_reg < region_thr) else 0  # sd\\\\telomeres or start with 1\n            new_line = line.split()[0] + '\\t' + str(cutoff) + '\\t' + str(len_reg + cutoff)\n            new_array.append(new_line)\n    new_file = open(basename + \".bed\", \"w\")\n    new_file.write(\"\\n\".join(new_array))\n    return basename + \".bed\"\n\ndef chr_intervals(no_of_chrms = 23):\n    \"\"\"returns all possible designations for chromosome intervals\"\"\"\n    \n    chrms = []\n    for i in range(1, no_of_chrms):\n        chrms.append(\"chr\" + str(i))\n        chrms.append(str(i))\n    chrms.extend([\"x\", \"y\", \"chrx\", \"chry\"])\n    return chrms\n\n\ndef mode_1(orig_file):\n    \"\"\"mode 1: every line is a new file\"\"\"\n\n    with open(orig_file, \"r\") as ins:\n        prev = \"\"\n        counter = 0\n        names = []\n        for line in ins:\n            if line.split()[0] == prev:\n                counter += 1\n            else:\n                counter = 0\n            suffix = \"\" if (counter == 0) else \"_\" + str(counter)\n            create_file(line, line.split()[0] + suffix)\n            names.append(line.split()[0] + suffix)\n            prev = line.split()[0]\n\n        create_file(str(names), \"names\", extension=\".txt\")\n\ndef mode_2(orig_file, others_name):\n    \"\"\"mode 2: separate file is created for each chromosome, and one file is created for other intervals\"\"\"\n\n    chrms = chr_intervals()\n    names = []\n\n    with open(orig_file, 'r') as ins:\n        for line in ins:\n            name = line.split()[0]\n            if name.lower() in chrms:\n                name = name.lower()\n            else:\n                name = others_name\n            try:\n                add_to_file(line, name)\n                if not name in names:\n                    names.append(name)\n            except:\n                raise Exception(\"Couldn't create or write in the file in mode 2\")\n\n        create_file(str(names), \"names\", extension = \".txt\")\n\n\ndef mode_3(orig_file, extension=default_extension):\n    \"\"\"mode 3: input file is staged to output\"\"\"\n\n    orig_name = orig_file.split(\"/\")[len(orig_file.split(\"/\")) - 1]\n    output_file = r\"./Intervals/\" + orig_name[0:orig_name.rfind('.')] + extension\n\n    shutil.copyfile(orig_file, output_file)\n\n    names = [orig_name[0:orig_name.rfind('.')]]\n    create_file(str(names), \"names\", extension=\".txt\")\n\n\ndef mode_4(orig_file, others_name):\n    \"\"\"mode 4: every interval in chromosomes is in a separate file. Other intervals are in a single file\"\"\"\n\n    chrms = chr_intervals()\n    names = []\n\n    with open(orig_file, \"r\") as ins:\n        counter = {}\n        for line in ins:\n            name = line.split()[0].lower()\n            if name in chrms:\n                if name in counter:\n                    counter[name] += 1\n                else:\n                    counter[name] = 0\n                suffix = \"\" if (counter[name] == 0) else \"_\" + str(counter[name])\n                create_file(line, name + suffix)\n                names.append(name + suffix)\n                prev = name\n            else:\n                name = others_name\n                if not name in names:\n                    names.append(name)\n                try:\n                    add_to_file(line, name)\n                except:\n                    raise Exception(\"Couldn't create or write in the file in mode 4\")\n\n        create_file(str(names), \"names\", extension=\".txt\")\n\n\ndef prepare_intervals():\n    # reading input files and split mode from command line\n    args = docopt(__doc__, version='1.0')\n\n    bed_file = args['--bed']\n    fai_file = args['--fai']\n    split_mode = int(args['--mode'])\n\n    \n    # define file name for non-chromosomal contigs\n    others_name = 'others' \n\n    global formats, lformat\n    formats = [\"chr start end\", \"chr:start-end\"]\n    lformat = args['--format']\n    if lformat == None:\n        lformat = formats[0]\n    if not lformat in formats:\n        raise Exception('Unsuported interval format')\n\n    if not os.path.exists(r\"./Intervals\"):\n        os.mkdir(r\"./Intervals\")\n    else:\n        files = glob.glob(r\"./Intervals/*\")\n        for f in files:\n            os.remove(f)\n\n    # create variable input_file taking bed_file as priority\n    if bed_file:\n        input_file = bed_file\n    elif fai_file:\n        input_file = fai2bed(fai_file)\n    else:\n        raise Exception('No input files are provided')\n\n    # calling adequate split mode function\n    if split_mode == 1:\n        mode_1(input_file)\n    elif split_mode == 2:\n        mode_2(input_file, others_name)\n    elif split_mode == 3:\n        if bed_file:\n            mode_3(input_file)\n        else:\n            raise Exception('Bed file is required for mode 3')\n    elif split_mode == 4:\n        mode_4(input_file, others_name)\n    else:\n        raise Exception('Split mode value is not set')\n\n\nif __name__ == '__main__':\n    prepare_intervals()"
                                        }
                                    ],
                                    "class": "CreateFileRequirement"
                                },
                                {
                                    "requirements": [
                                        {
                                            "class": "DockerRequirement",
                                            "dockerPull": "rabix/js-engine"
                                        }
                                    ],
                                    "id": "#cwl-js-engine",
                                    "class": "ExpressionEngineRequirement"
                                }
                            ],
                            "sbg:validationErrors": [],
                            "sbg:categories": [
                                "Converters"
                            ],
                            "label": "SBG Prepare Intervals",
                            "sbg:createdBy": "vladimirk",
                            "temporaryFailCodes": [],
                            "sbg:image_url": null,
                            "cwlVersion": "sbg:draft-2",
                            "sbg:revisionsInfo": [
                                {
                                    "sbg:revision": 0,
                                    "sbg:revisionNotes": "Copy of medjo/sbg-prepare-intervals/sbg-prepare-intervals/75",
                                    "sbg:modifiedBy": "vladimirk",
                                    "sbg:modifiedOn": 1473083821
                                },
                                {
                                    "sbg:revision": 1,
                                    "sbg:revisionNotes": "Copy of medjo/sbg-prepare-intervals/sbg-prepare-intervals/76",
                                    "sbg:modifiedBy": "bogdang",
                                    "sbg:modifiedOn": 1473084447
                                },
                                {
                                    "sbg:revision": 2,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "medjo",
                                    "sbg:modifiedOn": 1473928444
                                },
                                {
                                    "sbg:revision": 3,
                                    "sbg:revisionNotes": "split_mode set to required",
                                    "sbg:modifiedBy": "medjo",
                                    "sbg:modifiedOn": 1474970272
                                },
                                {
                                    "sbg:revision": 4,
                                    "sbg:revisionNotes": "Fixed Toolkit name.",
                                    "sbg:modifiedBy": "bix-demo",
                                    "sbg:modifiedOn": 1478525360
                                }
                            ],
                            "y": 494,
                            "sbg:modifiedOn": 1478525360,
                            "sbg:license": "Apache License 2.0",
                            "sbg:modifiedBy": "bix-demo",
                            "class": "CommandLineTool",
                            "inputs": [
                                {
                                    "label": "Split mode",
                                    "inputBinding": {
                                        "prefix": "--mode",
                                        "sbg:cmdInclude": true,
                                        "position": 3,
                                        "valueFrom": {
                                            "script": "{\n  mode = $job.inputs.split_mode\n  switch (mode) \n  {\n    case \"File per interval\": \n      return 1\n    case \"File per chr with alt contig in a single file\": \n      return 2\n    case \"Output original BED\": \n      return 3\n    case \"File per interval with alt contig in a single file\": \n      return 4  \n  }\n  return 3\n}",
                                            "engine": "#cwl-js-engine",
                                            "class": "Expression"
                                        },
                                        "separate": true
                                    },
                                    "description": "Depending on selected Split Mode value, output files are generated in accordance with description below:  1. Whole Genome - The tool creates one interval file per line of the input BED(FAI) file. Each interval file contains a single line (one of the lines of BED(FAI) input file).  2. Whole Genome with reduced number of jobs - For each contig(chromosome) a single file is created containing all the intervals corresponding to it . All the intervals (lines) other than (chr1, chr2 ... chrY or 1, 2 ... Y) are saved as (\"others.bed\").  3. Whole Exome - BED file is required for execution of this mode. If mode  3 is applied input is passed to the output.  4. Whole Exome Parallel - BED file is required for execution of this mode. For each contig(chromosome) a single file is created containing all the intervals corresponding to it. All the intervals (lines) other than (chr1, chr2 ... chrY or 1, 2 ... Y) are saved as (\"others.bed\"). If split mode is not selected, the tool will output original BED file.",
                                    "id": "#split_mode",
                                    "sbg:category": "Input",
                                    "type": [
                                        {
                                            "name": "split_mode",
                                            "symbols": [
                                                "File per interval",
                                                "File per chr with alt contig in a single file",
                                                "Output original BED",
                                                "File per interval with alt contig in a single file"
                                            ],
                                            "type": "enum"
                                        }
                                    ]
                                },
                                {
                                    "description": "Format of the intervals in the generated files.",
                                    "id": "#format",
                                    "label": "Interval format",
                                    "sbg:category": "Input",
                                    "type": [
                                        "null",
                                        {
                                            "name": "format",
                                            "symbols": [
                                                "chr start end",
                                                "chr:start-end"
                                            ],
                                            "type": "enum"
                                        }
                                    ]
                                },
                                {
                                    "sbg:fileTypes": "FAI",
                                    "label": "Input FAI file",
                                    "required": false,
                                    "inputBinding": {
                                        "prefix": "--fai",
                                        "sbg:cmdInclude": true,
                                        "position": 2,
                                        "separate": true
                                    },
                                    "description": "FAI file is converted to BED format if BED file is not provided.",
                                    "id": "#fai_file",
                                    "sbg:category": "File Input",
                                    "type": [
                                        "null",
                                        "File"
                                    ]
                                },
                                {
                                    "sbg:fileTypes": "BED",
                                    "sbg:stageInput": "link",
                                    "type": [
                                        "null",
                                        "File"
                                    ],
                                    "required": false,
                                    "inputBinding": {
                                        "prefix": "--bed",
                                        "sbg:cmdInclude": true,
                                        "position": 1,
                                        "separate": true
                                    },
                                    "description": "Input BED file containing intervals. Required for modes 3 and 4.",
                                    "id": "#bed_file",
                                    "sbg:category": "File Input",
                                    "label": "Input BED file"
                                }
                            ],
                            "hints": [
                                {
                                    "value": 1000,
                                    "class": "sbg:MemRequirement"
                                },
                                {
                                    "dockerImageId": "",
                                    "class": "DockerRequirement",
                                    "dockerPull": "images.sbgenomics.com/bogdang/sbg_prepare_intervals:1.0"
                                },
                                {
                                    "value": 1,
                                    "class": "sbg:CPURequirement"
                                }
                            ],
                            "sbg:revision": 4,
                            "baseCommand": [
                                "python",
                                "sbg_prepare_intervals.py"
                            ],
                            "id": "bix-demo/sbgtools-demo/sbg-prepare-intervals/4",
                            "sbg:latestRevision": 4,
                            "outputs": [
                                {
                                    "type": [
                                        "null",
                                        "string"
                                    ],
                                    "label": "Output file names",
                                    "outputBinding": {
                                        "glob": "Intervals/names.txt",
                                        "outputEval": {
                                            "script": "{   \n content =  $self[0].contents.replace(/\\0/g, '')\n content = content.replace('[','')\n content = content.replace(']','')\n content = content.replace(/\\'/g, \"\")\n content = content.replace(/\\s/g, '')\n content_arr = content.split(\",\")\n\n return content_arr\n \n\n} ",
                                            "engine": "#cwl-js-engine",
                                            "class": "Expression"
                                        },
                                        "loadContents": true
                                    },
                                    "sbg:fileTypes": "TXT",
                                    "id": "#names",
                                    "description": "File containing the names of created files."
                                },
                                {
                                    "type": [
                                        "null",
                                        {
                                            "name": "intervals",
                                            "type": "array",
                                            "items": "File"
                                        }
                                    ],
                                    "label": "Intervals",
                                    "outputBinding": {
                                        "glob": "Intervals/*.bed",
                                        "sbg:metadata": {
                                            "sbg_scatter": "true"
                                        }
                                    },
                                    "sbg:fileTypes": "BED",
                                    "id": "#intervals",
                                    "description": "Array of BED files genereted as per selected Split Mode."
                                }
                            ],
                            "sbg:toolkitVersion": "1.0",
                            "sbg:toolAuthor": "Seven Bridges Genomics",
                            "successCodes": [],
                            "sbg:revisionNotes": "Fixed Toolkit name.",
                            "arguments": [
                                {
                                    "separate": true,
                                    "valueFrom": {
                                        "script": "{\t\n  if (typeof($job.inputs.format) !== \"undefined\")\n  \treturn \"--format \" + \"\\\"\" + $job.inputs.format + \"\\\"\"\n}",
                                        "engine": "#cwl-js-engine",
                                        "class": "Expression"
                                    }
                                }
                            ],
                            "sbg:id": "admin/sbg-public-data/sbg-prepare-intervals/82",
                            "sbg:job": {
                                "allocatedResources": {
                                    "cpu": 1,
                                    "mem": 1000
                                },
                                "inputs": {
                                    "format": "chr start end",
                                    "fai_file": {
                                        "path": "/path/to/fai_file.ext",
                                        "size": 0,
                                        "class": "File",
                                        "secondaryFiles": []
                                    },
                                    "bed_file": {
                                        "path": "/path/to/bed_file.ext",
                                        "size": 0,
                                        "class": "File",
                                        "secondaryFiles": []
                                    },
                                    "split_mode": null
                                }
                            },
                            "stdin": "",
                            "sbg:project": "bix-demo/sbgtools-demo",
                            "description": "Depending on selected Split Mode value, output files are generated in accordance with description below:\n\n1. Whole Genome - The tool creates one interval file per line of the input BED(FAI) file.\nEach interval file contains a single line (one of the lines of BED(FAI) input file).\n\n2. Whole Genome with reduced number of jobs - For each contig(chromosome) a single file\nis created containing all the intervals corresponding to it .\nAll the intervals (lines) other than (chr1, chr2 ... chrY or 1, 2 ... Y) are saved as\n(\"others.bed\").\n\n3. Whole Exome - BED file is required for execution of this mode. If mode  3 is applied input is passed to the output.\n\n4. Whole Genome Hybrid - For each chromosome a single file is created for each interval.\nAll the intervals (lines) other than (chr1, chr2 ... chrY or 1, 2 ... Y) are saved as\n(\"others.bed\").",
                            "x": 782,
                            "sbg:createdOn": 1473083821,
                            "sbg:cmdPreview": "python sbg_prepare_intervals.py  --format \"chr start end\" --mode 3",
                            "sbg:contributors": [
                                "bix-demo",
                                "medjo",
                                "bogdang",
                                "vladimirk"
                            ],
                            "sbg:sbgMaintained": false,
                            "stdout": ""
                        },
                        "inputs": [
                            {
                                "default": "File per chr with alt contig in a single file",
                                "id": "#SBG_Prepare_Intervals.split_mode"
                            },
                            {
                                "default": "chr start end",
                                "id": "#SBG_Prepare_Intervals.format"
                            },
                            {
                                "id": "#SBG_Prepare_Intervals.fai_file",
                                "source": [
                                    "#fai_file"
                                ]
                            },
                            {
                                "id": "#SBG_Prepare_Intervals.bed_file"
                            }
                        ],
                        "id": "#SBG_Prepare_Intervals",
                        "sbg:y": 494
                    },
                    {
                        "sbg:x": 1109,
                        "outputs": [
                            {
                                "id": "#SBG_SAMtools_Merge_Mpileup.output_file"
                            }
                        ],
                        "run": {
                            "sbg:toolkit": "SBGTools",
                            "requirements": [
                                {
                                    "fileDef": [
                                        {
                                            "filename": "merge.py",
                                            "fileContent": "\"\"\"\nUsage: merge.py     --file-list FILE --output-format FRMT  \n\nDescription:\n  Main function of this tool is to provide option to the user weather to merge list of files from SAMtools Mpileup\n  tool or to pass list of files for each contig to the output. If --Merge option is selected, regardless of the type,\n  files will be merged following the contig order from the BAM header.\n\nOptions:\n\n    --help                              This message.\n\n    --version                           Tool version.\n    \n    --output-format FRMT                Output_format.\n\n    --file-list FILE                    List of contigs in BCF or PILEUP format produced by SAMtools Mpileup tool.\n\n\"\"\"\n\nimport docopt\nfrom subprocess import Popen\n\nargs = docopt.docopt(__doc__, version=\"1.0\") \n\nBCFTOOLS_ROOT = \"bcftools\"\n\nsorted_file_list = list()\nfile_list = args['--file-list'].split(',')\n\noutput_format =  args['--output-format']\n\n\nfor fp in file_list:\n    sorted_file_list.append(fp)\n\n\n\nif output_format ==\"pileup\":\n    with open(sorted_file_list[0].split('/')[-1].replace('.pileup', '.merged.pileup'), 'w')  as rr:\n        cmd = ['cat'] + sorted_file_list\n        p = Popen(cmd, stdout=rr)\n        p.wait()\nelif output_format ==\"bcf\":\n    with open(sorted_file_list[0].split('/')[-1].replace('.bcf', '.merged.bcf'), 'w')  as rr:\n        cmd = [BCFTOOLS_ROOT, 'concat'] + sorted_file_list\n        p = Popen(cmd, stdout=rr)\n        p.wait()\nelif output_format ==\"vcf\":\n    with open(sorted_file_list[0].split('/')[-1].replace('.vcf', '.merged.vcf'), 'w')  as rr:\n        cmd = [BCFTOOLS_ROOT, 'concat'] + sorted_file_list\n        p = Popen(cmd, stdout=rr)\n        p.wait()"
                                        },
                                        {
                                            "filename": "delete_empty.py",
                                            "fileContent": "\"\"\"\nUsage: delete_empty.py     --file-list FILE  \n\nDescription:\n  Main function of this tool is to remove empty pileup files.\n\nOptions:\n\n    --help                              This message.\n\n    --version                           Tool version.\n\n    --file-list FILE                    List of contigs in BCF or PILEUP format produced by SAMtools Mpileup tool.\n\n\"\"\"\n\nimport os\nimport docopt\nfrom subprocess import Popen\n\nargs = docopt.docopt(__doc__, version=\"1.0\") \n\n\nsorted_file_list = list()\nfile_list = args['--file-list'].split(',')\n\nfor fp in file_list:\n    os.remove(fp)"
                                        }
                                    ],
                                    "class": "CreateFileRequirement"
                                },
                                {
                                    "requirements": [
                                        {
                                            "class": "DockerRequirement",
                                            "dockerPull": "rabix/js-engine"
                                        }
                                    ],
                                    "id": "#cwl-js-engine",
                                    "class": "ExpressionEngineRequirement"
                                }
                            ],
                            "sbg:validationErrors": [],
                            "sbg:categories": [
                                "Text-Processing",
                                "SAM/BAM-Processing"
                            ],
                            "label": "SBG SAMtools Merge Mpileup",
                            "sbg:createdBy": "bix-demo",
                            "temporaryFailCodes": [],
                            "sbg:image_url": null,
                            "sbg:revisionsInfo": [
                                {
                                    "sbg:revision": 0,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "bix-demo",
                                    "sbg:modifiedOn": 1450911287
                                },
                                {
                                    "sbg:revision": 1,
                                    "sbg:revisionNotes": null,
                                    "sbg:modifiedBy": "bix-demo",
                                    "sbg:modifiedOn": 1450911287
                                },
                                {
                                    "sbg:revision": 2,
                                    "sbg:revisionNotes": "import from dev project",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1473929938
                                },
                                {
                                    "sbg:revision": 3,
                                    "sbg:revisionNotes": "add some verification to avoid deleting null in case there are not empty files to delete.",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1474019719
                                },
                                {
                                    "sbg:revision": 4,
                                    "sbg:revisionNotes": "Update the merge.py script and fix bug with the output file name",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1474973977
                                },
                                {
                                    "sbg:revision": 5,
                                    "sbg:revisionNotes": "change the python code in merge.py",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1475052403
                                },
                                {
                                    "sbg:revision": 6,
                                    "sbg:revisionNotes": "change the merge.py",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1475058095
                                },
                                {
                                    "sbg:revision": 7,
                                    "sbg:revisionNotes": "add . to the output file name",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1475083435
                                },
                                {
                                    "sbg:revision": 8,
                                    "sbg:revisionNotes": "update parameters ID, and label",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1475237770
                                },
                                {
                                    "sbg:revision": 9,
                                    "sbg:revisionNotes": "link instead of copy inputs",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1476440477
                                },
                                {
                                    "sbg:revision": 10,
                                    "sbg:revisionNotes": "Added inherit metadata from input.",
                                    "sbg:modifiedBy": "milan.domazet.sudo",
                                    "sbg:modifiedOn": 1482334679
                                },
                                {
                                    "sbg:revision": 11,
                                    "sbg:revisionNotes": "fix metadata issue",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1482417569
                                },
                                {
                                    "sbg:revision": 12,
                                    "sbg:revisionNotes": "fix metadata issue when there is not metadata at input files",
                                    "sbg:modifiedBy": "marouf",
                                    "sbg:modifiedOn": 1482421824
                                }
                            ],
                            "sbg:toolkitVersion": "0.2",
                            "sbg:license": "Apache License 2.0",
                            "sbg:cmdPreview": "",
                            "sbg:modifiedBy": "marouf",
                            "class": "CommandLineTool",
                            "inputs": [
                                {
                                    "type": [
                                        {
                                            "name": "file_list",
                                            "type": "array",
                                            "items": "File"
                                        }
                                    ],
                                    "sbg:stageInput": "link",
                                    "label": "Input file list",
                                    "sbg:fileTypes": "PILEUP,VCF,BCF",
                                    "id": "#file_list",
                                    "description": "Input files for merging."
                                },
                                {
                                    "type": [
                                        {
                                            "name": "contig_order_names",
                                            "type": "array",
                                            "items": "string"
                                        }
                                    ],
                                    "description": "Ordered names of contigs that are used for merging of file list loaded via file_list input.",
                                    "label": "merging_order",
                                    "id": "#contig_order_names",
                                    "sbg:includeInPorts": true
                                },
                                {
                                    "description": "Mode to tell how to merge mpileup files or to pass them through.",
                                    "label": "Output state",
                                    "id": "#output_state",
                                    "type": [
                                        {
                                            "name": "output_state",
                                            "symbols": [
                                                "Merge",
                                                "Pass nonempty files",
                                                "Pass all files"
                                            ],
                                            "type": "enum"
                                        }
                                    ]
                                }
                            ],
                            "hints": [
                                {
                                    "value": 1,
                                    "class": "sbg:CPURequirement"
                                },
                                {
                                    "value": 1000,
                                    "class": "sbg:MemRequirement"
                                },
                                {
                                    "dockerImageId": "",
                                    "class": "DockerRequirement",
                                    "dockerPull": "images.sbgenomics.com/marouf/sbg-pileup-merge-out:0.1"
                                }
                            ],
                            "sbg:revision": 12,
                            "baseCommand": [
                                ""
                            ],
                            "id": "https://api.sbgenomics.com/v2/apps/bix-demo/sbgtools-demo/sbg-samtools-mpileup-merge-out/12/raw/",
                            "sbg:latestRevision": 12,
                            "outputs": [
                                {
                                    "type": [
                                        "null",
                                        {
                                            "type": "array",
                                            "items": "File"
                                        }
                                    ],
                                    "label": "output_file",
                                    "outputBinding": {
                                        "glob": {
                                            "script": "{\n  \n  \n   if($job.inputs.output_state.includes(\"Merge\"))\n   {\n    \n    format = $job.inputs.file_list[0].path.split(\"/\").pop()\n    format = format.split('.').pop()\n\n  \treturn \"*.merged.\"+format.toLowerCase()\n   }\n  else\n  {\n    return \"{*.pileup,*.vcf,*.bcf}\"\n  }\n \n}",
                                            "engine": "#cwl-js-engine",
                                            "class": "Expression"
                                        },
                                        "outputEval": {
                                            "script": "{\n  \n  var inputs = {};\n\n  to_ret = \"\"\n\n  for ( fn in $job.inputs.file_list)\n  {\n    if($job.inputs.file_list[fn].size!==0)\n    {\n\n      name = $job.inputs.file_list[fn].path.split(\"/\").pop()  \n       \n\n\t if(\"metadata\" in $job.inputs.file_list[fn])\n     { \n     \n      inputs[name] = $job.inputs.file_list[fn].metadata\n      \n     }\n\n    }\n  }\n             \n  content = $self\n\n  if($job.inputs.output_state.includes(\"Merge\"))\n  {  \n    \n    // set metadata from any input to this tool\n    name =  content[0][\"name\"]\n    \n    if (typeof name !== 'undefined')\n    {\n      metadata = inputs[name]\n      \n\t if(typeof metadata !== 'undefined')\n     {\n      if (\"ScatteredUsing\" in metadata)\n\t\tdelete metadata[\"ScatteredUsing\"]\n        \n      for ( var fp in content )\n       {\n\n          content[fp][\"metadata\"] = metadata\n\n       }\n     }\n    }\n  \n  }\n   else\n   {\n     for ( var fp in content )\n     {\n\n       name =  content[fp][\"name\"]\n       \n       if (name in inputs)\n       {\n         metadata = inputs[name]\n\n         content[fp][\"metadata\"] = metadata\n       }  \n      \n     }\n\n     return content\n   }\n\n  return content\n}\n\n\n\n\n\n\n\n",
                                            "engine": "#cwl-js-engine",
                                            "class": "Expression"
                                        }
                                    },
                                    "sbg:fileTypes": "PILEUP,BCF,VCF",
                                    "id": "#output_file",
                                    "description": "File obtained by merging files form #file_list."
                                }
                            ],
                            "successCodes": [],
                            "sbg:toolAuthor": "Mohamed Marouf, Seven Bridges Genomics, <mohamed.marouf@sbgenomics.com>",
                            "sbg:revisionNotes": "fix metadata issue when there is not metadata at input files",
                            "arguments": [
                                {
                                    "prefix": "",
                                    "position": 3,
                                    "valueFrom": {
                                        "script": "{\n\n  if($job.inputs.output_state.includes(\"Merge\"))\n  {\n    to_ret = \"\"\n    for(f in $job.inputs.contig_order_names)\n    {\n\n      for ( fn in $job.inputs.file_list)\n      {\n        if($job.inputs.file_list[fn].metadata)\n        {\n\n\n          if($job.inputs.contig_order_names[f] == $job.inputs.file_list[fn].metadata.ScatteredUsing)\n          {\n            if(to_ret == \"\")\n              to_ret =  $job.inputs.file_list[fn].path\n            else\n    \t\t  to_ret =  to_ret + \",\"+ $job.inputs.file_list[fn].path\n\n          }\n        }\n      }\n\n    }\n    \n    format = $job.inputs.file_list[0].path.split(\"/\").pop()\n    format = format.split('.').pop()\n    \n    \n    if(to_ret==\"\")\n    \treturn \n    else\n      \treturn \"python merge.py --output-format \" + format.toLowerCase()+ \" --file-list \"+ to_ret\n  \n  }\n  else if($job.inputs.output_state == \"Pass nonempty files\")\n  {\n    to_ret = \"\"\n          \n    for ( fn in $job.inputs.file_list)\n      {\n        if($job.inputs.file_list[fn].size==0)\n        {\n\n          if(to_ret == \"\")\n            to_ret =  $job.inputs.file_list[fn].path\n          else\n            to_ret =  to_ret + \",\"+ $job.inputs.file_list[fn].path\n\n        }\n      }\n    \n    if(to_ret==\"\")\n    \treturn \n    else\n  \t\treturn \"python delete_empty.py --file-list \"+ to_ret\n          \n  }\n\n\n\n          \n}",
                                        "engine": "#cwl-js-engine",
                                        "class": "Expression"
                                    },
                                    "separate": true
                                }
                            ],
                            "sbg:id": "admin/sbg-public-data/sbg-samtools-mpileup-merge-out/12",
                            "sbg:job": {
                                "allocatedResources": {
                                    "cpu": 1,
                                    "mem": 1000
                                },
                                "inputs": {
                                    "contig_order_names": [
                                        "1",
                                        "2",
                                        "3"
                                    ],
                                    "file_list": [
                                        {
                                            "metadata": {
                                                "ScatteredUsing": "3"
                                            },
                                            "path": "/ovde/1994060146_RNASeq_R.Aligned.toTranscriptome.out_22.vcf",
                                            "size": 0,
                                            "class": "File",
                                            "secondaryFiles": []
                                        },
                                        {
                                            "metadata": {
                                                "SID": "111",
                                                "ScatteredUsing": "1"
                                            },
                                            "path": "1994060146_RNASeq_R.Aligned.toTranscriptome.out_11.vcf",
                                            "size": 1,
                                            "class": "File",
                                            "secondaryFiles": []
                                        },
                                        {
                                            "path": "file.txt",
                                            "size": 1,
                                            "class": "File",
                                            "secondaryFiles": []
                                        }
                                    ],
                                    "output_state": "Pass all files"
                                }
                            },
                            "stdin": "",
                            "sbg:project": "bix-demo/sbgtools-demo",
                            "description": "Main function of this tool is to provide option to the user whether to merge list of files from SAMtools Mpileup tool or to pass list of files for each contig to the output. \n\nIf MERGE option is selected, regardless of the type, files will be merged following the contig order from the reference file index file or BED file. This order is provided using input contig_order_names.  So, this tool will receive list of pileup, vcf or BC files and list of files names of BED files used in creating input files. Pairing and finding the correct order will be done using metadata field  ScatteredUsing.\n\n####Common issue:\nThis tools is built specially for mpileup parallel and Varscan pipelines, so please don't use it outside these pipeline unless you are sure how to use it. Contact our support team for information.",
                            "sbg:createdOn": 1450911287,
                            "cwlVersion": "sbg:draft-2",
                            "sbg:contributors": [
                                "milan.domazet.sudo",
                                "bix-demo",
                                "marouf"
                            ],
                            "sbg:modifiedOn": 1482421824,
                            "sbg:sbgMaintained": false,
                            "stdout": ""
                        },
                        "inputs": [
                            {
                                "id": "#SBG_SAMtools_Merge_Mpileup.file_list",
                                "source": [
                                    "#SAMtools_Mpileup.output_pileup_vcf_or_bcf_file"
                                ]
                            },
                            {
                                "id": "#SBG_SAMtools_Merge_Mpileup.contig_order_names",
                                "source": [
                                    "#SBG_Prepare_Intervals.names"
                                ]
                            },
                            {
                                "id": "#SBG_SAMtools_Merge_Mpileup.output_state",
                                "source": [
                                    "#output_state"
                                ]
                            }
                        ],
                        "id": "#SBG_SAMtools_Merge_Mpileup",
                        "sbg:y": 447
                    }
                ],
                "sbg:toolAuthor": "Heng Li, Sanger Institute",
                "sbg:revisionNotes": "fix metadata issue when there is no metadata at inputs",
                "sbg:modifiedBy": "marouf",
                "sbg:id": "admin/sbg-public-data/samtools-mpileup-parallel/22",
                "sbg:canvas_y": -51,
                "sbg:project": "marouf/samtools-mpileup-parallel-demo",
                "description": "The purpose of this workflow is to allow user to execute the SAMtools Mpileup tool in parallel using all of the instance resources. \n\nInstead of analyzing whole BAM at once, this workflow obtains contigs information from FASTA reference index (FAI). This file contains a list of contigs where pileup or BCF should be generated. This is done by using parameter (-I or --positions FILE) of the SAMtools Mpileup tool. Jobs will be then executed per chromosome on each of the available core of the instance, while all small contigs will be executed together in a single job on one core.\n\nAs a result, execution time will be order of magnitude faster compared to the execution of the whole BAM at once. Also, output of the workflow can be an integral file for the whole BAM, or list of files per chromosome which is defined by the parameter \"output_state\". Finally, user can choose format of the output file or list of files by the SAMtools mpileup parameter output_type. Available output formats are PILEUP, VCF and BCF.\n\n######Required inputs:\n\n- Reference file-The faidx-indexed reference file in the FASTA format.\n\n- BAM file-One or a list of input BAM files.\n\n- FAI file-One of these two is required input. FAI file is used to extract intervals that will be used to scatter the SAMtools mpileup job. FAI file can be generated using SAMtools faidx. \n\n- Merging mode-This defines the way this pipeline outputs its results. Options are:\n      - \"Merge\": to produce VCF, PILEUP, or BCF file acquired after merging results of the scattered pileup.\n      - \"Pass nonempty files\"  to pass through all nonempty PILEUP, VF, BCF files without merging.\n      - \"Pass all files\": to output all PILEUP, VF, BCF Separately.\n\n- Output file format -This parameter can be used to change the output format (available choices are VCF/BCF and pileup)\n\n###### Outputs:\nThis pipeline has one output port and the output depends on the selected input parameters (merge mode and output file format). Please refer to these parameters for more details.\n\n###### Common Issues\nBAM files should be sorted by coordinates before using SAMmpileup.",
                "sbg:createdOn": 1474020918,
                "sbg:canvas_zoom": 1,
                "outputs": [
                    {
                        "type": [
                            "null",
                            "File"
                        ],
                        "sbg:x": 1257,
                        "label": "output_pileup",
                        "source": [
                            "#SBG_SAMtools_Merge_Mpileup.output_file"
                        ],
                        "required": false,
                        "id": "#output_file",
                        "sbg:y": 447,
                        "sbg:includeInPorts": true
                    }
                ],
                "cwlVersion": "sbg:draft-2",
                "sbg:contributors": [
                    "milan.domazet.sudo",
                    "marouf"
                ],
                "sbg:sbgMaintained": false
            },
            "inputs": [
                {
                    "id": "#SAMtools_Mpileup_parallel_1.bams",
                    "source": [
                        "#BamTools_Index_Normal.output_bam_file"
                    ]
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.reference_fasta",
                    "source": [
                        "#SAMtools_Index_FASTA.output_fasta_file"
                    ]
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.fai_file",
                    "source": [
                        "#SAMtools_Index_FASTA.generated_index"
                    ]
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.skip_bases_with_baq_smaller_than_defined"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.required_flags"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.region_pileup_generation"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.recalc_BAQ_fly"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.per_sample_mF"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.parameter_for_adjusting_mapq"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.output_uncompressed_bcf_or_vcf"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.output_mapping_quality"
                },
                {
                    "default": "PILEUP",
                    "id": "#SAMtools_Mpileup_parallel_1.output_format"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.output_base_positions_on_reads"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.no_indel_calling"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.more_info_to_output"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.minimum_fraction_of_gapped_reads"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.min_gapped_reads_for_indel"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.max_per_bam_depth"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.max_idepth"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.mapq_threshold"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.ignore_rg_tags"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.ignore_overlaps"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.homopolymer_err_coeficient"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.gap_open_sequencing_error_probability"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.gap_extension_seq_error_probability"
                },
                {
                    "default": true,
                    "id": "#SAMtools_Mpileup_parallel_1.filter_zero_coverage_lines"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.filter_flags"
                },
                {
                    "default": true,
                    "id": "#SAMtools_Mpileup_parallel_1.disable_baq_computation"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.count_anomalous_read_pairs"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.comma_separated_list_of_platforms_for_indels"
                },
                {
                    "id": "#SAMtools_Mpileup_parallel_1.assume_the_quality_Illumina_encoding"
                },
                {
                    "default": "Pass nonempty files",
                    "id": "#SAMtools_Mpileup_parallel_1.output_state"
                }
            ],
            "id": "#SAMtools_Mpileup_parallel_1",
            "sbg:y": 431.46393369353075
        }
    ],
    "sbg:toolkitVersion": "2.3.9",
    "sbg:toolAuthor": "Genome Institute at Washington University",
    "sbg:appVersion": [
        "sbg:draft-2"
    ],
    "sbg:revisionNotes": "fix metadata issue when it is empty",
    "sbg:modifiedBy": "bixqa",
    "sbg:id": "bixqa/qa-load-2017-07-31-18/varscan2-workflow-from-bam-v2-3-9/29",
    "sbg:canvas_y": 30,
    "sbg:projectName": "qa-load-2017-07-31-18",
    "sbg:project": "bixqa/qa-load-2017-07-31-18",
    "description": "VarScan workflow is extended methodology suggested by the authors, where user can start analysis from BAM files for tumor and normal samples. \nFirst step in the workflow is to prepare pileups for the both samples. For this purpose, parallel realization of the SAMtools mpileup tool is used, where BAM files are splitted into smaller files using the intervals provided in the FAI file. It then runs the mpileup on each file in parallel. \n\nAfterwards, the VarScan Somatic is used. This tool accepts pileups for tumor and normal samples and calls for indels and SNPs in these files. Scatter option is enabled in this tool, so each tumor pileup file with its interval corresponding normal pileup file resulted from the previous step will be considered as single job.\n\nNext downstream tool is VarScan SomaticFilter which is also designed to perform analysis per chromosome. So, indels and SNP's files from the previous step are paired according to the chromosome they are located in. Simple filtering is then done to remove SNPs around INDELs (refer to publication for more details). \n\nClassification of mutations using VarsScan2 ProcessSomatic tool is done for all mutations (SNPs and INDELs). Mutations are segregated into three groups: germline, LOH (loss of heterozygosity), and somatic. Each of these groups will have two files on the tool output: one contains all variants from this group; the second is the subset of these variants that is considered high confidence. The output format of the tool is VCF by default, where only high confidence (HC) mutations are taken into consideration for the further downstream analysis.  \n\nSince scattering of the pipeline is done per chromosome, it is necessary to concat resulting VCF files, sort and reorder them. This is done using VCFtool Concat and Sort, while the reordering is performed by a custom Seven Bridges tool SBG Reorder VCF.\n\n##### Required inputs:\n1. A Reference FASTA file (ID: *input_fasta_file*) - The reference file in the FASTA format.\n2. Input tumor BAM file (ID: *Tumor_bam_file*) \n2. Input normal BAM file (ID: *Normal_bam_file*) \n\n\n##### Outputs:\nTwo VCF files are outputted from this pipeline.\n1.  Somatic INDELs (ID: *High_Confidence_INDELs*)\n2. Somatic SNPs (ID: *High_Confidence_SNPs*)\n\n##### Common issues:\n1. BAM files should be coordinates sorted.\n2. The input FASTA file should be the same as the one used for mapping the input tumor/normal BAM files otherwise the pipeline return an error like \"Inputs are null for parallelization\".\n3. BAQ calculations were disable as the authors of Varscan proposed.\n4.Please notice that you should set the instance according to the input BAM files. Instance should have enough memory to store the intermediate pileup files for the input BAMs.\n5. zero-coverage lines of pileup files are filtered, this is important to avoid any varscan failures.",
    "sbg:createdOn": 1501518730,
    "sbg:canvas_zoom": 0.7499999999999998,
    "outputs": [
        {
            "type": [
                "null",
                "File"
            ],
            "sbg:x": 1992.6351702929237,
            "label": "#High_Confidence_SNPs",
            "source": [
                "#SBG_VCF_Reorder.reordered_vcf"
            ],
            "required": false,
            "sbg:fileTypes": "VCF",
            "id": "#High_Confidence_SNPs",
            "sbg:y": 309.12334402230044,
            "sbg:includeInPorts": true
        },
        {
            "type": [
                "null",
                "File"
            ],
            "sbg:x": 1985.5430476483887,
            "label": "#High_Confidence_INDELs",
            "source": [
                "#SBG_VCF_Reorder_1.reordered_vcf"
            ],
            "required": false,
            "sbg:fileTypes": "VCF",
            "id": "#High_Confidence_INDELs",
            "sbg:y": 453.6660233470865,
            "sbg:includeInPorts": true
        }
    ],
    "sbg:links": [
        {
            "label": "Homepage",
            "id": "http://varscan.sourceforge.net/"
        },
        {
            "label": "Sourcecode",
            "id": "https://github.com/dkoboldt/varscan"
        },
        {
            "label": "Wiki",
            "id": "http://sourceforge.net/p/varscan/wiki/Home/"
        },
        {
            "label": "Download",
            "id": "http://sourceforge.net/projects/varscan/files/latest/download?source=files"
        },
        {
            "label": "Publication",
            "id": "http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2734323/"
        },
        {
            "label": "Documentation",
            "id": "http://varscan.sourceforge.net/using-varscan.html"
        }
    ],
    "sbg:contributors": [
        "bixqa"
    ],
    "sbg:copyOf": "admin/sbg-public-data/varscan2-workflow-from-bam-v2-3-9/29",
    "sbg:sbgMaintained": false
}